{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "90a88efe-ebb3-424c-8704-fb4ee189af2c",
   "metadata": {},
   "source": [
    "# OLS Library Preprocessing Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d058003-a667-4b45-9ddc-e47efe48e0e1",
   "metadata": {
    "tags": []
   },
   "source": [
    "**Authorship:**\n",
    "Adam Klie, *09/28/2021*\n",
    "***\n",
    "**Description:**\n",
    "This notebook can be used to preprocess the OLS library dataset for training machine learning (ML) models on MPRA data. So far, there are a total of **6** feature sets to generate:\n",
    "\n",
    "1. block encoding\n",
    "2. fasta sequences\n",
    "3. mixed-1.0 encoding\n",
    "4. mixed-2.0 encoding\n",
    "5. mixed-3.0 encoding\n",
    "5. one-hot encoding (forward and reverse complement)\n",
    "\n",
    "The general worklow is as follows:\n",
    "1. Read in dataset\n",
    "2. Perform train-test splitting\n",
    "3. Save relevant information (sequence IDs, sequence, lables etc.)\n",
    "4. Perform feature encoding/engineering/selection\n",
    "5. Save full encoding, as well as standardized train and test set to specified directory\n",
    "\n",
    "Details on the encodings can be found in the appropriate sections. Make sure to use a standard naming convention. Here I am using:\n",
    "`PREPROCESS_SET_ENCODING.extension`.\n",
    "\n",
    "An example: `0.18-0.4_X-train-0.9_block.npy`. This file contain a 0.9 split of the total sample set as a training set with the *block* encoding. These samples where preprocessed with the 0.18 and 0.4 low and high activity cut-offs respectively.\n",
    "\n",
    "Note that separation with `_` versus the `-`. `-` separates descriptors of the main `_` separated field.\n",
    "***\n",
    "**TODOs:**\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca7c07f-c31c-4d7d-b83d-264c9c3e4ffd",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Set-up"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df553e20-fb54-4a78-8fff-ec45f958e2b7",
   "metadata": {},
   "source": [
    "## Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5cb2a818-9f0c-4c1b-856a-984d3e59ad9a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T16:46:38.428847Z",
     "start_time": "2021-05-25T16:46:18.911625Z"
    },
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:13.588718Z",
     "iopub.status.busy": "2022-06-04T19:05:13.588493Z",
     "iopub.status.idle": "2022-06-04T19:05:25.938379Z",
     "shell.execute_reply": "2022-06-04T19:05:25.937962Z",
     "shell.execute_reply.started": "2022-06-04T19:05:13.588677Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Classics\n",
    "import os\n",
    "import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Helpful libraries\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder\n",
    "\n",
    "# Autoreload extension\n",
    "if 'autoreload' not in get_ipython().extension_manager.loaded:\n",
    "    %load_ext autoreload\n",
    "\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3685fa33-16af-489b-8d5d-624d6e137389",
   "metadata": {},
   "source": [
    "## Notebook parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "32127569-3c9d-4476-b231-482a9bdb6e71",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:26.064808Z",
     "iopub.status.busy": "2022-06-04T19:05:26.064687Z",
     "iopub.status.idle": "2022-06-04T19:05:26.658581Z",
     "shell.execute_reply": "2022-06-04T19:05:26.658204Z",
     "shell.execute_reply.started": "2022-06-04T19:05:26.064792Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Param definitions\n",
    "ACTIVE_LOW = np.inf  # The activity below which to define inactive enhancers\n",
    "ACTIVE_HIGH = -np.inf  # The activity above which to define active enhancers\n",
    "PREPROCESS = \"{}-{}\".format(ACTIVE_LOW, ACTIVE_HIGH)  # String defining the preprocessing for saving\n",
    "SPLIT = 0.9  # Split into training and test sets\n",
    "SUBSET = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88fd5092-fe40-4c4f-95d1-1e1f596af80d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load data to preprocess\n",
    "**Note**: There are two versions of the dataset I have to work with. See shared Google Drive for version history. The data has the following descriptors:\n",
    "1. **NAME** - Site list \n",
    "2. **SEQUENCE** - Sequence\n",
    "3. **MPRA_FXN** - Functional by MPRA (see slide 2)\n",
    "    - Activity >.4 = functional (1)\n",
    "    - Activity <.18 = non-functional (0)\n",
    "    - If activity is in between => ‘na’\n",
    "4. **MICROSCPOPE_FXN** - Microscope functional group\n",
    "    - Only be for experiments with low variance and good controls (lacking *)\n",
    "5. **ACTIVITY_SUMRNA_NUMDNA**\n",
    "    - Activity score => sum MRPM all RNA barcodes / number DNA barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5464b73b-deaa-4416-b75e-043c1a65e3d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T16:46:50.740320Z",
     "start_time": "2021-05-25T16:46:49.949482Z"
    },
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:29.581438Z",
     "iopub.status.busy": "2022-06-04T19:05:29.581291Z",
     "iopub.status.idle": "2022-06-04T19:05:36.671258Z",
     "shell.execute_reply": "2022-06-04T19:05:36.670903Z",
     "shell.execute_reply.started": "2022-06-04T19:05:29.581421Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>SEQ</th>\n",
       "      <th>MPRA_FXN</th>\n",
       "      <th>MICROSCOPE_FXN</th>\n",
       "      <th>ACTIVITY_SUMRNA_NUMDNA</th>\n",
       "      <th>SEQ_LEN</th>\n",
       "      <th>linker_1</th>\n",
       "      <th>TFBS_1</th>\n",
       "      <th>linker_2</th>\n",
       "      <th>TFBS_2</th>\n",
       "      <th>linker_3</th>\n",
       "      <th>TFBS_3</th>\n",
       "      <th>linker_4</th>\n",
       "      <th>TFBS_4</th>\n",
       "      <th>linker_5</th>\n",
       "      <th>TFBS_5</th>\n",
       "      <th>linker_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S1-G1R-S2-E1F-S3-E2F-S4-G2R-S5-G3F-S6</td>\n",
       "      <td>CATCTGAAGCTCGTTATCTCTAACGGAAGTTTTCGAAAAGGAAATT...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Neural Enhancer</td>\n",
       "      <td>0.611767</td>\n",
       "      <td>66</td>\n",
       "      <td>S1</td>\n",
       "      <td>G1R</td>\n",
       "      <td>S2</td>\n",
       "      <td>E1F</td>\n",
       "      <td>S3</td>\n",
       "      <td>E2F</td>\n",
       "      <td>S4</td>\n",
       "      <td>G2R</td>\n",
       "      <td>S5</td>\n",
       "      <td>G3F</td>\n",
       "      <td>S6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    NAME  \\\n",
       "0  S1-G1R-S2-E1F-S3-E2F-S4-G2R-S5-G3F-S6   \n",
       "\n",
       "                                                 SEQ  MPRA_FXN  \\\n",
       "0  CATCTGAAGCTCGTTATCTCTAACGGAAGTTTTCGAAAAGGAAATT...       1.0   \n",
       "\n",
       "    MICROSCOPE_FXN  ACTIVITY_SUMRNA_NUMDNA  SEQ_LEN linker_1 TFBS_1 linker_2  \\\n",
       "0  Neural Enhancer                0.611767       66       S1    G1R       S2   \n",
       "\n",
       "  TFBS_2 linker_3 TFBS_3 linker_4 TFBS_4 linker_5 TFBS_5 linker_6  \n",
       "0    E1F       S3    E2F       S4    G2R       S5    G3F       S6  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load from tsv provided by Joe\n",
    "OLS_data = pd.read_csv(\n",
    "    \"20210728-3.EnhancerTable.ForAdam.FunctionalEnhancers.WT-detected.ABL-notDetected.10R-20U-0.1P.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    na_values=\"na\",\n",
    ")\n",
    "\n",
    "# Add sequence length\n",
    "OLS_data = OLS_data.rename({\"SEQUENCE\":\"SEQ\"}, axis=1)\n",
    "OLS_data[\"SEQ_LEN\"] = OLS_data[\"SEQ\"].apply(len)\n",
    "\n",
    "# Define a set black features for current lack of a better term\n",
    "block_features = [\"linker_1\", \"TFBS_1\", \"linker_2\", \"TFBS_2\", \"linker_3\", \"TFBS_3\", \"linker_4\", \"TFBS_4\", \"linker_5\", \"TFBS_5\", \"linker_6\"]\n",
    "\n",
    "# Add these as columns to the dataframe\n",
    "OLS_data[block_features] = OLS_data[\"NAME\"].str.split(\"-\").to_list()\n",
    "\n",
    "# Save non-tiled dataframe in standard format\n",
    "OLS_data.to_csv(\"2021_OLS_Library.tsv\", index=False, sep=\"\\t\")\n",
    "\n",
    "# Check it\n",
    "OLS_data.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0d8cda25-d46e-4ac2-858b-71c5c6519402",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:36.671964Z",
     "iopub.status.busy": "2022-06-04T19:05:36.671853Z",
     "iopub.status.idle": "2022-06-04T19:05:37.572227Z",
     "shell.execute_reply": "2022-06-04T19:05:37.571774Z",
     "shell.execute_reply.started": "2022-06-04T19:05:36.671948Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0    208729\n",
       " NaN    157864\n",
       " 1.0     94207\n",
       " Name: MPRA_FXN, dtype: int64,\n",
       " 1.0    460777\n",
       " NaN        23\n",
       " Name: FXN_LABEL, dtype: int64)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add labels based on decided cutoffs of activity\n",
    "OLS_data[\"FXN_LABEL\"] = np.nan\n",
    "OLS_data.loc[OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"] <= ACTIVE_LOW, \"FXN_LABEL\"] = 0\n",
    "OLS_data.loc[OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"] >= ACTIVE_HIGH, \"FXN_LABEL\"] = 1\n",
    "\n",
    "# Impute activity of 23 NaN and 25 inf enhancers to mean avoid regression errors\n",
    "OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"] = OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"].replace([np.inf, -np.inf], np.nan)\n",
    "OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"] = OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"].fillna(OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"].mean())\n",
    "\n",
    "# Sanity check to make sure things match up with what Joe did if using the 0.18-0.4 thresholds\n",
    "OLS_data[\"MPRA_FXN\"].value_counts(dropna=False), OLS_data[\"FXN_LABEL\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7e0afac-d81c-487f-aa3f-d9334c29c482",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:37.573129Z",
     "iopub.status.busy": "2022-06-04T19:05:37.573014Z",
     "iopub.status.idle": "2022-06-04T19:05:38.296917Z",
     "shell.execute_reply": "2022-06-04T19:05:38.296577Z",
     "shell.execute_reply.started": "2022-06-04T19:05:37.573112Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23, 78, 101)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Hold out enhancer with good labels already and middle \"ambiguous\" enhancers\n",
    "holdout_mask = ((OLS_data[\"FXN_LABEL\"].isna()) | (~OLS_data[\"MICROSCOPE_FXN\"].isna()))\n",
    "\n",
    "# How many seqs with annotated microscope function\n",
    "sum(OLS_data[\"FXN_LABEL\"].isna()), sum(~OLS_data[\"MICROSCOPE_FXN\"].isna()), sum(holdout_mask)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cace238-5336-4738-b960-34d28b53fff2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Get indexes for each set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fff13bc-35a2-4031-9fdc-ce05ae51d01a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **All sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "14eaf10b-5c60-4ae0-bcfb-38f694e059a2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:38.297637Z",
     "iopub.status.busy": "2022-06-04T19:05:38.297527Z",
     "iopub.status.idle": "2022-06-04T19:05:38.990707Z",
     "shell.execute_reply": "2022-06-04T19:05:38.990336Z",
     "shell.execute_reply.started": "2022-06-04T19:05:38.297621Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((460800,), (460800,), (460800,), (460800,))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All sequences\n",
    "binary = OLS_data[\"FXN_LABEL\"].values\n",
    "activity = OLS_data[\"ACTIVITY_SUMRNA_NUMDNA\"].values\n",
    "ID = OLS_data[\"NAME\"].values\n",
    "X_seqs = OLS_data[\"SEQ\"].values\n",
    "binary.shape, activity.shape, ID.shape, X_seqs.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "823b0f3a-62a8-45ac-aa04-9b99311f06d7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Deployment sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4eafe8e1-a2ce-46c4-88d0-e68c5693b54f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:05:38.991562Z",
     "iopub.status.busy": "2022-06-04T19:05:38.991426Z",
     "iopub.status.idle": "2022-06-04T19:05:39.802553Z",
     "shell.execute_reply": "2022-06-04T19:05:39.802214Z",
     "shell.execute_reply.started": "2022-06-04T19:05:38.991526Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((460699,), (460699,), (460699,))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove ambiguous enhancers (aka the ones with NaNs). This will be the final training data\n",
    "training_index = OLS_data[~holdout_mask].index\n",
    "binary_training = binary[training_index]\n",
    "activity_training = activity[training_index]\n",
    "binary_training.shape, activity_training.shape, training_index.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1984bdd9-3a7e-4cee-8d33-70335053e046",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Train/test split sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4f4e4134-9101-4403-958e-dc94db2ea278",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:17.581718Z",
     "iopub.status.busy": "2022-06-04T19:07:17.581552Z",
     "iopub.status.idle": "2022-06-04T19:07:18.329881Z",
     "shell.execute_reply": "2022-06-04T19:07:18.329468Z",
     "shell.execute_reply.started": "2022-06-04T19:07:17.581700Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from eugene.preprocessing import split_train_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02840999-921f-41f9-ab48-1a5ada83000e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:19.581799Z",
     "iopub.status.busy": "2022-06-04T19:07:19.581659Z",
     "iopub.status.idle": "2022-06-04T19:07:20.296526Z",
     "shell.execute_reply": "2022-06-04T19:07:20.296229Z",
     "shell.execute_reply.started": "2022-06-04T19:07:19.581783Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((414629,), (46070,), (414629,), (46070,), (414629,), (46070,))"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the indexes and the targets for a train set and a test set\n",
    "train_index, test_index, binary_train, binary_test = split_train_test(training_index, binary_training, split=SPLIT, subset=SUBSET)\n",
    "activity_train = activity[train_index]\n",
    "activity_test = activity[test_index]\n",
    "train_index.shape, test_index.shape, binary_train.shape, binary_test.shape, activity_train.shape, activity_test.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d102c9a1-4890-4f7a-9e7c-ed0f9e162d54",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Holdout Sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8454badd-1307-4a26-a1f5-3a9b52e06ec1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:21.791710Z",
     "iopub.status.busy": "2022-06-04T19:07:21.791562Z",
     "iopub.status.idle": "2022-06-04T19:07:22.511444Z",
     "shell.execute_reply": "2022-06-04T19:07:22.511122Z",
     "shell.execute_reply.started": "2022-06-04T19:07:21.791693Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((101,), (101,), (101,))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the heldout data as well\n",
    "holdout_index = OLS_data[holdout_mask].index\n",
    "binary_holdout = binary[holdout_index]\n",
    "activity_holdout = activity[holdout_index]\n",
    "binary_holdout.shape, activity_holdout.shape, holdout_index.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fce99500-1ffc-45fb-a638-730a986d3da0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **!!!Sanity Check!!!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5c02041d-9914-4589-945f-ec881398fd68",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:22.605334Z",
     "iopub.status.busy": "2022-06-04T19:07:22.605213Z",
     "iopub.status.idle": "2022-06-04T19:07:23.233971Z",
     "shell.execute_reply": "2022-06-04T19:07:23.233660Z",
     "shell.execute_reply.started": "2022-06-04T19:07:22.605318Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True, True, True, True, True)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(train_index) + len(test_index) == len(training_index), \\\n",
    "len(training_index) + len(holdout_index) == len(OLS_data), \\\n",
    "len(binary_train) + len(binary_test) == len(binary_training), \\\n",
    "len(binary_training) + len(binary_holdout) == len(binary), \\\n",
    "len(activity_train) + len(activity_test) == len(activity_training), \\\n",
    "len(activity_training) + len(activity_holdout) == len(activity)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e9f975-5e6b-4e36-92f2-5e6bc9313a52",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Save relavant information (labels, IDs, sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88130e70-0743-4c73-bcd7-b634434ea9e9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Binary Labels**\n",
    "0 (non-functional) and 1 (functional). These are all the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "586677ae-76b5-44be-8bc0-593d83ff6ed7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:24.030528Z",
     "iopub.status.busy": "2022-06-04T19:07:24.030280Z",
     "iopub.status.idle": "2022-06-04T19:07:27.879095Z",
     "shell.execute_reply": "2022-06-04T19:07:27.878733Z",
     "shell.execute_reply.started": "2022-06-04T19:07:24.030497Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  460800 binary/0.09-0.4_y-all_binary.txt\n",
      "  263252 binary/0.09-0.4_y-holdout_binary.txt\n",
      "   19755 binary/0.09-0.4_y-test-0.1_binary.txt\n",
      "  177793 binary/0.09-0.4_y-train-0.9_binary.txt\n",
      "  197548 binary/0.09-0.4_y-training_binary.txt\n",
      "  460800 binary/0.18-0.4_y-all_binary.txt\n",
      "   30288 binary/0.18-0.4_y-test-0.1_binary.txt\n",
      "  272591 binary/0.18-0.4_y-train-0.9_binary.txt\n",
      "  302879 binary/0.18-0.4_y-training_binary.txt\n",
      "  460800 binary/inf--inf_y-all_binary.txt\n",
      "     101 binary/inf--inf_y-holdout_binary.txt\n",
      "   46070 binary/inf--inf_y-test-0.1_binary.txt\n",
      "  414629 binary/inf--inf_y-train-0.9_binary.txt\n",
      "  460699 binary/inf--inf_y-training_binary.txt\n",
      " 3568005 total\n"
     ]
    }
   ],
   "source": [
    "# Save the labels\n",
    "if not os.path.isdir(\"binary\"):\n",
    "    os.makedirs(\"binary\")  \n",
    "\n",
    "np.savetxt(\"binary/{}_y-all_binary.txt\".format(PREPROCESS), X=binary, fmt=\"%f\")\n",
    "np.savetxt(\"binary/{}_y-training_binary.txt\".format(PREPROCESS), X=binary_training, fmt=\"%f\")\n",
    "np.savetxt(\"binary/{}_y-train-{}_binary.txt\".format(PREPROCESS, SPLIT), X=binary_train, fmt=\"%f\")\n",
    "np.savetxt(\"binary/{}_y-test-{}_binary.txt\".format(PREPROCESS, round(1-SPLIT, 1)), X=binary_test, fmt=\"%f\")\n",
    "np.savetxt(\"binary/{}_y-holdout_binary.txt\".format(PREPROCESS), X=binary_holdout, fmt=\"%f\")\n",
    "!wc -l binary/*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b4dfd91-f671-4fc5-9cf3-bad340029b54",
   "metadata": {},
   "source": [
    "## **Continuous Labels**\n",
    "Normalized MPRA counts of RNA reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ed79c846-ebd6-4ce6-886f-c57fd01bce12",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:27.879993Z",
     "iopub.status.busy": "2022-06-04T19:07:27.879865Z",
     "iopub.status.idle": "2022-06-04T19:07:31.645441Z",
     "shell.execute_reply": "2022-06-04T19:07:31.645016Z",
     "shell.execute_reply.started": "2022-06-04T19:07:27.879974Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  460800 activity/0.09-0.4_y-all_activity.txt\n",
      "  263252 activity/0.09-0.4_y-holdout_activity.txt\n",
      "   19755 activity/0.09-0.4_y-test-0.1_activity.txt\n",
      "  177793 activity/0.09-0.4_y-train-0.9_activity.txt\n",
      "  197548 activity/0.09-0.4_y-training_activity.txt\n",
      "  460800 activity/inf--inf_y-all_activity.txt\n",
      "     101 activity/inf--inf_y-holdout_activity.txt\n",
      "   46070 activity/inf--inf_y-test-0.1_activity.txt\n",
      "  414629 activity/inf--inf_y-train-0.9_activity.txt\n",
      "  460699 activity/inf--inf_y-training_activity.txt\n",
      " 2501447 total\n"
     ]
    }
   ],
   "source": [
    "# Save the labels\n",
    "if not os.path.isdir(\"activity\"):\n",
    "    os.makedirs(\"activity\")  \n",
    "\n",
    "np.savetxt(\"activity/{}_y-all_activity.txt\".format(PREPROCESS), X=activity, fmt=\"%f\")\n",
    "np.savetxt(\"activity/{}_y-training_activity.txt\".format(PREPROCESS), X=activity_training, fmt=\"%f\")\n",
    "np.savetxt(\"activity/{}_y-train-{}_activity.txt\".format(PREPROCESS, SPLIT), X=activity_train, fmt=\"%f\")\n",
    "np.savetxt(\"activity/{}_y-test-{}_activity.txt\".format(PREPROCESS, round(1-SPLIT, 1)), X=activity_test, fmt=\"%f\")\n",
    "np.savetxt(\"activity/{}_y-holdout_activity.txt\".format(PREPROCESS), X=activity_holdout, fmt=\"%f\")\n",
    "!wc -l activity/*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4456d4c3-aa72-43a1-8027-c24f7a407154",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Identifiers**\n",
    "L1-TFBS1-... Also all the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d08c3e2e-d4c1-4ac4-8ac4-e968d83f6eb3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:31.646616Z",
     "iopub.status.busy": "2022-06-04T19:07:31.646384Z",
     "iopub.status.idle": "2022-06-04T19:07:35.988608Z",
     "shell.execute_reply": "2022-06-04T19:07:35.988200Z",
     "shell.execute_reply.started": "2022-06-04T19:07:31.646592Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   460800 id/0.09-0.4_id-all.txt\n",
      "   263252 id/0.09-0.4_id-holdout.txt\n",
      "    19755 id/0.09-0.4_id-test-0.1.txt\n",
      "   177793 id/0.09-0.4_id-train-0.9.txt\n",
      "   197548 id/0.09-0.4_id-training.txt\n",
      "   460800 id/0.18-0.4_id-all.txt\n",
      "   157921 id/0.18-0.4_id-holdout.txt\n",
      "    30288 id/0.18-0.4_id-test-0.1.txt\n",
      "   272591 id/0.18-0.4_id-train-0.9.txt\n",
      "   302879 id/0.18-0.4_id-training.txt\n",
      "   460800 id/inf--inf_id-all.txt\n",
      "      101 id/inf--inf_id-holdout.txt\n",
      "    46070 id/inf--inf_id-test-0.1.txt\n",
      "   414629 id/inf--inf_id-train-0.9.txt\n",
      "   460699 id/inf--inf_id-training.txt\n",
      "  3725926 total\n"
     ]
    }
   ],
   "source": [
    "# Save the names of the seqs\n",
    "if not os.path.isdir(\"id\"):\n",
    "    os.makedirs(\"id\")\n",
    "    \n",
    "np.savetxt(\"id/{}_id-all.txt\".format(PREPROCESS), X=ID, fmt=\"%s\")\n",
    "np.savetxt(\"id/{}_id-training.txt\".format(PREPROCESS), X=ID[training_index], fmt=\"%s\")\n",
    "np.savetxt(\"id/{}_id-train-{}.txt\".format(PREPROCESS, SPLIT), X=ID[train_index], fmt=\"%s\")\n",
    "np.savetxt(\"id/{}_id-test-{}.txt\".format(PREPROCESS, round(1-SPLIT, 1)), X=ID[test_index], fmt=\"%s\")\n",
    "np.savetxt(\"id/{}_id-holdout.txt\".format(PREPROCESS), X=ID[holdout_index], fmt=\"%s\")\n",
    "\n",
    "!wc -l id/*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b17a5c71-cde1-4e98-9a17-bf1453809938",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **Sequences**\n",
    "ACGT..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2e23e40e-7528-4381-a067-48914587d83c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:07:35.989424Z",
     "iopub.status.busy": "2022-06-04T19:07:35.989303Z",
     "iopub.status.idle": "2022-06-04T19:07:41.205766Z",
     "shell.execute_reply": "2022-06-04T19:07:41.205309Z",
     "shell.execute_reply.started": "2022-06-04T19:07:35.989405Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   460800 seqs/0.09-0.4_seqs-all.txt\n",
      "   263252 seqs/0.09-0.4_seqs-holdout.txt\n",
      "    19755 seqs/0.09-0.4_seqs-test-0.1.txt\n",
      "   177793 seqs/0.09-0.4_seqs-train-0.9.txt\n",
      "   197548 seqs/0.09-0.4_seqs-training.txt\n",
      "   460800 seqs/0.18-0.4_seqs-all.txt\n",
      "   157921 seqs/0.18-0.4_seqs-holdout.txt\n",
      "    30288 seqs/0.18-0.4_seqs-test-0.1.txt\n",
      "   272591 seqs/0.18-0.4_seqs-train-0.9.txt\n",
      "   302879 seqs/0.18-0.4_seqs-training.txt\n",
      "   460800 seqs/inf--inf_seqs-all.txt\n",
      "      101 seqs/inf--inf_seqs-holdout.txt\n",
      "    46070 seqs/inf--inf_seqs-test-0.1.txt\n",
      "   414629 seqs/inf--inf_seqs-train-0.9.txt\n",
      "   460699 seqs/inf--inf_seqs-training.txt\n",
      "  3725926 total\n"
     ]
    }
   ],
   "source": [
    "# Save the different files as numpy objects\n",
    "if not os.path.isdir(\"seqs\"):\n",
    "    os.makedirs(\"seqs\")  \n",
    "    \n",
    "np.savetxt(\"seqs/{}_seqs-all.txt\".format(PREPROCESS), X=X_seqs, fmt=\"%s\")\n",
    "np.savetxt(\"seqs/{}_seqs-training.txt\".format(PREPROCESS), X=X_seqs[training_index], fmt=\"%s\")\n",
    "np.savetxt(\"seqs/{}_seqs-train-{}.txt\".format(PREPROCESS, SPLIT), X=X_seqs[train_index], fmt=\"%s\")\n",
    "np.savetxt(\"seqs/{}_seqs-test-{}.txt\".format(PREPROCESS, round(1-SPLIT, 1)), X=X_seqs[test_index], fmt=\"%s\")\n",
    "np.savetxt(\"seqs/{}_seqs-holdout.txt\".format(PREPROCESS), X=X_seqs[holdout_index], fmt=\"%s\")\n",
    "!wc -l seqs/*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3ced0aa-de42-405b-b072-36e9e1ec1fec",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Preprocess and save different feature sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a8258f-5457-4617-9bae-ebb03582dfed",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-07-25T17:07:34.934142Z",
     "iopub.status.busy": "2021-07-25T17:07:34.933858Z",
     "iopub.status.idle": "2021-07-25T17:07:35.592499Z",
     "shell.execute_reply": "2021-07-25T17:07:35.590921Z",
     "shell.execute_reply.started": "2021-07-25T17:07:34.934106Z"
    },
    "tags": []
   },
   "source": [
    "## **<u>Sequence feature idea 1 </u>**: One hot encoding block features\n",
    " - linker_{1-5} could be a one hot encoded vector of length 6 that can be any of L1-L5 --> e.g., [0, 1, 0, 0, 0] encodes S2\n",
    " - TFBS_{1-5} could be a one hot encoded vector of length 10 that can be G{1-3}R, G{1-3}F, E{1,2}F, E{1,2}R\n",
    " - **Note**: L6 is only S6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d97c2aa-6e2c-4325-b7cc-49e47fb87e26",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:09:35.583071Z",
     "iopub.status.busy": "2022-06-04T19:09:35.582898Z",
     "iopub.status.idle": "2022-06-04T19:09:39.330826Z",
     "shell.execute_reply": "2022-06-04T19:09:39.330514Z",
     "shell.execute_reply.started": "2022-06-04T19:09:35.583051Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([array(['S1', 'S2', 'S3', 'S4', 'S5'], dtype=object),\n",
       "  array(['E1F', 'E1R', 'E2F', 'E2R', 'G1F', 'G1R', 'G2F', 'G2R', 'G3F',\n",
       "         'G3R'], dtype=object),\n",
       "  array(['S1', 'S2', 'S3', 'S4', 'S5'], dtype=object),\n",
       "  array(['E1F', 'E1R', 'E2F', 'E2R', 'G1F', 'G1R', 'G2F', 'G2R', 'G3F',\n",
       "         'G3R'], dtype=object),\n",
       "  array(['S1', 'S2', 'S3', 'S4', 'S5'], dtype=object),\n",
       "  array(['E1F', 'E1R', 'E2F', 'E2R', 'G1F', 'G1R', 'G2F', 'G2R', 'G3F',\n",
       "         'G3R'], dtype=object),\n",
       "  array(['S1', 'S2', 'S3', 'S4', 'S5'], dtype=object),\n",
       "  array(['E1F', 'E1R', 'E2F', 'E2R', 'G1F', 'G1R', 'G2F', 'G2R', 'G3F',\n",
       "         'G3R'], dtype=object),\n",
       "  array(['S1', 'S2', 'S3', 'S4', 'S5'], dtype=object),\n",
       "  array(['E1F', 'E1R', 'E2F', 'E2R', 'G1F', 'G1R', 'G2F', 'G2R', 'G3F',\n",
       "         'G3R'], dtype=object),\n",
       "  array(['S6'], dtype=object)],\n",
       " 'S1-G1R-S2-E1R-S3-E2F-S5-G2R-S4-G3R-S6',\n",
       " array([1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1.,\n",
       "        0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,\n",
       "        0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.,\n",
       "        0., 0., 0., 0., 0., 0., 1., 1.]))"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fit to overall dataframe\n",
    "ohe_block = OneHotEncoder(sparse=False)\n",
    "X = OLS_data[block_features]\n",
    "ohe_block.fit(X)\n",
    "\n",
    "# Transform\n",
    "X_block = ohe_block.fit_transform(X)\n",
    "\n",
    "# Sanity check one\n",
    "ohe_block.categories_, OLS_data.loc[45][\"NAME\"], X_block[45]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2728ddf1-7169-4157-a39e-43184ee6a7e3",
   "metadata": {},
   "source": [
    "**Sanity Check!!!** How many features should we be expecting?\n",
    " - Each non-L6 linker has 5 options and there are 5 of these --> 5\\*5=25\n",
    " - Each TFBS has 10 options and there are 5 of these --> 10\\*5=25\n",
    " - L6 has one option S6 --> 1\n",
    " - Should have:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6e9e6b9a-3c1a-42ee-a8db-828c24d010b3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:09:39.331603Z",
     "iopub.status.busy": "2022-06-04T19:09:39.331488Z",
     "iopub.status.idle": "2022-06-04T19:09:39.980267Z",
     "shell.execute_reply": "2022-06-04T19:09:39.979936Z",
     "shell.execute_reply.started": "2022-06-04T19:09:39.331587Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expecting 76 features\n",
      "Observed 76 features\n",
      "Expecting 75 features\n",
      "Observed 75 features\n"
     ]
    }
   ],
   "source": [
    "print(\"Expecting %d features\" % ((5 * 5) + (10 * 5) + 1))\n",
    "print(\"Observed %d features\" % X_block.shape[1])\n",
    "if (5 * 5) + (10 * 5) + 1 != X_block.shape[1]:\n",
    "    print(\"Something is amiss...\")\n",
    "\n",
    "# Remove non-unique feature at end\n",
    "if len(np.unique(X_block[:, -1])) == 1:\n",
    "    X_block = X_block[:, :-1]\n",
    "\n",
    "print(\"Expecting %d features\" % ((5 * 5) + (10 * 5)))\n",
    "print(\"Observed %d features\" % X_block.shape[1])\n",
    "if (5 * 5) + (10 * 5) != X_block.shape[1]:\n",
    "    print(\"Something is amiss...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1cb1629d-12c1-4862-9b32-b7108d636957",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:09:39.981091Z",
     "iopub.status.busy": "2022-06-04T19:09:39.980923Z",
     "iopub.status.idle": "2022-06-04T19:09:43.037668Z",
     "shell.execute_reply": "2022-06-04T19:09:43.037249Z",
     "shell.execute_reply.started": "2022-06-04T19:09:39.981075Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 2183165\n",
      "-rw-r--r-- 1 aklie carter-users 276480128 May  5 18:58 0.09-0.4_X-all_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 157951328 May  5 18:58 0.09-0.4_X-holdout_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users  11853128 May  5 18:58 0.09-0.4_X-test-0.1_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 106675928 May  5 18:58 0.09-0.4_X-train-0.9_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 118528928 May  5 18:58 0.09-0.4_X-training_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 276480128 Nov 26  2021 0.18-0.4_X-all_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users  94752728 Nov 26  2021 0.18-0.4_X-holdout_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users  18172928 Nov 26  2021 0.18-0.4_X-test-0.1_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 163554728 Nov 26  2021 0.18-0.4_X-train-0.9_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 181727528 Nov 26  2021 0.18-0.4_X-training_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users       274 Jun  4 12:09 block_header.txt\n",
      "-rw-r--r-- 1 aklie carter-users 276480128 Jun  4 12:09 inf--inf_X-all_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users     60728 Jun  4 12:09 inf--inf_X-holdout_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users  27642128 Jun  4 12:09 inf--inf_X-test-0.1_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 248777528 Jun  4 12:09 inf--inf_X-train-0.9_block.npy\n",
      "-rw-r--r-- 1 aklie carter-users 276419528 Jun  4 12:09 inf--inf_X-training_block.npy\n"
     ]
    }
   ],
   "source": [
    "# Save the different files as numpy objects\n",
    "if not os.path.isdir(\"block\"):\n",
    "    os.makedirs(\"block\")  \n",
    "    \n",
    "# Save the different files as numpy arrays\n",
    "np.save(\"block/{}_X-all_block\".format(PREPROCESS), arr=X_block)\n",
    "np.save(\"block/{}_X-training_block\".format(PREPROCESS), arr=X_block[training_index, :])\n",
    "np.save(\"block/{}_X-train-{}_block\".format(PREPROCESS, SPLIT), arr=X_block[train_index, :])\n",
    "np.save(\"block/{}_X-test-{}_block\".format(PREPROCESS, round(1-SPLIT, 1)), arr=X_block[test_index, :])\n",
    "np.save(\"block/{}_X-holdout_block\".format(PREPROCESS), arr=X_block[holdout_index, :])\n",
    "\n",
    "# Save the header for use down the line. Save as one column per line\n",
    "with open (\"block/block_header.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(np.concatenate(ohe_block.categories_[:-1])))\n",
    "\n",
    "# Sanity check\n",
    "!ls -l block"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dd1a4a4-b1a8-49f3-8cff-9d78f3d8f135",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **<u>Sequence feature idea 2 </u>**: Mixed encodings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eb0ae3a4-9d93-477f-8ac5-5c4cb545fce0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:10:05.582412Z",
     "iopub.status.busy": "2022-06-04T19:10:05.582231Z",
     "iopub.status.idle": "2022-06-04T19:10:06.310723Z",
     "shell.execute_reply": "2022-06-04T19:10:06.310329Z",
     "shell.execute_reply.started": "2022-06-04T19:10:05.582391Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from eugene.preprocessing import mixed_OLS_encode\n",
    "from eugene.preprocessing import standardize_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e05c0b2a-14b9-4c4a-b2e4-8eb271a23219",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:10:09.583125Z",
     "iopub.status.busy": "2022-06-04T19:10:09.582951Z",
     "iopub.status.idle": "2022-06-04T19:10:58.275299Z",
     "shell.execute_reply": "2022-06-04T19:10:58.274972Z",
     "shell.execute_reply.started": "2022-06-04T19:10:09.583107Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "460800it [00:38, 11880.24it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "((460800, 21), (460800, 26), (460800, 21))"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_mixed1, X_mixed2, X_mixed3 = mixed_OLS_encode(OLS_data[block_features])\n",
    "X_mixed1.shape, X_mixed2.shape, X_mixed3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e5d73554-9422-468f-927b-e6421fa084c2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:10:58.276052Z",
     "iopub.status.busy": "2022-06-04T19:10:58.275936Z",
     "iopub.status.idle": "2022-06-04T19:10:59.017286Z",
     "shell.execute_reply": "2022-06-04T19:10:59.016913Z",
     "shell.execute_reply.started": "2022-06-04T19:10:58.276035Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('S1-G1R-S2-E1F-S3-E2F-S4-G2R-S5-G3F-S6',\n",
       " array([12. ,  0. ,  0. ,  0.9,  2. ,  1. ,  1. ,  0.6,  7. ,  1. ,  1. ,\n",
       "         0.4,  5. ,  0. ,  0. ,  0.3,  0. ,  0. ,  1. ,  0.5,  1. ]),\n",
       " array([12. ,  0. ,  0. ,  0.9, -1. ,  2. ,  0.6,  1. ,  0. ,  0. ,  7. ,\n",
       "         0.4,  1. ,  0. ,  0. ,  5. ,  0. ,  0. ,  0.3, -1. ,  0. ,  0. ,\n",
       "         0. ,  0.5,  1. ,  1. ]),\n",
       " array([12. ,  0. ,  0.9,  0. ,  2. ,  0.6,  0. ,  1. ,  7. ,  0.4,  0. ,\n",
       "         1. ,  5. ,  0. ,  0.3,  0. ,  0. ,  0. ,  0.5,  1. ,  1. ]))"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OLS_data.iloc[0][\"NAME\"], X_mixed1[0], X_mixed2[0], X_mixed3[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d08738-6dd8-4138-8dfb-1dd339116acf",
   "metadata": {
    "tags": []
   },
   "source": [
    "### *Mixed 1.0*\n",
    " - Replace binding sites using dictionary\n",
    " - Separate based on these binding sites and add create \"dummy variables\"\n",
    " - Get lengths of linkers around binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "53b5bef7-08f4-41df-b426-43ad66041395",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:01.581507Z",
     "iopub.status.busy": "2022-06-04T19:13:01.581341Z",
     "iopub.status.idle": "2022-06-04T19:13:04.439889Z",
     "shell.execute_reply": "2022-06-04T19:13:04.439571Z",
     "shell.execute_reply.started": "2022-06-04T19:13:01.581490Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 614739\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Nov 27  2021 0.09-0.4_X-all_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 44226464 Nov 27  2021 0.09-0.4_X-holdout_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  3318968 Nov 27  2021 0.09-0.4_X-test-0.1_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 29869352 Nov 27  2021 0.09-0.4_X-train-0.9_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      534 Nov 27  2021 0.09-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 33188192 Nov 27  2021 0.09-0.4_X-training_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      534 Nov 27  2021 0.09-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users  3523378 Oct 28  2021 0.18-0.4_mixed-1.0_eda.html\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Nov 26  2021 0.18-0.4_X-all_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 26530856 Nov 26  2021 0.18-0.4_X-holdout_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  5088512 Nov 26  2021 0.18-0.4_X-test-0.1_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 45795416 Nov 26  2021 0.18-0.4_X-train-0.9_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      534 Nov 26  2021 0.18-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 50883800 Nov 26  2021 0.18-0.4_X-training_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      534 Nov 26  2021 0.18-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Jun  4 12:13 inf--inf_X-all_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users    17096 Jun  4 12:13 inf--inf_X-holdout_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  7739888 Jun  4 12:13 inf--inf_X-test-0.1_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 69657800 Jun  4 12:13 inf--inf_X-train-0.9_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      812 Jun  4 12:13 inf--inf_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 77397560 Jun  4 12:13 inf--inf_X-training_mixed-1.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      812 Jun  4 12:13 inf--inf_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users      254 Nov 26  2021 mixed-1.0_header.txt\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([5.12276999, 0.53998225, 5.03947625, 0.53976495, 5.04224017,\n",
       "        0.54005726, 5.03799541, 0.53998563, 5.03761676, 0.54020992,\n",
       "        0.9001927 ]),\n",
       " array([4.15142491, 0.20594161, 4.13440734, 0.20593279, 4.13205647,\n",
       "        0.20595367, 4.13085135, 0.20578677, 4.13232906, 0.20594787,\n",
       "        0.29974289]),\n",
       " array([ 2.12325344e-17, -7.14948615e-16,  5.66886390e-17,  3.78895347e-16,\n",
       "        -8.03717403e-17, -9.64460884e-17,  3.83865028e-18, -7.59521514e-16,\n",
       "         7.18718718e-17, -2.80975491e-16,  4.63037190e-17]),\n",
       " array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]),\n",
       " array([5.12014786, 0.54000551, 5.03976349, 0.53999401, 5.04012598,\n",
       "        0.53999531, 5.0399914 , 0.53999423, 5.03993931, 0.54001094,\n",
       "        0.90000847]),\n",
       " array([4.15033187, 0.20591487, 4.13259336, 0.20590925, 4.1326689 ,\n",
       "        0.20591554, 4.13262294, 0.205908  , 4.13255096, 0.20591535,\n",
       "        0.29998871]),\n",
       " array([-6.88489181e-17,  7.12919442e-16, -1.89951451e-16,  2.84279404e-16,\n",
       "         7.50798686e-17,  6.69241096e-16, -3.57816958e-17,  1.32639045e-15,\n",
       "         2.56641128e-17,  1.12329848e-15, -5.70039430e-17]),\n",
       " array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]))"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset train and test\n",
    "X_mixed1_train = X_mixed1[train_index, :]\n",
    "X_mixed1_test = X_mixed1[test_index, :]\n",
    "X_mixed1_training = X_mixed1[training_index, :]\n",
    "X_mixed1_holdout = X_mixed1[holdout_index, :]\n",
    "\n",
    "# Standardize features and save both training set stats\n",
    "scale_indeces = np.array([0, 3, 4, 7, 8, 11, 12, 15, 16, 19, 20])  # Mixed 1.0\n",
    "_, X_mixed1_training_scaled = standardize_features(train_X=X_mixed1_training, test_X=X_mixed1_training, indeces=scale_indeces, stats_file=\"mixed_1.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed1_train_scaled, X_mixed1_test_scaled = standardize_features(train_X=X_mixed1_train, test_X=X_mixed1_test, indeces=scale_indeces, stats_file=\"mixed_1.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed1_train_scaled, X_mixed1_holdout_scaled = standardize_features(train_X=X_mixed1_train, test_X=X_mixed1_holdout, indeces=scale_indeces)\n",
    "\n",
    "# Save the different files as numpy objects\n",
    "if not os.path.isdir(\"mixed_1.0\"):\n",
    "    os.makedirs(\"mixed_1.0\")  \n",
    "    \n",
    "np.save(\"mixed_1.0/{}_X-all_mixed-1.0\".format(PREPROCESS), X_mixed1)\n",
    "np.save(\"mixed_1.0/{}_X-training_mixed-1.0\".format(PREPROCESS), X_mixed1_training_scaled)\n",
    "np.save(\"mixed_1.0/{}_X-train-{}_mixed-1.0\".format(PREPROCESS, SPLIT), X_mixed1_train_scaled)\n",
    "np.save(\"mixed_1.0/{}_X-test-{}_mixed-1.0\".format(PREPROCESS, round(1-SPLIT, 1)), X_mixed1_test_scaled)\n",
    "np.save(\"mixed_1.0/{}_X-holdout_mixed-1.0\".format(PREPROCESS), X_mixed1_holdout_scaled)\n",
    "    \n",
    "!ls -l mixed_1.0\n",
    "\n",
    "# !!!Sanity Check!!!\n",
    "X_mixed1_train[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed1_train[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed1_train_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed1_train_scaled[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed1_training[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed1_training[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed1_training_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed1_training_scaled[:, scale_indeces].std(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b3eefa-3d35-42de-9550-1af3830fec3e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### *Mixed 2.0*\n",
    " - Replace binding sites using dictionary\n",
    " - 4 bit vector for each binding site [ets_affinity ets_orientation gata_affinity gata_orientation] - ties together the identity to affinity\n",
    " - Get lengths of linkers around binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e37f99b2-4496-4354-833c-0eeff2cd7a24",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:06.812435Z",
     "iopub.status.busy": "2022-06-04T19:13:06.812244Z",
     "iopub.status.idle": "2022-06-04T19:13:09.804470Z",
     "shell.execute_reply": "2022-06-04T19:13:09.804132Z",
     "shell.execute_reply.started": "2022-06-04T19:13:06.812415Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 750671\n",
      "-rw-r--r-- 1 aklie carter-users 95846528 Nov 27  2021 0.09-0.4_X-all_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 54756544 Nov 27  2021 0.09-0.4_X-holdout_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  4109168 Nov 27  2021 0.09-0.4_X-test-0.1_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 36981072 Nov 27  2021 0.09-0.4_X-train-0.9_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 27  2021 0.09-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 41090112 Nov 27  2021 0.09-0.4_X-training_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 27  2021 0.09-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 95846528 Nov 26  2021 0.18-0.4_X-all_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 26530856 Nov 26  2021 0.18-0.4_X-holdout_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  6300032 Nov 26  2021 0.18-0.4_X-test-0.1_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 56699056 Nov 26  2021 0.18-0.4_X-train-0.9_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 26  2021 0.18-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 62998960 Nov 26  2021 0.18-0.4_X-training_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 26  2021 0.18-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 95846528 Jun  4 12:13 inf--inf_X-all_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users    21136 Jun  4 12:13 inf--inf_X-holdout_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  9582688 Jun  4 12:13 inf--inf_X-test-0.1_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 86242960 Jun  4 12:13 inf--inf_X-train-0.9_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      932 Jun  4 12:13 inf--inf_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 95825520 Jun  4 12:13 inf--inf_X-training_mixed-2.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      932 Jun  4 12:13 inf--inf_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users      429 Nov 26  2021 mixed-2.0_header.txt\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([5.12276999, 5.03947625, 5.04224017, 5.03799541, 5.03761676,\n",
       "        0.9001927 ]),\n",
       " array([4.15142491, 4.13440734, 4.13205647, 4.13085135, 4.13232906,\n",
       "        0.29974289]),\n",
       " array([ 2.12325344e-17,  5.66886390e-17, -8.03717403e-17,  3.83865028e-18,\n",
       "         7.18718718e-17,  4.63037190e-17]),\n",
       " array([1., 1., 1., 1., 1., 1.]),\n",
       " array([5.12014786, 5.03976349, 5.04012598, 5.0399914 , 5.03993931,\n",
       "        0.90000847]),\n",
       " array([4.15033187, 4.13259336, 4.1326689 , 4.13262294, 4.13255096,\n",
       "        0.29998871]),\n",
       " array([-6.88489181e-17, -1.89951451e-16,  7.50798686e-17, -3.57816958e-17,\n",
       "         2.56641128e-17, -5.70039430e-17]),\n",
       " array([1., 1., 1., 1., 1., 1.]))"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset train and test\n",
    "X_mixed2_train = X_mixed2[train_index, :]\n",
    "X_mixed2_test = X_mixed2[test_index, :]\n",
    "X_mixed2_training = X_mixed2[training_index, :]\n",
    "X_mixed2_holdout = X_mixed2[holdout_index, :]\n",
    "\n",
    "# Standardize features and save both training set stats\n",
    "scale_indeces = np.array([0, 5, 10, 15, 20, 25])  # Mixed 2.0\n",
    "_, X_mixed2_training_scaled = standardize_features(train_X=X_mixed2_training, test_X=X_mixed2_training, indeces=scale_indeces, stats_file=\"mixed_2.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed2_train_scaled, X_mixed2_test_scaled = standardize_features(train_X=X_mixed2_train, test_X=X_mixed2_test, indeces=scale_indeces, stats_file=\"mixed_2.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed2_train_scaled, X_mixed2_holdout_scaled = standardize_features(train_X=X_mixed2_train, test_X=X_mixed2_holdout, indeces=scale_indeces)\n",
    "\n",
    "# Save the different files as numpy objects\n",
    "if not os.path.isdir(\"mixed_2.0\"):\n",
    "    os.makedirs(\"mixed_2.0\")  \n",
    "    \n",
    "np.save(\"mixed_2.0/{}_X-all_mixed-2.0\".format(PREPROCESS), X_mixed2)\n",
    "np.save(\"mixed_2.0/{}_X-training_mixed-2.0\".format(PREPROCESS), X_mixed2_training_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-train-{}_mixed-2.0\".format(PREPROCESS, SPLIT), X_mixed2_train_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-test-{}_mixed-2.0\".format(PREPROCESS, round(1-SPLIT, 1)), X_mixed2_test_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-holdout_mixed-2.0\".format(PREPROCESS), X_mixed2_holdout_scaled)\n",
    "    \n",
    "!ls -l mixed_2.0\n",
    "\n",
    "# !!!Sanity Check!!!\n",
    "X_mixed2_train[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_train[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed2_train_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_train_scaled[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed2_training[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_training[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed2_training_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_training_scaled[:, scale_indeces].std(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e758f09-8d55-480d-adab-8c76cdec65de",
   "metadata": {
    "tags": []
   },
   "source": [
    "### *Mixed 3.0*\n",
    " - Replace binding sites using dictionary\n",
    " - 3 bit vector for each binding site [ets_affinity gata_affinity orientation] - ties together the identity to affinity while removing redundant info from mixed-2.0\n",
    " - Get lengths of linkers around binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a6c10b3d-c371-447d-a168-67e6313cc758",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:09.806396Z",
     "iopub.status.busy": "2022-06-04T19:13:09.806272Z",
     "iopub.status.idle": "2022-06-04T19:13:12.482420Z",
     "shell.execute_reply": "2022-06-04T19:13:12.482096Z",
     "shell.execute_reply.started": "2022-06-04T19:13:09.806376Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 611296\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Nov 27  2021 0.09-0.4_X-all_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 44226464 Nov 27  2021 0.09-0.4_X-holdout_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  3318968 Nov 27  2021 0.09-0.4_X-test-0.1_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 29869352 Nov 27  2021 0.09-0.4_X-train-0.9_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 27  2021 0.09-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 33188192 Nov 27  2021 0.09-0.4_X-training_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 27  2021 0.09-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Nov 26  2021 0.18-0.4_X-all_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 26530856 Nov 26  2021 0.18-0.4_X-holdout_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  5088512 Nov 26  2021 0.18-0.4_X-test-0.1_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 45795416 Nov 26  2021 0.18-0.4_X-train-0.9_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 26  2021 0.18-0.4_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 50883800 Nov 26  2021 0.18-0.4_X-training_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      414 Nov 26  2021 0.18-0.4_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 77414528 Jun  4 12:13 inf--inf_X-all_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users    17096 Jun  4 12:13 inf--inf_X-holdout_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users  7739888 Jun  4 12:13 inf--inf_X-test-0.1_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users 69657800 Jun  4 12:13 inf--inf_X-train-0.9_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      812 Jun  4 12:13 inf--inf_X-train-0.9_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users 77397560 Jun  4 12:13 inf--inf_X-training_mixed-3.0.npy\n",
      "-rw-r--r-- 1 aklie carter-users      812 Jun  4 12:13 inf--inf_X-training_stats.pickle\n",
      "-rw-r--r-- 1 aklie carter-users      319 Nov 26  2021 mixed-3.0_header.txt\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([5.12276999, 5.03947625, 5.04224017, 5.03799541, 5.03761676,\n",
       "        0.9001927 ]),\n",
       " array([4.15142491, 4.13440734, 4.13205647, 4.13085135, 4.13232906,\n",
       "        0.29974289]),\n",
       " array([ 2.12325344e-17,  5.66886390e-17, -8.03717403e-17,  3.83865028e-18,\n",
       "         7.18718718e-17,  4.63037190e-17]),\n",
       " array([1., 1., 1., 1., 1., 1.]),\n",
       " array([5.12014786, 5.03976349, 5.04012598, 5.0399914 , 5.03993931,\n",
       "        0.90000847]),\n",
       " array([4.15033187, 4.13259336, 4.1326689 , 4.13262294, 4.13255096,\n",
       "        0.29998871]),\n",
       " array([-6.88489181e-17, -1.89951451e-16,  7.50798686e-17, -3.57816958e-17,\n",
       "         2.56641128e-17, -5.70039430e-17]),\n",
       " array([1., 1., 1., 1., 1., 1.]))"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset train and test\n",
    "X_mixed3_train = X_mixed3[train_index, :]\n",
    "X_mixed3_test = X_mixed3[test_index, :]\n",
    "X_mixed3_training = X_mixed3[training_index, :]\n",
    "X_mixed3_holdout = X_mixed3[holdout_index, :]\n",
    "\n",
    "# Standardize features and save both training set stats\n",
    "scale_indeces = np.array([0, 4, 8, 12, 16, 20])  # Mixed 3.0\n",
    "_, X_mixed3_training_scaled = standardize_features(train_X=X_mixed3_training, test_X=X_mixed3_training, indeces=scale_indeces, stats_file=\"mixed_3.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed3_train_scaled, X_mixed3_test_scaled = standardize_features(train_X=X_mixed3_train, test_X=X_mixed3_test, indeces=scale_indeces, stats_file=\"mixed_3.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed3_train_scaled, X_mixed3_holdout_scaled = standardize_features(train_X=X_mixed3_train, test_X=X_mixed3_holdout, indeces=scale_indeces)\n",
    "\n",
    "# Save the different files as numpy objects\n",
    "if not os.path.isdir(\"mixed_3.0\"):\n",
    "    os.makedirs(\"mixed_3.0\")  \n",
    "    \n",
    "np.save(\"mixed_3.0/{}_X-all_mixed-3.0\".format(PREPROCESS), X_mixed3)\n",
    "np.save(\"mixed_3.0/{}_X-training_mixed-3.0\".format(PREPROCESS), X_mixed3_training_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-train-{}_mixed-3.0\".format(PREPROCESS, SPLIT), X_mixed3_train_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-test-{}_mixed-3.0\".format(PREPROCESS, round(1-SPLIT, 1)), X_mixed3_test_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-holdout_mixed-3.0\".format(PREPROCESS), X_mixed3_holdout_scaled)\n",
    "    \n",
    "!ls -l mixed_3.0\n",
    "\n",
    "# !!!Sanity Check!!!\n",
    "X_mixed3_train[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_train[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed3_train_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_train_scaled[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed3_training[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_training[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed3_training_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_training_scaled[:, scale_indeces].std(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47d0c4c4-5d78-4c44-a671-0582fc58493b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **<u>Sequence feature idea 3 </u>**: Use the actual sequence (one-hot encoded)\n",
    " - One hot encoded sequence: each position is encoded as a 1-D vector of size 4 e.g., AT is [[1,0,0,0], [0,0,0,1]]\n",
    " - Generally, we will get inputs of size (len(seq) X 4). The above example would be of size 2x4\n",
    " - Can also save the string seqs in case those are also useful down the line"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52459f96-05e9-4dad-801b-bb0e24171771",
   "metadata": {},
   "source": [
    "**Q** Are all sequences the same length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "db1c1df9-cc71-4c65-84b0-d331999724a7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:12.483359Z",
     "iopub.status.busy": "2022-06-04T19:13:12.483240Z",
     "iopub.status.idle": "2022-06-04T19:13:13.260534Z",
     "shell.execute_reply": "2022-06-04T19:13:13.260187Z",
     "shell.execute_reply.started": "2022-06-04T19:13:12.483341Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66    460800\n",
       "Name: SEQ, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the lengths of sequences to make sure they are all the same\n",
    "OLS_data[\"SEQ\"].apply(len).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aac97318-f8dd-4278-a67e-de15ea588a4d",
   "metadata": {},
   "source": [
    "**Answer**: Yes, 66bp long"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b11a3da-872e-4065-91e4-503f55ce778a",
   "metadata": {},
   "source": [
    "### *Forward Seqs*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f836918b-41fa-4b3e-9233-341d836b0107",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:13.261354Z",
     "iopub.status.busy": "2022-06-04T19:13:13.261240Z",
     "iopub.status.idle": "2022-06-04T19:13:13.847945Z",
     "shell.execute_reply": "2022-06-04T19:13:13.847452Z",
     "shell.execute_reply.started": "2022-06-04T19:13:13.261338Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from eugene.preprocessing import encodeDNA, decodeDNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "54caaf22-0d3f-4a8d-b8c3-ba75415988b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:13.848795Z",
     "iopub.status.busy": "2022-06-04T19:13:13.848682Z",
     "iopub.status.idle": "2022-06-04T19:13:14.509540Z",
     "shell.execute_reply": "2022-06-04T19:13:14.509225Z",
     "shell.execute_reply.started": "2022-06-04T19:13:13.848780Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quick test\n",
    "decodeDNA(encodeDNA(X_seqs[0:5])) == X_seqs[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "58f35871-41b1-423d-b118-a03b83bcf7a2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:14.510184Z",
     "iopub.status.busy": "2022-06-04T19:13:14.510076Z",
     "iopub.status.idle": "2022-06-04T19:13:30.585505Z",
     "shell.execute_reply": "2022-06-04T19:13:30.585041Z",
     "shell.execute_reply.started": "2022-06-04T19:13:14.510169Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATCT [[0. 1. 0. 0.]\n",
      " [1. 0. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 0. 1.]]\n",
      "total 15369456\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.09-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.09-0.4_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 26  2021 0.09-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 26  2021 0.09-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 26  2021 0.09-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 26  2021 0.09-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 26  2021 0.09-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 26  2021 0.09-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 417221504 Nov 26  2021 0.09-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 417221504 Nov 26  2021 0.09-0.4_X-training_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.18-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.18-0.4_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26  2021 0.18-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26  2021 0.18-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26  2021 0.18-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26  2021 0.18-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26  2021 0.18-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26  2021 0.18-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26  2021 0.18-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26  2021 0.18-0.4_X-training_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Jun  4 12:13 inf--inf_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Jun  4 11:59 inf--inf_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users    213440 Jun  4 12:13 inf--inf_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users    213440 Jun  4 11:59 inf--inf_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  97299968 Jun  4 12:13 inf--inf_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  97299968 Jun  4 11:59 inf--inf_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 875696576 Jun  4 12:13 inf--inf_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 875696576 Jun  4 11:59 inf--inf_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 972996416 Jun  4 12:13 inf--inf_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 972996416 Jun  4 11:59 inf--inf_X-training_ohe-seq-rev.npy\n"
     ]
    }
   ],
   "source": [
    "X_ohe_seq = encodeDNA(X_seqs)\n",
    "\n",
    "# Save the seqs\n",
    "if not os.path.isdir(\"ohe_seq\"):\n",
    "    os.makedirs(\"ohe_seq\")\n",
    "    \n",
    "np.save(\"ohe_seq/{}_X-all_ohe-seq\".format(PREPROCESS), X_ohe_seq)\n",
    "np.save(\"ohe_seq/{}_X-training_ohe-seq\".format(PREPROCESS), X_ohe_seq[training_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-train-{}_ohe-seq\".format(PREPROCESS, SPLIT), arr=X_ohe_seq[train_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-test-{}_ohe-seq\".format(PREPROCESS, round(1-SPLIT, 1)), arr=X_ohe_seq[test_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-holdout_ohe-seq\".format(PREPROCESS), X_ohe_seq[holdout_index, :, :])\n",
    "\n",
    "# Quick check\n",
    "print(X_seqs[0][:5], X_ohe_seq[0][:5])\n",
    "\n",
    "!ls -l ohe_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4f3478a-e3ec-4130-8f57-344ca01fc375",
   "metadata": {},
   "source": [
    "### *Reverse Seqs*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8814b146-dc23-41e5-b7ec-de3f25bf46f2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:30.586293Z",
     "iopub.status.busy": "2022-06-04T19:13:30.586162Z",
     "iopub.status.idle": "2022-06-04T19:13:31.226941Z",
     "shell.execute_reply": "2022-06-04T19:13:31.226563Z",
     "shell.execute_reply.started": "2022-06-04T19:13:30.586275Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from eugene.utils.seq_utils import reverse_complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a81b37cc-3a69-4e4b-b5fa-86939526e9f7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:31.228038Z",
     "iopub.status.busy": "2022-06-04T19:13:31.227922Z",
     "iopub.status.idle": "2022-06-04T19:13:50.045931Z",
     "shell.execute_reply": "2022-06-04T19:13:50.045361Z",
     "shell.execute_reply.started": "2022-06-04T19:13:31.228022Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCCTA [[0. 0. 0. 1.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [1. 0. 0. 0.]]\n",
      "total 15369456\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.09-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.09-0.4_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 26  2021 0.09-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 26  2021 0.09-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 26  2021 0.09-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 26  2021 0.09-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 26  2021 0.09-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 26  2021 0.09-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 417221504 Nov 26  2021 0.09-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 417221504 Nov 26  2021 0.09-0.4_X-training_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.18-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26  2021 0.18-0.4_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26  2021 0.18-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26  2021 0.18-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26  2021 0.18-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26  2021 0.18-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26  2021 0.18-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26  2021 0.18-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26  2021 0.18-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26  2021 0.18-0.4_X-training_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Jun  4 12:13 inf--inf_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Jun  4 12:13 inf--inf_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users    213440 Jun  4 12:13 inf--inf_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users    213440 Jun  4 12:13 inf--inf_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  97299968 Jun  4 12:13 inf--inf_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  97299968 Jun  4 12:13 inf--inf_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 875696576 Jun  4 12:13 inf--inf_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 875696576 Jun  4 12:13 inf--inf_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 972996416 Jun  4 12:13 inf--inf_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 972996416 Jun  4 12:13 inf--inf_X-training_ohe-seq-rev.npy\n"
     ]
    }
   ],
   "source": [
    "# Get the reverse encodings\n",
    "X_rev_seqs = np.array([reverse_complement(seq) for seq in X_seqs], dtype=\"object\")\n",
    "X_ohe_rev_seq = encodeDNA(X_rev_seqs)\n",
    "\n",
    "# Save the seqs\n",
    "if not os.path.isdir(\"ohe_seq\"):\n",
    "    os.makedirs(\"ohe_seq\")\n",
    "    \n",
    "np.save(\"ohe_seq/{}_X-all_ohe-seq-rev\".format(PREPROCESS), X_ohe_rev_seq)\n",
    "np.save(\"ohe_seq/{}_X-training_ohe-seq-rev\".format(PREPROCESS), arr=X_ohe_rev_seq[training_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-train-{}_ohe-seq-rev\".format(PREPROCESS, SPLIT), arr=X_ohe_rev_seq[train_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-test-{}_ohe-seq-rev\".format(PREPROCESS, round(1-SPLIT, 1)), arr=X_ohe_rev_seq[test_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-holdout_ohe-seq-rev\".format(PREPROCESS), X_ohe_rev_seq[holdout_index, :, :])\n",
    "\n",
    "# Quick check\n",
    "print(X_rev_seqs[0][:5], X_ohe_rev_seq[0][:5])\n",
    "\n",
    "!ls -l ohe_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "583e18b7-112e-4fc3-9f36-442747fba411",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:50.046901Z",
     "iopub.status.busy": "2022-06-04T19:13:50.046777Z",
     "iopub.status.idle": "2022-06-04T19:13:50.920387Z",
     "shell.execute_reply": "2022-06-04T19:13:50.919856Z",
     "shell.execute_reply.started": "2022-06-04T19:13:50.046882Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATCT GTAGA\n",
      "TGCTC ACGAG\n"
     ]
    }
   ],
   "source": [
    "print(X_seqs[0][:5], decodeDNA(X_ohe_rev_seq[0:100])[-1][-5:][::-1])\n",
    "print(X_seqs[10000][:5], decodeDNA(X_ohe_rev_seq[0:10000])[-1][-5:][::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59350c92-1067-47f6-a7b5-c51429274f62",
   "metadata": {
    "tags": []
   },
   "source": [
    "## **<u>Sequence feature idea 4 </u>**: Use the actual sequence (as fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "1cd248e4-9e7a-42f8-8328-41b29bec2587",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:50.921183Z",
     "iopub.status.busy": "2022-06-04T19:13:50.921065Z",
     "iopub.status.idle": "2022-06-04T19:13:51.562612Z",
     "shell.execute_reply": "2022-06-04T19:13:51.562120Z",
     "shell.execute_reply.started": "2022-06-04T19:13:50.921166Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from eugene.utils.seq_utils import seq2Fasta, gkmSeq2Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8d04e2b8-108e-48f2-80d1-f4e73f6ec5be",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:51.563375Z",
     "iopub.status.busy": "2022-06-04T19:13:51.563262Z",
     "iopub.status.idle": "2022-06-04T19:13:58.331127Z",
     "shell.execute_reply": "2022-06-04T19:13:58.330743Z",
     "shell.execute_reply.started": "2022-06-04T19:13:51.563360Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "921600\n",
      "921398\n",
      "829258\n",
      "92140\n",
      "202\n",
      "2764598\n",
      "   921600 fasta/0.09-0.4_X-all_fasta.fa\n",
      "   526504 fasta/0.09-0.4_X-holdout_fasta.fa\n",
      "    39510 fasta/0.09-0.4_X-test-0.1_fasta.fa\n",
      "   185928 fasta/0.09-0.4_X-train-0.9_fasta-neg.fa\n",
      "   169658 fasta/0.09-0.4_X-train-0.9_fasta-pos.fa\n",
      "   206718 fasta/0.09-0.4_X-training_fasta-neg.fa\n",
      "   188378 fasta/0.09-0.4_X-training_fasta-pos.fa\n",
      "  2238296 total\n",
      "   921600 fasta/0.18-0.4_X-all_fasta.fa\n",
      "   315842 fasta/0.18-0.4_X-holdout_fasta.fa\n",
      "    60576 fasta/0.18-0.4_X-test-0.1_fasta.fa\n",
      "   375660 fasta/0.18-0.4_X-train-0.9_fasta-neg.fa\n",
      "   169522 fasta/0.18-0.4_X-train-0.9_fasta-pos.fa\n",
      "   417380 fasta/0.18-0.4_X-training_fasta-neg.fa\n",
      "   188378 fasta/0.18-0.4_X-training_fasta-pos.fa\n",
      "  2448958 total\n",
      "   921600 fasta/inf--inf_X-all_fasta.fa\n",
      "      202 fasta/inf--inf_X-holdout_fasta.fa\n",
      "    92140 fasta/inf--inf_X-test-0.1_fasta.fa\n",
      "        0 fasta/inf--inf_X-train-0.9_fasta-neg.fa\n",
      "   829258 fasta/inf--inf_X-train-0.9_fasta-pos.fa\n",
      "        0 fasta/inf--inf_X-training_fasta-neg.fa\n",
      "   921398 fasta/inf--inf_X-training_fasta-pos.fa\n",
      "  2764598 total\n"
     ]
    }
   ],
   "source": [
    "# All seqs\n",
    "file_name = \"fasta/{}_X-all_fasta\".format(PREPROCESS)\n",
    "seq2Fasta(X_seqs, ID, name=file_name)\n",
    "\n",
    "# Training seqs\n",
    "file_name = \"fasta/{}_X-training_fasta\".format(PREPROCESS)\n",
    "gkmSeq2Fasta(X_seqs[training_index], ID[training_index], binary_training, name=file_name)\n",
    "\n",
    "# Train seqs\n",
    "file_name = \"fasta/{}_X-train-{}_fasta\".format(PREPROCESS, SPLIT)\n",
    "gkmSeq2Fasta(X_seqs[train_index], ID[train_index], binary_train, name=file_name)\n",
    "\n",
    "# Test seqs\n",
    "file_name = \"fasta/{}_X-test-{}_fasta\".format(PREPROCESS, round(1-SPLIT, 1))\n",
    "seq2Fasta(X_seqs[test_index], ID[test_index], name=file_name)\n",
    "\n",
    "# Holdout seqs\n",
    "file_name = \"fasta/{}_X-holdout_fasta\".format(PREPROCESS)\n",
    "seq2Fasta(X_seqs[holdout_index], ID[holdout_index], name=file_name)\n",
    "\n",
    "# Sanity checks\n",
    "print(len(ID)*2)\n",
    "print(len(training_index)*2)\n",
    "print(len(train_index)*2)\n",
    "print(len(test_index)*2)\n",
    "print(len(holdout_index)*2)\n",
    "print(len(ID)*2 + len(training_index)*2 + len(train_index)*2 + len(test_index)*2 + len(holdout_index)*2)\n",
    "\n",
    "!wc -l fasta/0.09-0.4*\n",
    "!wc -l fasta/0.18-0.4*\n",
    "!wc -l fasta/inf--inf*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df3c533f-820a-4b64-b297-b8c778e21c35",
   "metadata": {},
   "source": [
    "# Final Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "90a1320f-269c-401a-8e3c-209bb27fc0fd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-04T19:13:58.331922Z",
     "iopub.status.busy": "2022-06-04T19:13:58.331795Z",
     "iopub.status.idle": "2022-06-04T19:13:59.459501Z",
     "shell.execute_reply": "2022-06-04T19:13:59.459066Z",
     "shell.execute_reply.started": "2022-06-04T19:13:58.331901Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[38;5;33m.\u001b[0m\n",
      "├── 20210728-3.EnhancerTable.ForAdam.FunctionalEnhancers.WT-detected.ABL-notDetected.10R-20U-0.1P.tsv\n",
      "├── 2021_OLS_Library_EDA.ipynb\n",
      "├── 2021_OLS_Library_preprocess.ipynb\n",
      "├── 2021_OLS_Library.tsv\n",
      "├── 3.ForAdam.EnhancerTable.FunctionalEnhancers.WT-detected.ABL-notDetected.10R-20U-0.1P.tsv\n",
      "├── \u001b[38;5;33mactivity\u001b[0m\n",
      "│   ├── 0.09-0.4_y-all_activity.txt\n",
      "│   ├── 0.09-0.4_y-holdout_activity.txt\n",
      "│   ├── 0.09-0.4_y-test-0.1_activity.txt\n",
      "│   ├── 0.09-0.4_y-train-0.9_activity.txt\n",
      "│   ├── 0.09-0.4_y-training_activity.txt\n",
      "│   ├── inf--inf_y-all_activity.txt\n",
      "│   ├── inf--inf_y-holdout_activity.txt\n",
      "│   ├── inf--inf_y-test-0.1_activity.txt\n",
      "│   ├── inf--inf_y-train-0.9_activity.txt\n",
      "│   └── inf--inf_y-training_activity.txt\n",
      "├── \u001b[38;5;33mbinary\u001b[0m\n",
      "│   ├── 0.09-0.4_y-all_binary.txt\n",
      "│   ├── 0.09-0.4_y-holdout_binary.txt\n",
      "│   ├── 0.09-0.4_y-test-0.1_binary.txt\n",
      "│   ├── 0.09-0.4_y-train-0.9_binary.txt\n",
      "│   ├── 0.09-0.4_y-training_binary.txt\n",
      "│   ├── 0.18-0.4_y-all_binary.txt\n",
      "│   ├── 0.18-0.4_y-test-0.1_binary.txt\n",
      "│   ├── 0.18-0.4_y-train-0.9_binary.txt\n",
      "│   ├── 0.18-0.4_y-training_binary.txt\n",
      "│   ├── inf--inf_y-all_binary.txt\n",
      "│   ├── inf--inf_y-holdout_binary.txt\n",
      "│   ├── inf--inf_y-test-0.1_binary.txt\n",
      "│   ├── inf--inf_y-train-0.9_binary.txt\n",
      "│   └── inf--inf_y-training_binary.txt\n",
      "├── \u001b[38;5;33mblock\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_block.npy\n",
      "│   ├── 0.09-0.4_X-holdout_block.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_block.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_block.npy\n",
      "│   ├── 0.09-0.4_X-training_block.npy\n",
      "│   ├── 0.18-0.4_X-all_block.npy\n",
      "│   ├── 0.18-0.4_X-holdout_block.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_block.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_block.npy\n",
      "│   ├── 0.18-0.4_X-training_block.npy\n",
      "│   ├── block_header.txt\n",
      "│   ├── inf--inf_X-all_block.npy\n",
      "│   ├── inf--inf_X-holdout_block.npy\n",
      "│   ├── inf--inf_X-test-0.1_block.npy\n",
      "│   ├── inf--inf_X-train-0.9_block.npy\n",
      "│   └── inf--inf_X-training_block.npy\n",
      "├── \u001b[38;5;33mfasta\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_fasta.fa\n",
      "│   ├── 0.09-0.4_X-holdout_fasta.fa\n",
      "│   ├── 0.09-0.4_X-test-0.1_fasta.fa\n",
      "│   ├── 0.09-0.4_X-train-0.9_fasta-neg.fa\n",
      "│   ├── 0.09-0.4_X-train-0.9_fasta-pos.fa\n",
      "│   ├── 0.09-0.4_X-training_fasta-neg.fa\n",
      "│   ├── 0.09-0.4_X-training_fasta-pos.fa\n",
      "│   ├── 0.18-0.4_X-all_fasta.fa\n",
      "│   ├── 0.18-0.4_X-holdout_fasta.fa\n",
      "│   ├── 0.18-0.4_X-test-0.1_fasta.fa\n",
      "│   ├── 0.18-0.4_X-train-0.9_fasta-neg.fa\n",
      "│   ├── 0.18-0.4_X-train-0.9_fasta-pos.fa\n",
      "│   ├── 0.18-0.4_X-training_fasta-neg.fa\n",
      "│   ├── 0.18-0.4_X-training_fasta-pos.fa\n",
      "│   ├── inf--inf_X-all_fasta.fa\n",
      "│   ├── inf--inf_X-holdout_fasta.fa\n",
      "│   ├── inf--inf_X-test-0.1_fasta.fa\n",
      "│   ├── inf--inf_X-train-0.9_fasta-neg.fa\n",
      "│   ├── inf--inf_X-train-0.9_fasta-pos.fa\n",
      "│   ├── inf--inf_X-training_fasta-neg.fa\n",
      "│   ├── inf--inf_X-training_fasta-pos.fa\n",
      "│   └── \u001b[38;5;33msubset\u001b[0m\n",
      "├── \u001b[38;5;33mfigs\u001b[0m\n",
      "│   ├── \u001b[38;5;13m0.09-0.4_class-balance.png\u001b[0m\n",
      "│   ├── \u001b[38;5;13m0.18-0.4_class-balance.png\u001b[0m\n",
      "│   ├── \u001b[38;5;13m0.18-0.4_feature-counts.png\u001b[0m\n",
      "│   ├── \u001b[38;5;33modds_ratios\u001b[0m\n",
      "│   └── \u001b[38;5;33mset_disribution\u001b[0m\n",
      "├── \u001b[38;5;33mid\u001b[0m\n",
      "│   ├── 0.09-0.4_id-all.txt\n",
      "│   ├── 0.09-0.4_id-holdout.txt\n",
      "│   ├── 0.09-0.4_id-test-0.1.txt\n",
      "│   ├── 0.09-0.4_id-train-0.9.txt\n",
      "│   ├── 0.09-0.4_id-training.txt\n",
      "│   ├── 0.18-0.4_id-all.txt\n",
      "│   ├── 0.18-0.4_id-holdout.txt\n",
      "│   ├── 0.18-0.4_id-test-0.1.txt\n",
      "│   ├── 0.18-0.4_id-train-0.9.txt\n",
      "│   ├── 0.18-0.4_id-training.txt\n",
      "│   ├── inf--inf_id-all.txt\n",
      "│   ├── inf--inf_id-holdout.txt\n",
      "│   ├── inf--inf_id-test-0.1.txt\n",
      "│   ├── inf--inf_id-train-0.9.txt\n",
      "│   └── inf--inf_id-training.txt\n",
      "├── \u001b[38;5;33mmixed_1.0\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_mixed-1.0.npy\n",
      "│   ├── 0.09-0.4_X-holdout_mixed-1.0.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_mixed-1.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_mixed-1.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.09-0.4_X-training_mixed-1.0.npy\n",
      "│   ├── 0.09-0.4_X-training_stats.pickle\n",
      "│   ├── 0.18-0.4_mixed-1.0_eda.html\n",
      "│   ├── 0.18-0.4_X-all_mixed-1.0.npy\n",
      "│   ├── 0.18-0.4_X-holdout_mixed-1.0.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_mixed-1.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_mixed-1.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.18-0.4_X-training_mixed-1.0.npy\n",
      "│   ├── 0.18-0.4_X-training_stats.pickle\n",
      "│   ├── inf--inf_X-all_mixed-1.0.npy\n",
      "│   ├── inf--inf_X-holdout_mixed-1.0.npy\n",
      "│   ├── inf--inf_X-test-0.1_mixed-1.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_mixed-1.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_stats.pickle\n",
      "│   ├── inf--inf_X-training_mixed-1.0.npy\n",
      "│   ├── inf--inf_X-training_stats.pickle\n",
      "│   └── mixed-1.0_header.txt\n",
      "├── \u001b[38;5;33mmixed_2.0\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_mixed-2.0.npy\n",
      "│   ├── 0.09-0.4_X-holdout_mixed-2.0.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_mixed-2.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_mixed-2.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.09-0.4_X-training_mixed-2.0.npy\n",
      "│   ├── 0.09-0.4_X-training_stats.pickle\n",
      "│   ├── 0.18-0.4_X-all_mixed-2.0.npy\n",
      "│   ├── 0.18-0.4_X-holdout_mixed-2.0.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_mixed-2.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_mixed-2.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.18-0.4_X-training_mixed-2.0.npy\n",
      "│   ├── 0.18-0.4_X-training_stats.pickle\n",
      "│   ├── inf--inf_X-all_mixed-2.0.npy\n",
      "│   ├── inf--inf_X-holdout_mixed-2.0.npy\n",
      "│   ├── inf--inf_X-test-0.1_mixed-2.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_mixed-2.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_stats.pickle\n",
      "│   ├── inf--inf_X-training_mixed-2.0.npy\n",
      "│   ├── inf--inf_X-training_stats.pickle\n",
      "│   └── mixed-2.0_header.txt\n",
      "├── \u001b[38;5;33mmixed_3.0\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_mixed-3.0.npy\n",
      "│   ├── 0.09-0.4_X-holdout_mixed-3.0.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_mixed-3.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_mixed-3.0.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.09-0.4_X-training_mixed-3.0.npy\n",
      "│   ├── 0.09-0.4_X-training_stats.pickle\n",
      "│   ├── 0.18-0.4_X-all_mixed-3.0.npy\n",
      "│   ├── 0.18-0.4_X-holdout_mixed-3.0.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_mixed-3.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_mixed-3.0.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_stats.pickle\n",
      "│   ├── 0.18-0.4_X-training_mixed-3.0.npy\n",
      "│   ├── 0.18-0.4_X-training_stats.pickle\n",
      "│   ├── inf--inf_X-all_mixed-3.0.npy\n",
      "│   ├── inf--inf_X-holdout_mixed-3.0.npy\n",
      "│   ├── inf--inf_X-test-0.1_mixed-3.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_mixed-3.0.npy\n",
      "│   ├── inf--inf_X-train-0.9_stats.pickle\n",
      "│   ├── inf--inf_X-training_mixed-3.0.npy\n",
      "│   ├── inf--inf_X-training_stats.pickle\n",
      "│   └── mixed-3.0_header.txt\n",
      "├── \u001b[38;5;33mohe_seq\u001b[0m\n",
      "│   ├── 0.09-0.4_X-all_ohe-seq.npy\n",
      "│   ├── 0.09-0.4_X-all_ohe-seq-rev.npy\n",
      "│   ├── 0.09-0.4_X-holdout_ohe-seq.npy\n",
      "│   ├── 0.09-0.4_X-holdout_ohe-seq-rev.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_ohe-seq.npy\n",
      "│   ├── 0.09-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_ohe-seq.npy\n",
      "│   ├── 0.09-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "│   ├── 0.09-0.4_X-training_ohe-seq.npy\n",
      "│   ├── 0.09-0.4_X-training_ohe-seq-rev.npy\n",
      "│   ├── 0.18-0.4_X-all_ohe-seq.npy\n",
      "│   ├── 0.18-0.4_X-all_ohe-seq-rev.npy\n",
      "│   ├── 0.18-0.4_X-holdout_ohe-seq.npy\n",
      "│   ├── 0.18-0.4_X-holdout_ohe-seq-rev.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_ohe-seq.npy\n",
      "│   ├── 0.18-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_ohe-seq.npy\n",
      "│   ├── 0.18-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "│   ├── 0.18-0.4_X-training_ohe-seq.npy\n",
      "│   ├── 0.18-0.4_X-training_ohe-seq-rev.npy\n",
      "│   ├── inf--inf_X-all_ohe-seq.npy\n",
      "│   ├── inf--inf_X-all_ohe-seq-rev.npy\n",
      "│   ├── inf--inf_X-holdout_ohe-seq.npy\n",
      "│   ├── inf--inf_X-holdout_ohe-seq-rev.npy\n",
      "│   ├── inf--inf_X-test-0.1_ohe-seq.npy\n",
      "│   ├── inf--inf_X-test-0.1_ohe-seq-rev.npy\n",
      "│   ├── inf--inf_X-train-0.9_ohe-seq.npy\n",
      "│   ├── inf--inf_X-train-0.9_ohe-seq-rev.npy\n",
      "│   ├── inf--inf_X-training_ohe-seq.npy\n",
      "│   └── inf--inf_X-training_ohe-seq-rev.npy\n",
      "├── \u001b[38;5;33mphrase_encodings\u001b[0m\n",
      "│   ├── parse_ord_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_aff_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_aff_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_aff_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_aff_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_aff_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_aff_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_aff_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_aff_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_spc_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_spc_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_spc_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_spc_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_spc_aff_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_spc_aff_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_ori_spc_aff_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_ori_spc_aff_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_spc_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_spc_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_spc_3_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_spc_3_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_spc_aff_2_weightedByCount_False.tsv\n",
      "│   ├── parse_ord_spc_aff_2_weightedByCount_True.tsv\n",
      "│   ├── parse_ord_spc_aff_3_weightedByCount_False.tsv\n",
      "│   └── parse_ord_spc_aff_3_weightedByCount_True.tsv\n",
      "└── \u001b[38;5;33mseqs\u001b[0m\n",
      "    ├── 0.09-0.4_seqs-all.txt\n",
      "    ├── 0.09-0.4_seqs-holdout.txt\n",
      "    ├── 0.09-0.4_seqs-test-0.1.txt\n",
      "    ├── 0.09-0.4_seqs-train-0.9.txt\n",
      "    ├── 0.09-0.4_seqs-training.txt\n",
      "    ├── 0.18-0.4_seqs-all.txt\n",
      "    ├── 0.18-0.4_seqs-holdout.txt\n",
      "    ├── 0.18-0.4_seqs-test-0.1.txt\n",
      "    ├── 0.18-0.4_seqs-train-0.9.txt\n",
      "    ├── 0.18-0.4_seqs-training.txt\n",
      "    ├── inf--inf_seqs-all.txt\n",
      "    ├── inf--inf_seqs-holdout.txt\n",
      "    ├── inf--inf_seqs-test-0.1.txt\n",
      "    ├── inf--inf_seqs-train-0.9.txt\n",
      "    └── inf--inf_seqs-training.txt\n",
      "\n",
      "15 directories, 228 files\n"
     ]
    }
   ],
   "source": [
    "!tree -L 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fddfcab-8659-471f-ac58-cd76fdde58cd",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "a00ef7f5-22af-4d32-8b3d-36aede4eec3b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-27T06:43:35.174371Z",
     "iopub.status.busy": "2021-11-27T06:43:35.174136Z",
     "iopub.status.idle": "2021-11-27T06:49:29.803287Z",
     "shell.execute_reply": "2021-11-27T06:49:29.802205Z",
     "shell.execute_reply.started": "2021-11-27T06:43:35.174342Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 460800/460800 [05:46<00:00, 1327.99it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Encoded 460800 seqs\n",
      "Checking 1000 random seqs for proper encoding\n",
      "Sequence encoding was great success\n",
      "CATCT [[0. 1. 0. 0.]\n",
      " [1. 0. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 0. 1.]]\n",
      "total 8309627\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26 22:49 0.09-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 26 22:49 0.09-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 555988352 Nov 24 17:45 0.09-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 26 22:49 0.09-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  41722688 Nov 24 17:45 0.09-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 26 22:49 0.09-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 375498944 Nov 24 17:45 0.09-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 417221504 Nov 26 22:49 0.09-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26 14:24 0.18-0.4_X-all_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 973209728 Nov 26 16:15 0.18-0.4_X-all_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26 14:24 0.18-0.4_X-holdout_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 333529280 Nov 26 16:15 0.18-0.4_X-holdout_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26 14:24 0.18-0.4_X-test-0.1_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users  63968384 Nov 26 16:15 0.18-0.4_X-test-0.1_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26 14:24 0.18-0.4_X-train-0.9_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 575712320 Nov 26 16:15 0.18-0.4_X-train-0.9_ohe-seq-rev.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26 14:24 0.18-0.4_X-training_ohe-seq.npy\n",
      "-rw-r--r-- 1 aklie carter-users 639680576 Nov 26 16:15 0.18-0.4_X-training_ohe-seq-rev.npy\n"
     ]
    }
   ],
   "source": [
    "# Get the ohe\n",
    "X_ohe_seq = ohe_seqs(X_seqs)\n",
    "\n",
    "# Save the seqs\n",
    "if not os.path.isdir(\"ohe_seq\"):\n",
    "    os.makedirs(\"ohe_seq\")\n",
    "    \n",
    "np.save(\"ohe_seq/{}_X-all_ohe-seq\".format(PREPROCESS), X_ohe_seq)\n",
    "np.save(\"ohe_seq/{}_X-training_ohe-seq\".format(PREPROCESS), X_ohe_seq[training_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-train-{}_ohe-seq\".format(PREPROCESS, SPLIT), arr=X_ohe_seq[train_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-test-{}_ohe-seq\".format(PREPROCESS, round(1-SPLIT, 1)), arr=X_ohe_seq[test_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-holdout_ohe-seq\".format(PREPROCESS), X_ohe_seq[holdout_index, :, :])\n",
    "\n",
    "# Quick check\n",
    "print(X_seqs[0][:5], X_ohe_seq[0][:5])\n",
    "\n",
    "!ls -l ohe_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a00f39-b476-467e-b16a-5fb44c9b3e0e",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.707000Z",
     "iopub.status.idle": "2021-11-27T00:18:14.707359Z",
     "shell.execute_reply": "2021-11-27T00:18:14.707187Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create the encoding\n",
    "mixed3_encoding = []\n",
    "\n",
    "# Loop through each enhancer\n",
    "for i, (row_num, enh_data) in enumerate(tqdm.tqdm(OLS_data[block_features].iterrows())):\n",
    "\n",
    "    enh_enc = []  # Single enhancer encoding\n",
    "\n",
    "    # Loop through each position\n",
    "    for col_num in range(len(enh_data.index)):\n",
    "\n",
    "        # If we have a spacer in the current position we need to check for surrounding GATA-2 sites\n",
    "        if \"S\" in enh_data.iloc[col_num]:\n",
    "\n",
    "            # If the spacer is the empty spacer, just add a 0 and go to the next\n",
    "            if enh_data.iloc[col_num] == \"S5\":\n",
    "                enh_enc.append(\n",
    "                    len(siteName2bindingSiteSequence[enh_data.iloc[col_num]])\n",
    "                )\n",
    "                continue\n",
    "\n",
    "            # If the spacer is downstream of a GATA-2 reverse, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num > 0:\n",
    "                if enh_data.iloc[col_num - 1] == \"G2R\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # If the spacer is upstream of a GATA-2 forward, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num < len(enh_data.index) - 1:\n",
    "                if enh_data.iloc[col_num + 1] == \"G2F\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # Finally if no G2 or S5 is involved, just add the normal len of the spacer\n",
    "            enh_enc.append(len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]))\n",
    "            continue\n",
    "\n",
    "        # If we are at a TFBS, add the TFBS type, orientation and affinity\n",
    "        else:\n",
    "            tfbs = enh_data.iloc[col_num]\n",
    "            tf = tfbs[0]\n",
    "            aff = bindingSiteName2affinities[tfbs[:2]]\n",
    "            orient = tfbs[2]\n",
    "            if tf == \"E\":\n",
    "                enh_enc += [aff, 0, orient]\n",
    "            elif tf == \"G\":\n",
    "                enh_enc += [0, aff, orient]\n",
    "    # Poisin pill\n",
    "    # if row_num == 10000:\n",
    "    #    break\n",
    "    mixed3_encoding.append(enh_enc + [\"-\".join(list(enh_data.values))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a498d2-47a2-48fd-b9aa-a3fcfead1f09",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.708439Z",
     "iopub.status.idle": "2021-11-27T00:18:14.708782Z",
     "shell.execute_reply": "2021-11-27T00:18:14.708613Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mixed-3.0 header\n",
    "header = [\n",
    "    \"L1_length\",\n",
    "    \"TFBS1_ETS_affinity\",\n",
    "    \"TFBS1_GATA_affinity\",\n",
    "    \"TFBS1_orient\",\n",
    "    \"L2_length\",\n",
    "    \"TFBS2_ETS_affinity\",\n",
    "    \"TFBS2_GATA_affinity\",\n",
    "    \"TFBS2_orient\",\n",
    "    \"L3_length\",\n",
    "    \"TFBS3_ETS_affinity\",\n",
    "    \"TFBS3_GATA_affinity\",\n",
    "    \"TFBS3_orient\",\n",
    "    \"L4_length\",\n",
    "    \"TFBS4_ETS_affinity\",\n",
    "    \"TFBS4_GATA_affinity\",\n",
    "    \"TFBS4_orient\",\n",
    "    \"L5_length\",\n",
    "    \"TFBS5_ETS_affinity\",\n",
    "    \"TFBS5_GATA_affinity\",\n",
    "    \"TFBS5_orient\",\n",
    "    \"L6_length\",\n",
    "    \"Enhancer\"\n",
    "]\n",
    "# Save the header for use down the line. Save as one column per line\n",
    "with open (\"mixed_3.0/mixed-3.0_header.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(header[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fd6f37e-f51f-4d1d-95b0-bcb8bc8c566c",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.710024Z",
     "iopub.status.idle": "2021-11-27T00:18:14.710383Z",
     "shell.execute_reply": "2021-11-27T00:18:14.710212Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create a dataframe and replace occurrences of TFs and orientation with numbers\n",
    "X_mixed3 = pd.DataFrame(mixed3_encoding, columns=header).replace({\"R\": -1, \"F\": 1})\n",
    "\n",
    "# Drop the enhancer column\n",
    "X_mixed3 = X_mixed3.drop(\"Enhancer\", axis=1).values\n",
    "\n",
    "# Subset train and test\n",
    "X_mixed3_train = X_mixed3[train_index, :]\n",
    "X_mixed3_test = X_mixed3[test_index, :]\n",
    "X_mixed3_trian_and_test = X_mixed3[training_index, :]\n",
    "X_mixed3_holdout = X_mixed3[holdout_index, :]\n",
    "\n",
    "# Standardize features and double check means are 0 and stds are 1\n",
    "scale_indeces = np.array([0, 4, 8, 12, 16, 20])  # Mixed 3.0\n",
    "_, X_mixed3_scaled = project_utils.standardize_features(train_X=X_mixed3_trian_and_test, test_X=X_mixed3_trian_and_test, indeces=scale_indeces, stats_file=\"mixed_3.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed3_train_scaled, X_mixed3_test_scaled = project_utils.standardize_features(train_X=X_mixed3_train, test_X=X_mixed3_test, indeces=scale_indeces, stats_file=\"mixed_3.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed3_train_scaled, X_mixed3_holdout_scaled = project_utils.standardize_features(train_X=X_mixed3_train, test_X=X_mixed3_holdout, indeces=scale_indeces)\n",
    "\n",
    "# Check\n",
    "X_mixed3_train[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_train[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed3_train_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_train_scaled[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed3_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed3_scaled[:, scale_indeces].std(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93a80740-f7e3-469f-b58d-a2a29e0d6602",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.711454Z",
     "iopub.status.idle": "2021-11-27T00:18:14.711799Z",
     "shell.execute_reply": "2021-11-27T00:18:14.711629Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Save the different files as plain text for now\n",
    "if not os.path.isdir(\"mixed_3.0\"):\n",
    "    os.makedirs(\"mixed_3.0\")  \n",
    "    \n",
    "np.save(\"mixed_3.0/{}_X-all_mixed-3.0\".format(PREPROCESS), X_mixed3)\n",
    "np.save(\"mixed_3.0/{}_X-training_mixed-3.0\".format(PREPROCESS), X_mixed3_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-train-{}_mixed-3.0\".format(PREPROCESS, SPLIT), X_mixed3_train_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-test-{}_mixed-3.0\".format(PREPROCESS, round(1-SPLIT, 1)), X_mixed3_test_scaled)\n",
    "np.save(\"mixed_3.0/{}_X-holdout_mixed-3.0\".format(PREPROCESS), X_mixed3_holdout_scaled)\n",
    "    \n",
    "!ls -l mixed_3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a482d13-1775-4168-a668-b6724c763528",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.700932Z",
     "iopub.status.idle": "2021-11-27T00:18:14.701317Z",
     "shell.execute_reply": "2021-11-27T00:18:14.701112Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create the encoding\n",
    "mixed2_encoding = []\n",
    "\n",
    "# Loop through each enhancer\n",
    "for i, (row_num, enh_data) in tqdm.tqdm(enumerate(OLS_data[block_features].iterrows())):\n",
    "\n",
    "    enh_enc = []  # Single enhancer encoding\n",
    "\n",
    "    # Loop through each position\n",
    "    for col_num in range(len(enh_data.index)):\n",
    "\n",
    "        # If we have a spacer in the current position we need to check for surrounding GATA-2 sites\n",
    "        if \"S\" in enh_data.iloc[col_num]:\n",
    "\n",
    "            # If the spacer is the empty spacer, just add a 0 and go to the next\n",
    "            if enh_data.iloc[col_num] == \"S5\":\n",
    "                enh_enc.append(\n",
    "                    len(siteName2bindingSiteSequence[enh_data.iloc[col_num]])\n",
    "                )\n",
    "                continue\n",
    "\n",
    "            # If the spacer is downstream of a GATA-2 reverse, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num > 0:\n",
    "                if enh_data.iloc[col_num - 1] == \"G2R\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # If the spacer is upstream of a GATA-2 forward, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num < len(enh_data.index) - 1:\n",
    "                if enh_data.iloc[col_num + 1] == \"G2F\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # Finally if no G2 or S5 is involved, just add the normal len of the spacer\n",
    "            enh_enc.append(len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]))\n",
    "            continue\n",
    "\n",
    "        # If we are at a TFBS, add the TFBS type, orientation and affinity\n",
    "        else:\n",
    "            tfbs = enh_data.iloc[col_num]\n",
    "            tf = tfbs[0]\n",
    "            aff = bindingSiteName2affinities[tfbs[:2]]\n",
    "            orient = tfbs[2]\n",
    "            if tf == \"E\":\n",
    "                enh_enc += [aff, orient, 0, 0]\n",
    "            elif tf == \"G\":\n",
    "                enh_enc += [0, 0, aff, orient]\n",
    "    # Poisin pill\n",
    "    # if row_num == 10000:\n",
    "    #    break\n",
    "    mixed2_encoding.append(enh_enc + [\"-\".join(list(enh_data.values))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a5f2df-bdfd-4697-ba58-69575bda8299",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.702452Z",
     "iopub.status.idle": "2021-11-27T00:18:14.702797Z",
     "shell.execute_reply": "2021-11-27T00:18:14.702630Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mixed-2.0 header\n",
    "header = [\n",
    "    \"L1_length\",\n",
    "    \"TFBS1_ETS_affinity\",\n",
    "    \"TFBS1_ETS_orient\",\n",
    "    \"TFBS1_GATA_affinity\",\n",
    "    \"TFBS1_GATA_orient\",\n",
    "    \"L2_length\",\n",
    "    \"TFBS2_ETS_affinity\",\n",
    "    \"TFBS2_ETS_orient\",\n",
    "    \"TFBS2_GATA_affinity\",\n",
    "    \"TFBS2_GATA_orient\",\n",
    "    \"L3_length\",\n",
    "    \"TFBS3_ETS_affinity\",\n",
    "    \"TFBS3_ETS_orient\",\n",
    "    \"TFBS3_GATA_affinity\",\n",
    "    \"TFBS3_GATA_orient\",\n",
    "    \"L4_length\",\n",
    "    \"TFBS4_ETS_affinity\",\n",
    "    \"TFBS4_ETS_orient\",\n",
    "    \"TFBS4_GATA_affinity\",\n",
    "    \"TFBS4_GATA_orient\",\n",
    "    \"L5_length\",\n",
    "    \"TFBS5_ETS_affinity\",\n",
    "    \"TFBS5_ETS_orient\",\n",
    "    \"TFBS5_GATA_affinity\",\n",
    "    \"TFBS5_GATA_orient\",\n",
    "    \"L6_length\",\n",
    "    \"Enhancer\",\n",
    "]\n",
    "# Save the header for use down the line. Save as one column per line\n",
    "with open (\"mixed_2.0/mixed-2.0_header.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(header[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a79baa8-8d20-4468-ab19-b7895cada1a9",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.703920Z",
     "iopub.status.idle": "2021-11-27T00:18:14.704282Z",
     "shell.execute_reply": "2021-11-27T00:18:14.704094Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create a dataframe and replace occurrences of TFs and orientation with numbers\n",
    "X_mixed2 = (\n",
    "    pd.DataFrame(mixed2_encoding, columns=header)\n",
    "    .replace({\"G\": 0, \"E\": 1, \"R\": 0, \"F\": 1})\n",
    ")\n",
    "\n",
    "# Drop the enhancer column\n",
    "X_mixed2 = X_mixed2.drop(\"Enhancer\", axis=1).values\n",
    "\n",
    "# Subset train and test\n",
    "X_mixed2_train = X_mixed2[train_index, :]\n",
    "X_mixed2_test = X_mixed2[test_index, :]\n",
    "X_mixed2_trian_and_test = X_mixed2[training_index, :]\n",
    "X_mixed2_holdout = X_mixed2[holdout_index, :]\n",
    "\n",
    "# Standardize features and double check means are 0 and stds are 1\n",
    "scale_indeces = np.array([0, 5, 10, 15, 20, 25])  # Mixed 2.0\n",
    "_, X_mixed2_scaled = project_utils.standardize_features(train_X=X_mixed2_trian_and_test, test_X=X_mixed2_trian_and_test, indeces=scale_indeces, stats_file=\"mixed_2.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed2_train_scaled, X_mixed2_test_scaled = project_utils.standardize_features(train_X=X_mixed2_train, test_X=X_mixed2_test, indeces=scale_indeces, stats_file=\"mixed_2.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed2_train_scaled, X_mixed2_holdout_scaled = project_utils.standardize_features(train_X=X_mixed2_train, test_X=X_mixed2_holdout, indeces=scale_indeces)\n",
    "\n",
    "# Check\n",
    "X_mixed2_train[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_train[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed2_train_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_train_scaled[:, scale_indeces].std(axis=0), \\\n",
    "X_mixed2_scaled[:, scale_indeces].mean(axis=0), \\\n",
    "X_mixed2_scaled[:, scale_indeces].std(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f43fbb1f-8a41-4d05-bc35-2e2e86ebb22e",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.705425Z",
     "iopub.status.idle": "2021-11-27T00:18:14.705770Z",
     "shell.execute_reply": "2021-11-27T00:18:14.705600Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Save the different files as plain text for now\n",
    "if not os.path.isdir(\"mixed_2.0\"):\n",
    "    os.makedirs(\"mixed_2.0\")  \n",
    "    \n",
    "np.save(\"mixed_2.0/{}_X-all_mixed-2.0\".format(PREPROCESS), X_mixed2)\n",
    "np.save(\"mixed_2.0/{}_X-training_mixed-2.0\".format(PREPROCESS), X_mixed2_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-train-{}_mixed-2.0\".format(PREPROCESS, SPLIT), X_mixed2_train_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-test-{}_mixed-2.0\".format(PREPROCESS, round(1-SPLIT, 1)), X_mixed2_test_scaled)\n",
    "np.save(\"mixed_2.0/{}_X-holdout_mixed-2.0\".format(PREPROCESS), X_mixed1_holdout_scaled)\n",
    "    \n",
    "!ls -l mixed_2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6c83887-70b0-4bb1-b58c-9391ef463533",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.692753Z",
     "iopub.status.idle": "2021-11-27T00:18:14.693360Z",
     "shell.execute_reply": "2021-11-27T00:18:14.693038Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create the encoding\n",
    "mixed1_encoding = []\n",
    "\n",
    "# Loop through each enhancer\n",
    "for i, (row_num, enh_data) in enumerate(tqdm.tqdm(OLS_data[block_features].iterrows())):\n",
    "\n",
    "    enh_enc = []  # Single enhancer encoding\n",
    "\n",
    "    # Loop through each position\n",
    "    for col_num in range(len(enh_data.index)):\n",
    "\n",
    "        # If we have a spacer in the current position we need to check for surrounding GATA-2 sites\n",
    "        if \"S\" in enh_data.iloc[col_num]:\n",
    "\n",
    "            # If the spacer is the empty spacer, just add a 0 and go to the next\n",
    "            if enh_data.iloc[col_num] == \"S5\":\n",
    "                enh_enc.append(\n",
    "                    len(siteName2bindingSiteSequence[enh_data.iloc[col_num]])\n",
    "                )\n",
    "                continue\n",
    "\n",
    "            # If the spacer is downstream of a GATA-2 reverse, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num > 0:\n",
    "                if enh_data.iloc[col_num - 1] == \"G2R\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # If the spacer is upstream of a GATA-2 forward, we need to add a nucleotide to the GATA-2 (subtract one from spacer)\n",
    "            if col_num < len(enh_data.index) - 1:\n",
    "                if enh_data.iloc[col_num + 1] == \"G2F\":\n",
    "                    enh_enc.append(\n",
    "                        len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]) - 1\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "            # Finally if no G2F or S5 is involved, just add the normal len of the spacer\n",
    "            enh_enc.append(len(siteName2bindingSiteSequence[enh_data.iloc[col_num]]))\n",
    "            continue\n",
    "\n",
    "        # If we are at a TFBS, add the TFBS type, orientation and affinity\n",
    "        else:\n",
    "            tfbs = enh_data.iloc[col_num]\n",
    "            tf = tfbs[0]\n",
    "            aff = bindingSiteName2affinities[tfbs[:2]]\n",
    "            orient = tfbs[2]\n",
    "            enh_enc += [tf, orient, aff]\n",
    "    \n",
    "    # Poisin pill\n",
    "    #if row_num == 10000:\n",
    "    #    break\n",
    "    mixed1_encoding.append(enh_enc + [\"-\".join(list(enh_data.values))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885f12c6-900b-480d-b522-df371c9c9ac3",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.694873Z",
     "iopub.status.idle": "2021-11-27T00:18:14.695521Z",
     "shell.execute_reply": "2021-11-27T00:18:14.695212Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mixed-1.0 header\n",
    "header = [\n",
    "    \"L1_length\",\n",
    "    \"TFBS1_type\",\n",
    "    \"TFBS1_orient\",\n",
    "    \"TFBS1_affinity\",\n",
    "    \"L2_length\",\n",
    "    \"TFBS2_type\",\n",
    "    \"TFBS2_orient\",\n",
    "    \"TFBS2_affinity\",\n",
    "    \"L3_length\",\n",
    "    \"TFBS3_type\",\n",
    "    \"TFBS3_orient\",\n",
    "    \"TFBS3_affinity\",\n",
    "    \"L4_length\",\n",
    "    \"TFBS4_type\",\n",
    "    \"TFBS4_orient\",\n",
    "    \"TFBS4_affinity\",\n",
    "    \"L5_length\",\n",
    "    \"TFBS5_type\",\n",
    "    \"TFBS5_orient\",\n",
    "    \"TFBS5_affinity\",\n",
    "    \"L6_length\",\n",
    "    \"Enhancer\"\n",
    "]\n",
    "# Save the header for use down the line. Save as one column per line\n",
    "with open (\"mixed_1.0/mixed-1.0_header.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(header[:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9e67a33-d8f3-4398-afba-5170b972ec37",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2021-11-27T00:18:14.697114Z",
     "iopub.status.idle": "2021-11-27T00:18:14.697757Z",
     "shell.execute_reply": "2021-11-27T00:18:14.697451Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create a dataframe and replace occurrences of TFs and orientation with numbers\n",
    "X_mixed1 = (\n",
    "    pd.DataFrame(mixed1_encoding, columns=header)\n",
    "    .replace({\"G\": 0, \"E\": 1, \"R\": 0, \"F\": 1})\n",
    ")\n",
    "\n",
    "# Drop the enhancer column\n",
    "X_mixed1 = X_mixed1.drop(\"Enhancer\", axis=1).values\n",
    "\n",
    "# Subset train and test\n",
    "X_mixed1_train = X_mixed1[train_index, :]\n",
    "X_mixed1_test = X_mixed1[test_index, :]\n",
    "X_mixed1_trian_and_test = X_mixed1[training_index, :]\n",
    "X_mixed1_holdout = X_mixed1[holdout_index, :]\n",
    "\n",
    "# Standardize features and double check means are 0 and stds are 1\n",
    "scale_indeces = np.array([0, 3, 4, 7, 8, 11, 12, 15, 16, 19, 20])  # Mixed 1.0\n",
    "_, X_mixed1_scaled = project_utils.standardize_features(train_X=X_mixed1_trian_and_test, test_X=X_mixed1_trian_and_test, indeces=scale_indeces, stats_file=\"mixed_1.0/{}_X-training_stats.pickle\".format(PREPROCESS))\n",
    "X_mixed1_train_scaled, X_mixed1_test_scaled = project_utils.standardize_features(train_X=X_mixed1_train, test_X=X_mixed1_test, indeces=scale_indeces, stats_file=\"mixed_1.0/{}_X-train-{}_stats.pickle\".format(PREPROCESS, SPLIT))\n",
    "X_mixed1_train_scaled, X_mixed1_holdout_scaled = project_utils.standardize_features(train_X=X_mixed1_train, test_X=X_mixed1_holdout, indeces=scale_indeces)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad1a322-3856-4975-ad11-c4ac9420a7ea",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Sanity check ordering of current"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "34ef1904-9653-4680-95ad-30f53f58e53a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:11.876284Z",
     "iopub.status.busy": "2021-10-25T04:32:11.875991Z",
     "iopub.status.idle": "2021-10-25T04:32:13.122464Z",
     "shell.execute_reply": "2021-10-25T04:32:13.121750Z",
     "shell.execute_reply.started": "2021-10-25T04:32:11.876213Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "X1 = np.load(\"mixed_1.0/0.18-0.4_X_mixed-1.0.npy\")\n",
    "X1_train = np.load(\"mixed_1.0/0.18-0.4_X_mixed-1.0.npy\")\n",
    "X1_test = np.load(\"mixed_1.0/0.18-0.4_X_mixed-1.0.npy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "b305c1db-6586-40e4-bd18-1f75f0809e6a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:13.123737Z",
     "iopub.status.busy": "2021-10-25T04:32:13.123542Z",
     "iopub.status.idle": "2021-10-25T04:32:15.719541Z",
     "shell.execute_reply": "2021-10-25T04:32:15.718881Z",
     "shell.execute_reply.started": "2021-10-25T04:32:13.123712Z"
    }
   },
   "outputs": [],
   "source": [
    "y1 = np.loadtxt(\"y_binary/0.18-0.4_y-binary.txt\", dtype=float, delimiter=\" \")\n",
    "y1_train = np.loadtxt(\"binary_binary/0.18-0.4_y-train-0.9_binary.txt\", dtype=float, delimiter=\" \")\n",
    "y1_test = np.loadtxt(\"y_binary/0.18-0.4_y-test-0.1_binary.txt\", dtype=int, delimiter=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "efb87240-a9aa-4c30-b81f-0bf0a9fd313e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:15.720987Z",
     "iopub.status.busy": "2021-10-25T04:32:15.720755Z",
     "iopub.status.idle": "2021-10-25T04:32:22.192802Z",
     "shell.execute_reply": "2021-10-25T04:32:22.192116Z",
     "shell.execute_reply.started": "2021-10-25T04:32:15.720962Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "X2 = np.loadtxt(\"mixed_1.0/X_mixed-1.0_0.18-0.4.txt\", dtype=float, delimiter=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "f301b6fc-11ec-42a3-980f-7f74e2ee1daa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:22.193988Z",
     "iopub.status.busy": "2021-10-25T04:32:22.193787Z",
     "iopub.status.idle": "2021-10-25T04:32:23.269231Z",
     "shell.execute_reply": "2021-10-25T04:32:23.268611Z",
     "shell.execute_reply.started": "2021-10-25T04:32:22.193962Z"
    }
   },
   "outputs": [],
   "source": [
    "y2 = y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "cd3a5d0a-3d4e-4c9f-b6c3-5ed019a9a33c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:35.030192Z",
     "iopub.status.busy": "2021-10-25T04:32:35.029930Z",
     "iopub.status.idle": "2021-10-25T04:32:36.068398Z",
     "shell.execute_reply": "2021-10-25T04:32:36.067454Z",
     "shell.execute_reply.started": "2021-10-25T04:32:35.030159Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(y1 == y2).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "a943f720-80d5-4eeb-80af-e539b64b4ac9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:45.545184Z",
     "iopub.status.busy": "2021-10-25T04:32:45.544880Z",
     "iopub.status.idle": "2021-10-25T04:32:46.633677Z",
     "shell.execute_reply": "2021-10-25T04:32:46.632792Z",
     "shell.execute_reply.started": "2021-10-25T04:32:45.545152Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(X1 == X2).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "f9edd550-6891-48a4-97b7-0cefb1ec2446",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:32:47.681600Z",
     "iopub.status.busy": "2021-10-25T04:32:47.681295Z",
     "iopub.status.idle": "2021-10-25T04:32:48.992682Z",
     "shell.execute_reply": "2021-10-25T04:32:48.991987Z",
     "shell.execute_reply.started": "2021-10-25T04:32:47.681566Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "9dcde204-0979-49d5-a3b0-56386c487a10",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-25T04:33:04.512367Z",
     "iopub.status.busy": "2021-10-25T04:33:04.512064Z",
     "iopub.status.idle": "2021-10-25T04:33:05.633373Z",
     "shell.execute_reply": "2021-10-25T04:33:05.632694Z",
     "shell.execute_reply.started": "2021-10-25T04:33:04.512304Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, train_size=0.9, random_state=13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07e6dce3-d178-4c43-ac13-01d47db9bbf8",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Old mixed encoding 1.0 code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "f9241a6e-3c52-47fe-8810-61a01865801c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-21T18:44:15.181688Z",
     "iopub.status.busy": "2021-09-21T18:44:15.181395Z",
     "iopub.status.idle": "2021-09-21T18:44:16.051551Z",
     "shell.execute_reply": "2021-09-21T18:44:16.050669Z",
     "shell.execute_reply.started": "2021-09-21T18:44:15.181654Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2400it [00:00, 7178.25it/s]\n"
     ]
    }
   ],
   "source": [
    "j = 0\n",
    "for i, (row_num, enh_data) in tqdm.tqdm(enumerate(X_mixed2_df.iterrows())):\n",
    "    enhancer = enh_data[\"Enhancer\"]\n",
    "    if \"S5-G2F\" in enhancer:\n",
    "        s5_loc = int((np.where(np.array(enhancer.split(\"-\")) == \"S5\")[0][0] / 2) + 1)\n",
    "        if s5_loc == 1:\n",
    "            continue\n",
    "        linker_to_change_loc = s5_loc - 1\n",
    "        # print(i, row_num, enhancer, s5_loc, linker_to_change_loc, enh_data[\"L{}_length\".format(linker_to_change_loc)])\n",
    "        enh_data[\"L{}_length\".format(linker_to_change_loc)] -= 1\n",
    "        # print(i, row_num, enhancer, s5_loc, linker_to_change_loc, enh_data[\"L{}_length\".format(linker_to_change_loc)])\n",
    "        j += 1\n",
    "    elif \"G2R-S5\" in enhancer:\n",
    "        s5_loc = int((np.where(np.array(enhancer.split(\"-\")) == \"S5\")[0][0] / 2) + 1)\n",
    "        if s5_loc == 6:\n",
    "            continue\n",
    "        linker_to_change_loc = s5_loc + 1\n",
    "        # print(i, row_num, enhancer, s5_loc, linker_to_change_loc, enh_data[\"L{}_length\".format(linker_to_change_loc)])\n",
    "        enh_data[\"L{}_length\".format(linker_to_change_loc)] -= 1\n",
    "        # print(i, row_num, enhancer, s5_loc, linker_to_change_loc, enh_data[\"L{}_length\".format(linker_to_change_loc)])\n",
    "        j += 1\n",
    "    # if j==100:\n",
    "    #    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccf80c6f-c446-4553-8420-2af0952f32e9",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Different thresholds and methods for dealing with class imbalance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fae0345-62ab-4869-8675-6667e514d866",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-07-28T06:03:14.030000Z",
     "iopub.status.busy": "2021-07-28T06:03:14.029719Z",
     "iopub.status.idle": "2021-07-28T06:03:14.471012Z",
     "shell.execute_reply": "2021-07-28T06:03:14.469961Z",
     "shell.execute_reply.started": "2021-07-28T06:03:14.029971Z"
    },
    "tags": []
   },
   "source": [
    "**Dealing with class imbalance**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e1f89baf-1d14-4477-992c-c88f2019147a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:08.981389Z",
     "start_time": "2021-05-20T16:08:08.959611Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:03:17.616882Z",
     "iopub.status.busy": "2021-07-28T06:03:17.616494Z",
     "iopub.status.idle": "2021-07-28T06:03:18.080101Z",
     "shell.execute_reply": "2021-07-28T06:03:18.079307Z",
     "shell.execute_reply.started": "2021-07-28T06:03:17.616849Z"
    }
   },
   "outputs": [],
   "source": [
    "from sklearn.utils import resample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "902c7cc1-0018-4fb1-a9c7-e0c16891c71e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:09.528824Z",
     "start_time": "2021-05-20T16:08:09.500469Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:03:33.867422Z",
     "iopub.status.busy": "2021-07-28T06:03:33.867104Z",
     "iopub.status.idle": "2021-07-28T06:03:34.322075Z",
     "shell.execute_reply": "2021-07-28T06:03:34.321287Z",
     "shell.execute_reply.started": "2021-07-28T06:03:33.867391Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pos_mask = y_train == 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "abacf64a-5796-4d97-a364-4681e14d2ef5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:10.100066Z",
     "start_time": "2021-05-20T16:08:09.984623Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:04:07.086136Z",
     "iopub.status.busy": "2021-07-28T06:04:07.085853Z",
     "iopub.status.idle": "2021-07-28T06:04:07.699945Z",
     "shell.execute_reply": "2021-07-28T06:04:07.699227Z",
     "shell.execute_reply.started": "2021-07-28T06:04:07.086106Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "neg_y = y_train[~pos_mask]\n",
    "pos_y = y_train[pos_mask]\n",
    "\n",
    "neg_X = X_train[~pos_mask, :]\n",
    "pos_X = X_train[pos_mask, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "298c050e-4b4f-4fe4-a525-5e17cf134b20",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:10.573083Z",
     "start_time": "2021-05-20T16:08:10.548778Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:04:09.642591Z",
     "iopub.status.busy": "2021-07-28T06:04:09.642317Z",
     "iopub.status.idle": "2021-07-28T06:04:10.123532Z",
     "shell.execute_reply": "2021-07-28T06:04:10.122674Z",
     "shell.execute_reply.started": "2021-07-28T06:04:09.642561Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([1]), array([65561]))\n",
      "(array([0]), array([349159]))\n"
     ]
    }
   ],
   "source": [
    "print(np.unique(pos_y, return_counts=True))\n",
    "print(np.unique(neg_y, return_counts=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ebf3bbb-a10e-4824-9f41-e6cb11bca8bf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:07:03.797536Z",
     "start_time": "2021-05-20T16:07:03.487420Z"
    }
   },
   "source": [
    "**Downsample negative class**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "4e450c1c-91b1-4f41-b834-4e5b51fe5312",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:11.786396Z",
     "start_time": "2021-05-20T16:08:11.757571Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:08:44.069166Z",
     "iopub.status.busy": "2021-07-28T06:08:44.068868Z",
     "iopub.status.idle": "2021-07-28T06:08:44.507384Z",
     "shell.execute_reply": "2021-07-28T06:08:44.506545Z",
     "shell.execute_reply.started": "2021-07-28T06:08:44.069135Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "downsampled_neg_X, downsampled_neg_y = resample(\n",
    "    neg_X, neg_y, n_samples=len(pos_y), random_state=13\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "eeb37942-4a55-48be-888a-06f4aeb06444",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:08:59.749231Z",
     "start_time": "2021-05-20T16:08:59.696304Z"
    },
    "execution": {
     "iopub.execute_input": "2021-07-28T06:08:44.941585Z",
     "iopub.status.busy": "2021-07-28T06:08:44.941221Z",
     "iopub.status.idle": "2021-07-28T06:08:45.442282Z",
     "shell.execute_reply": "2021-07-28T06:08:45.441491Z",
     "shell.execute_reply.started": "2021-07-28T06:08:44.941540Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "X_train_down = np.concatenate([downsampled_neg_X, pos_X])\n",
    "y_train_down = np.concatenate([downsampled_neg_y, pos_y])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "51b5a98d-2497-4928-b59b-553c0aab0661",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-07-28T06:08:45.443754Z",
     "iopub.status.busy": "2021-07-28T06:08:45.443524Z",
     "iopub.status.idle": "2021-07-28T06:08:45.866128Z",
     "shell.execute_reply": "2021-07-28T06:08:45.865403Z",
     "shell.execute_reply.started": "2021-07-28T06:08:45.443727Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((131122, 75), (131122,))"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train_down.shape, y_train_down.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "90bc1133-4d82-48da-bcf0-1778c30d7964",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-07-28T06:08:45.867625Z",
     "iopub.status.busy": "2021-07-28T06:08:45.867372Z",
     "iopub.status.idle": "2021-07-28T06:08:46.289074Z",
     "shell.execute_reply": "2021-07-28T06:08:46.288265Z",
     "shell.execute_reply.started": "2021-07-28T06:08:45.867597Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.5, 0.5])"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(y_train_down, return_counts=True)[1] / len(y_train_down)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345159f7-4870-41b6-bf98-fe93e42b3307",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Testing mixed encoding code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "795a227c-d128-462b-bcd4-ade5cf6844a1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T22:51:32.449935Z",
     "iopub.status.busy": "2021-10-28T22:51:32.449680Z",
     "iopub.status.idle": "2021-10-28T22:51:33.367697Z",
     "shell.execute_reply": "2021-10-28T22:51:33.366860Z",
     "shell.execute_reply.started": "2021-10-28T22:51:32.449904Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # This is obsolete. DO NOT USE!!!\n",
    "# def correct_s5(x):\n",
    "#     enhancer = x[\"Enhancer\"]\n",
    "#     if \"S5-G2F\" in enhancer:\n",
    "#         s5_loc = int((np.where(np.array(enhancer.split(\"-\")) == \"S5\")[0][0] / 2) + 1)\n",
    "#         if s5_loc == 1:\n",
    "#             return x\n",
    "#         linker_to_change_loc = s5_loc - 1\n",
    "#         x[\"L{}_length\".format(linker_to_change_loc)] -= 1\n",
    "#         return x\n",
    "#     elif \"G2R-S5\" in enhancer:\n",
    "#         s5_loc = int((np.where(np.array(enhancer.split(\"-\")) == \"S5\")[0][0] / 2) + 1)\n",
    "#         if s5_loc == 6:\n",
    "#             return x\n",
    "#         linker_to_change_loc = s5_loc + 1\n",
    "#         x[\"L{}_length\".format(linker_to_change_loc)] -= 1\n",
    "#         return x\n",
    "#     else:\n",
    "#         return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "84ab039c-5b82-4cde-a38b-bd2688502a05",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-21T18:47:51.270972Z",
     "iopub.status.busy": "2021-09-21T18:47:51.270594Z",
     "iopub.status.idle": "2021-09-21T18:47:52.073637Z",
     "shell.execute_reply": "2021-09-21T18:47:52.072664Z",
     "shell.execute_reply.started": "2021-09-21T18:47:51.270926Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(29952, 23969)"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_mixed2_df[\"Enhancer\"].str.contains(\"S5-G2F\").sum(), X_mixed2_df[\n",
    "    \"Enhancer\"\n",
    "].str.contains(\"G2R-S5\").sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "d48e106a-ad91-4d46-9cc8-9815219a8de7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-21T18:50:01.916789Z",
     "iopub.status.busy": "2021-09-21T18:50:01.916516Z",
     "iopub.status.idle": "2021-09-21T18:50:02.404261Z",
     "shell.execute_reply": "2021-09-21T18:50:02.403423Z",
     "shell.execute_reply.started": "2021-09-21T18:50:01.916759Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1437\n",
       "Name: L2_length, dtype: int64"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_mixed2_df_corrected[\n",
    "    (X_mixed2_df_corrected[\"L1_length\"] == 11)\n",
    "    & (X_mixed2_df_corrected[\"TFBS1_GATA_affinity\"] != 0.3)\n",
    "][\"L2_length\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "917eb002-daf7-4395-a81e-4cab35b74706",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-21T18:50:25.011902Z",
     "iopub.status.busy": "2021-09-21T18:50:25.011604Z",
     "iopub.status.idle": "2021-09-21T18:50:25.455695Z",
     "shell.execute_reply": "2021-09-21T18:50:25.454710Z",
     "shell.execute_reply.started": "2021-09-21T18:50:25.011871Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    6037\n",
       "Name: L5_length, dtype: int64"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_mixed2_df_corrected[\n",
    "    (X_mixed2_df_corrected[\"L6_length\"] == 0)\n",
    "    & (X_mixed2_df_corrected[\"TFBS5_GATA_affinity\"] != 0.3)\n",
    "][\"L5_length\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "cdacf906-9f63-40a2-a5be-6971241ed781",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-21T18:52:46.338326Z",
     "iopub.status.busy": "2021-09-21T18:52:46.338011Z",
     "iopub.status.idle": "2021-09-21T18:52:46.983842Z",
     "shell.execute_reply": "2021-09-21T18:52:46.983006Z",
     "shell.execute_reply.started": "2021-09-21T18:52:46.338293Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "X_mixed2_df_corrected = X_mixed2_df_corrected.drop(\"Enhancer\", axis=1).values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fc2f529-eabb-4b8a-a4f6-dfa507e4bc0c",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Old one hot encoding code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "35b3c528-30c3-45f2-b737-6f85b7216701",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:41:05.759549Z",
     "iopub.status.busy": "2021-10-28T17:41:05.759372Z",
     "iopub.status.idle": "2021-10-28T17:41:10.227760Z",
     "shell.execute_reply": "2021-10-28T17:41:10.227212Z",
     "shell.execute_reply.started": "2021-10-28T17:41:05.759532Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get the sequences only\n",
    "X_rev_seqs = np.array([\"\".join({'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}.get(base, base) for base in reversed(seq)) for seq in X_seqs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "66cd9a80-f8d9-4d65-944f-c3c2ca772bad",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:41:24.840734Z",
     "iopub.status.busy": "2021-10-28T17:41:24.840569Z",
     "iopub.status.idle": "2021-10-28T17:41:25.682562Z",
     "shell.execute_reply": "2021-10-28T17:41:25.682195Z",
     "shell.execute_reply.started": "2021-10-28T17:41:24.840718Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define encoders\n",
    "integer_encoder = LabelEncoder()\n",
    "one_hot_encoder = OneHotEncoder(\n",
    "    categories=[np.array([0, 1, 2, 3])], handle_unknown=\"ignore\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6ef5842f-bbd3-40a9-b9b4-5e825d5658fc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:41:25.750074Z",
     "iopub.status.busy": "2021-10-28T17:41:25.749936Z",
     "iopub.status.idle": "2021-10-28T17:41:26.554730Z",
     "shell.execute_reply": "2021-10-28T17:41:26.554356Z",
     "shell.execute_reply.started": "2021-10-28T17:41:25.750059Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCCTA...\n",
      "[3 1 1 3 0]...\n",
      "[[0. 0. 0. 1.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [1. 0. 0. 0.]]...\n"
     ]
    }
   ],
   "source": [
    "# Example steps for one hot encoding\n",
    "test = X_seqs[0]\n",
    "print(\"{}...\".format(test[:5]))\n",
    "integer_encoded = integer_encoder.fit_transform(list(test))\n",
    "print(\"{}...\".format(integer_encoded[:5]))\n",
    "integer_encoded = np.array(integer_encoded).reshape(-1, 1)\n",
    "one_hot_encoder.fit(integer_encoded)  # convert to one hot\n",
    "one_hot_encoded = one_hot_encoder.fit_transform(integer_encoded)\n",
    "print(\"{}...\".format(one_hot_encoded.toarray()[:5]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50e4df25-ac96-4669-824e-5e7543108165",
   "metadata": {},
   "source": [
    "**One hot encoding of sequences...<u>takes about 2 minutes</u>**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6d883a74-d6a1-4223-8b04-ea95ae2316b9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:41:31.704471Z",
     "iopub.status.busy": "2021-10-28T17:41:31.704278Z",
     "iopub.status.idle": "2021-10-28T17:43:12.459524Z",
     "shell.execute_reply": "2021-10-28T17:43:12.458952Z",
     "shell.execute_reply.started": "2021-10-28T17:41:31.704446Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 302936/302936 [01:39<00:00, 3030.54it/s]\n"
     ]
    }
   ],
   "source": [
    "# Applying above for all seqs\n",
    "X_features = []  # will hold one hot encoded sequence\n",
    "for i, seq in enumerate(tqdm.tqdm(X_seqs)):\n",
    "    integer_encoded = integer_encoder.fit_transform(list(seq))  # convert to integer\n",
    "    integer_encoded = np.array(integer_encoded).reshape(-1, 1)\n",
    "    one_hot_encoder.fit(integer_encoded)  # convert to one hot\n",
    "    one_hot_encoded = one_hot_encoder.fit_transform(integer_encoded)\n",
    "    X_features.append(one_hot_encoded.toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "708c81d3-fafc-4b7c-8171-59607192b3cb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:43:14.884717Z",
     "iopub.status.busy": "2021-10-28T17:43:14.884549Z",
     "iopub.status.idle": "2021-10-28T17:43:15.746849Z",
     "shell.execute_reply": "2021-10-28T17:43:15.746457Z",
     "shell.execute_reply.started": "2021-10-28T17:43:14.884692Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "302936"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to make sure it is correct length\n",
    "len(X_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4f7402b0-9752-4462-9a0f-3a4f0cfeb9ca",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:43:16.597083Z",
     "iopub.status.busy": "2021-10-28T17:43:16.596907Z",
     "iopub.status.idle": "2021-10-28T17:43:17.841920Z",
     "shell.execute_reply": "2021-10-28T17:43:17.841483Z",
     "shell.execute_reply.started": "2021-10-28T17:43:16.597064Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# convert to numpy array\n",
    "X_ohe_seq = np.array(X_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2cc0d544-7970-4835-8a8b-7ac3fa07d20f",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2021-10-28T17:43:19.270748Z",
     "iopub.status.busy": "2021-10-28T17:43:19.270562Z",
     "iopub.status.idle": "2021-10-28T17:43:20.799025Z",
     "shell.execute_reply": "2021-10-28T17:43:20.798645Z",
     "shell.execute_reply.started": "2021-10-28T17:43:19.270725Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq #1 encoded correctly\n",
      "Seq #2 encoded correctly\n",
      "Seq #3 encoded correctly\n",
      "Seq #4 encoded correctly\n",
      "Seq #5 encoded correctly\n",
      "Seq #6 encoded correctly\n",
      "Seq #7 encoded correctly\n",
      "Seq #8 encoded correctly\n",
      "Seq #9 encoded correctly\n",
      "Seq #10 encoded correctly\n",
      "Seq #11 encoded correctly\n",
      "Seq #12 encoded correctly\n",
      "Seq #13 encoded correctly\n",
      "Seq #14 encoded correctly\n",
      "Seq #15 encoded correctly\n",
      "Seq #16 encoded correctly\n",
      "Seq #17 encoded correctly\n",
      "Seq #18 encoded correctly\n",
      "Seq #19 encoded correctly\n",
      "Seq #20 encoded correctly\n",
      "Seq #21 encoded correctly\n",
      "Seq #22 encoded correctly\n",
      "Seq #23 encoded correctly\n",
      "Seq #24 encoded correctly\n",
      "Seq #25 encoded correctly\n",
      "Seq #26 encoded correctly\n",
      "Seq #27 encoded correctly\n",
      "Seq #28 encoded correctly\n",
      "Seq #29 encoded correctly\n",
      "Seq #30 encoded correctly\n",
      "Seq #31 encoded correctly\n",
      "Seq #32 encoded correctly\n",
      "Seq #33 encoded correctly\n",
      "Seq #34 encoded correctly\n",
      "Seq #35 encoded correctly\n",
      "Seq #36 encoded correctly\n",
      "Seq #37 encoded correctly\n",
      "Seq #38 encoded correctly\n",
      "Seq #39 encoded correctly\n",
      "Seq #40 encoded correctly\n",
      "Seq #41 encoded correctly\n",
      "Seq #42 encoded correctly\n",
      "Seq #43 encoded correctly\n",
      "Seq #44 encoded correctly\n",
      "Seq #45 encoded correctly\n",
      "Seq #46 encoded correctly\n",
      "Seq #47 encoded correctly\n",
      "Seq #48 encoded correctly\n",
      "Seq #49 encoded correctly\n",
      "Seq #50 encoded correctly\n",
      "Seq #51 encoded correctly\n",
      "Seq #52 encoded correctly\n",
      "Seq #53 encoded correctly\n",
      "Seq #54 encoded correctly\n",
      "Seq #55 encoded correctly\n",
      "Seq #56 encoded correctly\n",
      "Seq #57 encoded correctly\n",
      "Seq #58 encoded correctly\n",
      "Seq #59 encoded correctly\n",
      "Seq #60 encoded correctly\n",
      "Seq #61 encoded correctly\n",
      "Seq #62 encoded correctly\n",
      "Seq #63 encoded correctly\n",
      "Seq #64 encoded correctly\n",
      "Seq #65 encoded correctly\n",
      "Seq #66 encoded correctly\n",
      "Seq #67 encoded correctly\n",
      "Seq #68 encoded correctly\n",
      "Seq #69 encoded correctly\n",
      "Seq #70 encoded correctly\n",
      "Seq #71 encoded correctly\n",
      "Seq #72 encoded correctly\n",
      "Seq #73 encoded correctly\n",
      "Seq #74 encoded correctly\n",
      "Seq #75 encoded correctly\n",
      "Seq #76 encoded correctly\n",
      "Seq #77 encoded correctly\n",
      "Seq #78 encoded correctly\n",
      "Seq #79 encoded correctly\n",
      "Seq #80 encoded correctly\n",
      "Seq #81 encoded correctly\n",
      "Seq #82 encoded correctly\n",
      "Seq #83 encoded correctly\n",
      "Seq #84 encoded correctly\n",
      "Seq #85 encoded correctly\n",
      "Seq #86 encoded correctly\n",
      "Seq #87 encoded correctly\n",
      "Seq #88 encoded correctly\n",
      "Seq #89 encoded correctly\n",
      "Seq #90 encoded correctly\n",
      "Seq #91 encoded correctly\n",
      "Seq #92 encoded correctly\n",
      "Seq #93 encoded correctly\n",
      "Seq #94 encoded correctly\n",
      "Seq #95 encoded correctly\n",
      "Seq #96 encoded correctly\n",
      "Seq #97 encoded correctly\n",
      "Seq #98 encoded correctly\n",
      "Seq #99 encoded correctly\n",
      "Seq #100 encoded correctly\n",
      "Seq #101 encoded correctly\n",
      "Seq #102 encoded correctly\n",
      "Seq #103 encoded correctly\n",
      "Seq #104 encoded correctly\n",
      "Seq #105 encoded correctly\n",
      "Seq #106 encoded correctly\n",
      "Seq #107 encoded correctly\n",
      "Seq #108 encoded correctly\n",
      "Seq #109 encoded correctly\n",
      "Seq #110 encoded correctly\n",
      "Seq #111 encoded correctly\n",
      "Seq #112 encoded correctly\n",
      "Seq #113 encoded correctly\n",
      "Seq #114 encoded correctly\n",
      "Seq #115 encoded correctly\n",
      "Seq #116 encoded correctly\n",
      "Seq #117 encoded correctly\n",
      "Seq #118 encoded correctly\n",
      "Seq #119 encoded correctly\n",
      "Seq #120 encoded correctly\n",
      "Seq #121 encoded correctly\n",
      "Seq #122 encoded correctly\n",
      "Seq #123 encoded correctly\n",
      "Seq #124 encoded correctly\n",
      "Seq #125 encoded correctly\n",
      "Seq #126 encoded correctly\n",
      "Seq #127 encoded correctly\n",
      "Seq #128 encoded correctly\n",
      "Seq #129 encoded correctly\n",
      "Seq #130 encoded correctly\n",
      "Seq #131 encoded correctly\n",
      "Seq #132 encoded correctly\n",
      "Seq #133 encoded correctly\n",
      "Seq #134 encoded correctly\n",
      "Seq #135 encoded correctly\n",
      "Seq #136 encoded correctly\n",
      "Seq #137 encoded correctly\n",
      "Seq #138 encoded correctly\n",
      "Seq #139 encoded correctly\n",
      "Seq #140 encoded correctly\n",
      "Seq #141 encoded correctly\n",
      "Seq #142 encoded correctly\n",
      "Seq #143 encoded correctly\n",
      "Seq #144 encoded correctly\n",
      "Seq #145 encoded correctly\n",
      "Seq #146 encoded correctly\n",
      "Seq #147 encoded correctly\n",
      "Seq #148 encoded correctly\n",
      "Seq #149 encoded correctly\n",
      "Seq #150 encoded correctly\n",
      "Seq #151 encoded correctly\n",
      "Seq #152 encoded correctly\n",
      "Seq #153 encoded correctly\n",
      "Seq #154 encoded correctly\n",
      "Seq #155 encoded correctly\n",
      "Seq #156 encoded correctly\n",
      "Seq #157 encoded correctly\n",
      "Seq #158 encoded correctly\n",
      "Seq #159 encoded correctly\n",
      "Seq #160 encoded correctly\n",
      "Seq #161 encoded correctly\n",
      "Seq #162 encoded correctly\n",
      "Seq #163 encoded correctly\n",
      "Seq #164 encoded correctly\n",
      "Seq #165 encoded correctly\n",
      "Seq #166 encoded correctly\n",
      "Seq #167 encoded correctly\n",
      "Seq #168 encoded correctly\n",
      "Seq #169 encoded correctly\n",
      "Seq #170 encoded correctly\n",
      "Seq #171 encoded correctly\n",
      "Seq #172 encoded correctly\n",
      "Seq #173 encoded correctly\n",
      "Seq #174 encoded correctly\n",
      "Seq #175 encoded correctly\n",
      "Seq #176 encoded correctly\n",
      "Seq #177 encoded correctly\n",
      "Seq #178 encoded correctly\n",
      "Seq #179 encoded correctly\n",
      "Seq #180 encoded correctly\n",
      "Seq #181 encoded correctly\n",
      "Seq #182 encoded correctly\n",
      "Seq #183 encoded correctly\n",
      "Seq #184 encoded correctly\n",
      "Seq #185 encoded correctly\n",
      "Seq #186 encoded correctly\n",
      "Seq #187 encoded correctly\n",
      "Seq #188 encoded correctly\n",
      "Seq #189 encoded correctly\n",
      "Seq #190 encoded correctly\n",
      "Seq #191 encoded correctly\n",
      "Seq #192 encoded correctly\n",
      "Seq #193 encoded correctly\n",
      "Seq #194 encoded correctly\n",
      "Seq #195 encoded correctly\n",
      "Seq #196 encoded correctly\n",
      "Seq #197 encoded correctly\n",
      "Seq #198 encoded correctly\n",
      "Seq #199 encoded correctly\n",
      "Seq #200 encoded correctly\n",
      "Seq #201 encoded correctly\n",
      "Seq #202 encoded correctly\n",
      "Seq #203 encoded correctly\n",
      "Seq #204 encoded correctly\n",
      "Seq #205 encoded correctly\n",
      "Seq #206 encoded correctly\n",
      "Seq #207 encoded correctly\n",
      "Seq #208 encoded correctly\n",
      "Seq #209 encoded correctly\n",
      "Seq #210 encoded correctly\n",
      "Seq #211 encoded correctly\n",
      "Seq #212 encoded correctly\n",
      "Seq #213 encoded correctly\n",
      "Seq #214 encoded correctly\n",
      "Seq #215 encoded correctly\n",
      "Seq #216 encoded correctly\n",
      "Seq #217 encoded correctly\n",
      "Seq #218 encoded correctly\n",
      "Seq #219 encoded correctly\n",
      "Seq #220 encoded correctly\n",
      "Seq #221 encoded correctly\n",
      "Seq #222 encoded correctly\n",
      "Seq #223 encoded correctly\n",
      "Seq #224 encoded correctly\n",
      "Seq #225 encoded correctly\n",
      "Seq #226 encoded correctly\n",
      "Seq #227 encoded correctly\n",
      "Seq #228 encoded correctly\n",
      "Seq #229 encoded correctly\n",
      "Seq #230 encoded correctly\n",
      "Seq #231 encoded correctly\n",
      "Seq #232 encoded correctly\n",
      "Seq #233 encoded correctly\n",
      "Seq #234 encoded correctly\n",
      "Seq #235 encoded correctly\n",
      "Seq #236 encoded correctly\n",
      "Seq #237 encoded correctly\n",
      "Seq #238 encoded correctly\n",
      "Seq #239 encoded correctly\n",
      "Seq #240 encoded correctly\n",
      "Seq #241 encoded correctly\n",
      "Seq #242 encoded correctly\n",
      "Seq #243 encoded correctly\n",
      "Seq #244 encoded correctly\n",
      "Seq #245 encoded correctly\n",
      "Seq #246 encoded correctly\n",
      "Seq #247 encoded correctly\n",
      "Seq #248 encoded correctly\n",
      "Seq #249 encoded correctly\n",
      "Seq #250 encoded correctly\n",
      "Seq #251 encoded correctly\n",
      "Seq #252 encoded correctly\n",
      "Seq #253 encoded correctly\n",
      "Seq #254 encoded correctly\n",
      "Seq #255 encoded correctly\n",
      "Seq #256 encoded correctly\n",
      "Seq #257 encoded correctly\n",
      "Seq #258 encoded correctly\n",
      "Seq #259 encoded correctly\n",
      "Seq #260 encoded correctly\n",
      "Seq #261 encoded correctly\n",
      "Seq #262 encoded correctly\n",
      "Seq #263 encoded correctly\n",
      "Seq #264 encoded correctly\n",
      "Seq #265 encoded correctly\n",
      "Seq #266 encoded correctly\n",
      "Seq #267 encoded correctly\n",
      "Seq #268 encoded correctly\n",
      "Seq #269 encoded correctly\n",
      "Seq #270 encoded correctly\n",
      "Seq #271 encoded correctly\n",
      "Seq #272 encoded correctly\n",
      "Seq #273 encoded correctly\n",
      "Seq #274 encoded correctly\n",
      "Seq #275 encoded correctly\n",
      "Seq #276 encoded correctly\n",
      "Seq #277 encoded correctly\n",
      "Seq #278 encoded correctly\n",
      "Seq #279 encoded correctly\n",
      "Seq #280 encoded correctly\n",
      "Seq #281 encoded correctly\n",
      "Seq #282 encoded correctly\n",
      "Seq #283 encoded correctly\n",
      "Seq #284 encoded correctly\n",
      "Seq #285 encoded correctly\n",
      "Seq #286 encoded correctly\n",
      "Seq #287 encoded correctly\n",
      "Seq #288 encoded correctly\n",
      "Seq #289 encoded correctly\n",
      "Seq #290 encoded correctly\n",
      "Seq #291 encoded correctly\n",
      "Seq #292 encoded correctly\n",
      "Seq #293 encoded correctly\n",
      "Seq #294 encoded correctly\n",
      "Seq #295 encoded correctly\n",
      "Seq #296 encoded correctly\n",
      "Seq #297 encoded correctly\n",
      "Seq #298 encoded correctly\n",
      "Seq #299 encoded correctly\n",
      "Seq #300 encoded correctly\n",
      "Seq #301 encoded correctly\n",
      "Seq #302 encoded correctly\n",
      "Seq #303 encoded correctly\n",
      "Seq #304 encoded correctly\n",
      "Seq #305 encoded correctly\n",
      "Seq #306 encoded correctly\n",
      "Seq #307 encoded correctly\n",
      "Seq #308 encoded correctly\n",
      "Seq #309 encoded correctly\n",
      "Seq #310 encoded correctly\n",
      "Seq #311 encoded correctly\n",
      "Seq #312 encoded correctly\n",
      "Seq #313 encoded correctly\n",
      "Seq #314 encoded correctly\n",
      "Seq #315 encoded correctly\n",
      "Seq #316 encoded correctly\n",
      "Seq #317 encoded correctly\n",
      "Seq #318 encoded correctly\n",
      "Seq #319 encoded correctly\n",
      "Seq #320 encoded correctly\n",
      "Seq #321 encoded correctly\n",
      "Seq #322 encoded correctly\n",
      "Seq #323 encoded correctly\n",
      "Seq #324 encoded correctly\n",
      "Seq #325 encoded correctly\n",
      "Seq #326 encoded correctly\n",
      "Seq #327 encoded correctly\n",
      "Seq #328 encoded correctly\n",
      "Seq #329 encoded correctly\n",
      "Seq #330 encoded correctly\n",
      "Seq #331 encoded correctly\n",
      "Seq #332 encoded correctly\n",
      "Seq #333 encoded correctly\n",
      "Seq #334 encoded correctly\n",
      "Seq #335 encoded correctly\n",
      "Seq #336 encoded correctly\n",
      "Seq #337 encoded correctly\n",
      "Seq #338 encoded correctly\n",
      "Seq #339 encoded correctly\n",
      "Seq #340 encoded correctly\n",
      "Seq #341 encoded correctly\n",
      "Seq #342 encoded correctly\n",
      "Seq #343 encoded correctly\n",
      "Seq #344 encoded correctly\n",
      "Seq #345 encoded correctly\n",
      "Seq #346 encoded correctly\n",
      "Seq #347 encoded correctly\n",
      "Seq #348 encoded correctly\n",
      "Seq #349 encoded correctly\n",
      "Seq #350 encoded correctly\n",
      "Seq #351 encoded correctly\n",
      "Seq #352 encoded correctly\n",
      "Seq #353 encoded correctly\n",
      "Seq #354 encoded correctly\n",
      "Seq #355 encoded correctly\n",
      "Seq #356 encoded correctly\n",
      "Seq #357 encoded correctly\n",
      "Seq #358 encoded correctly\n",
      "Seq #359 encoded correctly\n",
      "Seq #360 encoded correctly\n",
      "Seq #361 encoded correctly\n",
      "Seq #362 encoded correctly\n",
      "Seq #363 encoded correctly\n",
      "Seq #364 encoded correctly\n",
      "Seq #365 encoded correctly\n",
      "Seq #366 encoded correctly\n",
      "Seq #367 encoded correctly\n",
      "Seq #368 encoded correctly\n",
      "Seq #369 encoded correctly\n",
      "Seq #370 encoded correctly\n",
      "Seq #371 encoded correctly\n",
      "Seq #372 encoded correctly\n",
      "Seq #373 encoded correctly\n",
      "Seq #374 encoded correctly\n",
      "Seq #375 encoded correctly\n",
      "Seq #376 encoded correctly\n",
      "Seq #377 encoded correctly\n",
      "Seq #378 encoded correctly\n",
      "Seq #379 encoded correctly\n",
      "Seq #380 encoded correctly\n",
      "Seq #381 encoded correctly\n",
      "Seq #382 encoded correctly\n",
      "Seq #383 encoded correctly\n",
      "Seq #384 encoded correctly\n",
      "Seq #385 encoded correctly\n",
      "Seq #386 encoded correctly\n",
      "Seq #387 encoded correctly\n",
      "Seq #388 encoded correctly\n",
      "Seq #389 encoded correctly\n",
      "Seq #390 encoded correctly\n",
      "Seq #391 encoded correctly\n",
      "Seq #392 encoded correctly\n",
      "Seq #393 encoded correctly\n",
      "Seq #394 encoded correctly\n",
      "Seq #395 encoded correctly\n",
      "Seq #396 encoded correctly\n",
      "Seq #397 encoded correctly\n",
      "Seq #398 encoded correctly\n",
      "Seq #399 encoded correctly\n",
      "Seq #400 encoded correctly\n",
      "Seq #401 encoded correctly\n",
      "Seq #402 encoded correctly\n",
      "Seq #403 encoded correctly\n",
      "Seq #404 encoded correctly\n",
      "Seq #405 encoded correctly\n",
      "Seq #406 encoded correctly\n",
      "Seq #407 encoded correctly\n",
      "Seq #408 encoded correctly\n",
      "Seq #409 encoded correctly\n",
      "Seq #410 encoded correctly\n",
      "Seq #411 encoded correctly\n",
      "Seq #412 encoded correctly\n",
      "Seq #413 encoded correctly\n",
      "Seq #414 encoded correctly\n",
      "Seq #415 encoded correctly\n",
      "Seq #416 encoded correctly\n",
      "Seq #417 encoded correctly\n",
      "Seq #418 encoded correctly\n",
      "Seq #419 encoded correctly\n",
      "Seq #420 encoded correctly\n",
      "Seq #421 encoded correctly\n",
      "Seq #422 encoded correctly\n",
      "Seq #423 encoded correctly\n",
      "Seq #424 encoded correctly\n",
      "Seq #425 encoded correctly\n",
      "Seq #426 encoded correctly\n",
      "Seq #427 encoded correctly\n",
      "Seq #428 encoded correctly\n",
      "Seq #429 encoded correctly\n",
      "Seq #430 encoded correctly\n",
      "Seq #431 encoded correctly\n",
      "Seq #432 encoded correctly\n",
      "Seq #433 encoded correctly\n",
      "Seq #434 encoded correctly\n",
      "Seq #435 encoded correctly\n",
      "Seq #436 encoded correctly\n",
      "Seq #437 encoded correctly\n",
      "Seq #438 encoded correctly\n",
      "Seq #439 encoded correctly\n",
      "Seq #440 encoded correctly\n",
      "Seq #441 encoded correctly\n",
      "Seq #442 encoded correctly\n",
      "Seq #443 encoded correctly\n",
      "Seq #444 encoded correctly\n",
      "Seq #445 encoded correctly\n",
      "Seq #446 encoded correctly\n",
      "Seq #447 encoded correctly\n",
      "Seq #448 encoded correctly\n",
      "Seq #449 encoded correctly\n",
      "Seq #450 encoded correctly\n",
      "Seq #451 encoded correctly\n",
      "Seq #452 encoded correctly\n",
      "Seq #453 encoded correctly\n",
      "Seq #454 encoded correctly\n",
      "Seq #455 encoded correctly\n",
      "Seq #456 encoded correctly\n",
      "Seq #457 encoded correctly\n",
      "Seq #458 encoded correctly\n",
      "Seq #459 encoded correctly\n",
      "Seq #460 encoded correctly\n",
      "Seq #461 encoded correctly\n",
      "Seq #462 encoded correctly\n",
      "Seq #463 encoded correctly\n",
      "Seq #464 encoded correctly\n",
      "Seq #465 encoded correctly\n",
      "Seq #466 encoded correctly\n",
      "Seq #467 encoded correctly\n",
      "Seq #468 encoded correctly\n",
      "Seq #469 encoded correctly\n",
      "Seq #470 encoded correctly\n",
      "Seq #471 encoded correctly\n",
      "Seq #472 encoded correctly\n",
      "Seq #473 encoded correctly\n",
      "Seq #474 encoded correctly\n",
      "Seq #475 encoded correctly\n",
      "Seq #476 encoded correctly\n",
      "Seq #477 encoded correctly\n",
      "Seq #478 encoded correctly\n",
      "Seq #479 encoded correctly\n",
      "Seq #480 encoded correctly\n",
      "Seq #481 encoded correctly\n",
      "Seq #482 encoded correctly\n",
      "Seq #483 encoded correctly\n",
      "Seq #484 encoded correctly\n",
      "Seq #485 encoded correctly\n",
      "Seq #486 encoded correctly\n",
      "Seq #487 encoded correctly\n",
      "Seq #488 encoded correctly\n",
      "Seq #489 encoded correctly\n",
      "Seq #490 encoded correctly\n",
      "Seq #491 encoded correctly\n",
      "Seq #492 encoded correctly\n",
      "Seq #493 encoded correctly\n",
      "Seq #494 encoded correctly\n",
      "Seq #495 encoded correctly\n",
      "Seq #496 encoded correctly\n",
      "Seq #497 encoded correctly\n",
      "Seq #498 encoded correctly\n",
      "Seq #499 encoded correctly\n",
      "Seq #500 encoded correctly\n",
      "Seq #501 encoded correctly\n",
      "Seq #502 encoded correctly\n",
      "Seq #503 encoded correctly\n",
      "Seq #504 encoded correctly\n",
      "Seq #505 encoded correctly\n",
      "Seq #506 encoded correctly\n",
      "Seq #507 encoded correctly\n",
      "Seq #508 encoded correctly\n",
      "Seq #509 encoded correctly\n",
      "Seq #510 encoded correctly\n",
      "Seq #511 encoded correctly\n",
      "Seq #512 encoded correctly\n",
      "Seq #513 encoded correctly\n",
      "Seq #514 encoded correctly\n",
      "Seq #515 encoded correctly\n",
      "Seq #516 encoded correctly\n",
      "Seq #517 encoded correctly\n",
      "Seq #518 encoded correctly\n",
      "Seq #519 encoded correctly\n",
      "Seq #520 encoded correctly\n",
      "Seq #521 encoded correctly\n",
      "Seq #522 encoded correctly\n",
      "Seq #523 encoded correctly\n",
      "Seq #524 encoded correctly\n",
      "Seq #525 encoded correctly\n",
      "Seq #526 encoded correctly\n",
      "Seq #527 encoded correctly\n",
      "Seq #528 encoded correctly\n",
      "Seq #529 encoded correctly\n",
      "Seq #530 encoded correctly\n",
      "Seq #531 encoded correctly\n",
      "Seq #532 encoded correctly\n",
      "Seq #533 encoded correctly\n",
      "Seq #534 encoded correctly\n",
      "Seq #535 encoded correctly\n",
      "Seq #536 encoded correctly\n",
      "Seq #537 encoded correctly\n",
      "Seq #538 encoded correctly\n",
      "Seq #539 encoded correctly\n",
      "Seq #540 encoded correctly\n",
      "Seq #541 encoded correctly\n",
      "Seq #542 encoded correctly\n",
      "Seq #543 encoded correctly\n",
      "Seq #544 encoded correctly\n",
      "Seq #545 encoded correctly\n",
      "Seq #546 encoded correctly\n",
      "Seq #547 encoded correctly\n",
      "Seq #548 encoded correctly\n",
      "Seq #549 encoded correctly\n",
      "Seq #550 encoded correctly\n",
      "Seq #551 encoded correctly\n",
      "Seq #552 encoded correctly\n",
      "Seq #553 encoded correctly\n",
      "Seq #554 encoded correctly\n",
      "Seq #555 encoded correctly\n",
      "Seq #556 encoded correctly\n",
      "Seq #557 encoded correctly\n",
      "Seq #558 encoded correctly\n",
      "Seq #559 encoded correctly\n",
      "Seq #560 encoded correctly\n",
      "Seq #561 encoded correctly\n",
      "Seq #562 encoded correctly\n",
      "Seq #563 encoded correctly\n",
      "Seq #564 encoded correctly\n",
      "Seq #565 encoded correctly\n",
      "Seq #566 encoded correctly\n",
      "Seq #567 encoded correctly\n",
      "Seq #568 encoded correctly\n",
      "Seq #569 encoded correctly\n",
      "Seq #570 encoded correctly\n",
      "Seq #571 encoded correctly\n",
      "Seq #572 encoded correctly\n",
      "Seq #573 encoded correctly\n",
      "Seq #574 encoded correctly\n",
      "Seq #575 encoded correctly\n",
      "Seq #576 encoded correctly\n",
      "Seq #577 encoded correctly\n",
      "Seq #578 encoded correctly\n",
      "Seq #579 encoded correctly\n",
      "Seq #580 encoded correctly\n",
      "Seq #581 encoded correctly\n",
      "Seq #582 encoded correctly\n",
      "Seq #583 encoded correctly\n",
      "Seq #584 encoded correctly\n",
      "Seq #585 encoded correctly\n",
      "Seq #586 encoded correctly\n",
      "Seq #587 encoded correctly\n",
      "Seq #588 encoded correctly\n",
      "Seq #589 encoded correctly\n",
      "Seq #590 encoded correctly\n",
      "Seq #591 encoded correctly\n",
      "Seq #592 encoded correctly\n",
      "Seq #593 encoded correctly\n",
      "Seq #594 encoded correctly\n",
      "Seq #595 encoded correctly\n",
      "Seq #596 encoded correctly\n",
      "Seq #597 encoded correctly\n",
      "Seq #598 encoded correctly\n",
      "Seq #599 encoded correctly\n",
      "Seq #600 encoded correctly\n",
      "Seq #601 encoded correctly\n",
      "Seq #602 encoded correctly\n",
      "Seq #603 encoded correctly\n",
      "Seq #604 encoded correctly\n",
      "Seq #605 encoded correctly\n",
      "Seq #606 encoded correctly\n",
      "Seq #607 encoded correctly\n",
      "Seq #608 encoded correctly\n",
      "Seq #609 encoded correctly\n",
      "Seq #610 encoded correctly\n",
      "Seq #611 encoded correctly\n",
      "Seq #612 encoded correctly\n",
      "Seq #613 encoded correctly\n",
      "Seq #614 encoded correctly\n",
      "Seq #615 encoded correctly\n",
      "Seq #616 encoded correctly\n",
      "Seq #617 encoded correctly\n",
      "Seq #618 encoded correctly\n",
      "Seq #619 encoded correctly\n",
      "Seq #620 encoded correctly\n",
      "Seq #621 encoded correctly\n",
      "Seq #622 encoded correctly\n",
      "Seq #623 encoded correctly\n",
      "Seq #624 encoded correctly\n",
      "Seq #625 encoded correctly\n",
      "Seq #626 encoded correctly\n",
      "Seq #627 encoded correctly\n",
      "Seq #628 encoded correctly\n",
      "Seq #629 encoded correctly\n",
      "Seq #630 encoded correctly\n",
      "Seq #631 encoded correctly\n",
      "Seq #632 encoded correctly\n",
      "Seq #633 encoded correctly\n",
      "Seq #634 encoded correctly\n",
      "Seq #635 encoded correctly\n",
      "Seq #636 encoded correctly\n",
      "Seq #637 encoded correctly\n",
      "Seq #638 encoded correctly\n",
      "Seq #639 encoded correctly\n",
      "Seq #640 encoded correctly\n",
      "Seq #641 encoded correctly\n",
      "Seq #642 encoded correctly\n",
      "Seq #643 encoded correctly\n",
      "Seq #644 encoded correctly\n",
      "Seq #645 encoded correctly\n",
      "Seq #646 encoded correctly\n",
      "Seq #647 encoded correctly\n",
      "Seq #648 encoded correctly\n",
      "Seq #649 encoded correctly\n",
      "Seq #650 encoded correctly\n",
      "Seq #651 encoded correctly\n",
      "Seq #652 encoded correctly\n",
      "Seq #653 encoded correctly\n",
      "Seq #654 encoded correctly\n",
      "Seq #655 encoded correctly\n",
      "Seq #656 encoded correctly\n",
      "Seq #657 encoded correctly\n",
      "Seq #658 encoded correctly\n",
      "Seq #659 encoded correctly\n",
      "Seq #660 encoded correctly\n",
      "Seq #661 encoded correctly\n",
      "Seq #662 encoded correctly\n",
      "Seq #663 encoded correctly\n",
      "Seq #664 encoded correctly\n",
      "Seq #665 encoded correctly\n",
      "Seq #666 encoded correctly\n",
      "Seq #667 encoded correctly\n",
      "Seq #668 encoded correctly\n",
      "Seq #669 encoded correctly\n",
      "Seq #670 encoded correctly\n",
      "Seq #671 encoded correctly\n",
      "Seq #672 encoded correctly\n",
      "Seq #673 encoded correctly\n",
      "Seq #674 encoded correctly\n",
      "Seq #675 encoded correctly\n",
      "Seq #676 encoded correctly\n",
      "Seq #677 encoded correctly\n",
      "Seq #678 encoded correctly\n",
      "Seq #679 encoded correctly\n",
      "Seq #680 encoded correctly\n",
      "Seq #681 encoded correctly\n",
      "Seq #682 encoded correctly\n",
      "Seq #683 encoded correctly\n",
      "Seq #684 encoded correctly\n",
      "Seq #685 encoded correctly\n",
      "Seq #686 encoded correctly\n",
      "Seq #687 encoded correctly\n",
      "Seq #688 encoded correctly\n",
      "Seq #689 encoded correctly\n",
      "Seq #690 encoded correctly\n",
      "Seq #691 encoded correctly\n",
      "Seq #692 encoded correctly\n",
      "Seq #693 encoded correctly\n",
      "Seq #694 encoded correctly\n",
      "Seq #695 encoded correctly\n",
      "Seq #696 encoded correctly\n",
      "Seq #697 encoded correctly\n",
      "Seq #698 encoded correctly\n",
      "Seq #699 encoded correctly\n",
      "Seq #700 encoded correctly\n",
      "Seq #701 encoded correctly\n",
      "Seq #702 encoded correctly\n",
      "Seq #703 encoded correctly\n",
      "Seq #704 encoded correctly\n",
      "Seq #705 encoded correctly\n",
      "Seq #706 encoded correctly\n",
      "Seq #707 encoded correctly\n",
      "Seq #708 encoded correctly\n",
      "Seq #709 encoded correctly\n",
      "Seq #710 encoded correctly\n",
      "Seq #711 encoded correctly\n",
      "Seq #712 encoded correctly\n",
      "Seq #713 encoded correctly\n",
      "Seq #714 encoded correctly\n",
      "Seq #715 encoded correctly\n",
      "Seq #716 encoded correctly\n",
      "Seq #717 encoded correctly\n",
      "Seq #718 encoded correctly\n",
      "Seq #719 encoded correctly\n",
      "Seq #720 encoded correctly\n",
      "Seq #721 encoded correctly\n",
      "Seq #722 encoded correctly\n",
      "Seq #723 encoded correctly\n",
      "Seq #724 encoded correctly\n",
      "Seq #725 encoded correctly\n",
      "Seq #726 encoded correctly\n",
      "Seq #727 encoded correctly\n",
      "Seq #728 encoded correctly\n",
      "Seq #729 encoded correctly\n",
      "Seq #730 encoded correctly\n",
      "Seq #731 encoded correctly\n",
      "Seq #732 encoded correctly\n",
      "Seq #733 encoded correctly\n",
      "Seq #734 encoded correctly\n",
      "Seq #735 encoded correctly\n",
      "Seq #736 encoded correctly\n",
      "Seq #737 encoded correctly\n",
      "Seq #738 encoded correctly\n",
      "Seq #739 encoded correctly\n",
      "Seq #740 encoded correctly\n",
      "Seq #741 encoded correctly\n",
      "Seq #742 encoded correctly\n",
      "Seq #743 encoded correctly\n",
      "Seq #744 encoded correctly\n",
      "Seq #745 encoded correctly\n",
      "Seq #746 encoded correctly\n",
      "Seq #747 encoded correctly\n",
      "Seq #748 encoded correctly\n",
      "Seq #749 encoded correctly\n",
      "Seq #750 encoded correctly\n",
      "Seq #751 encoded correctly\n",
      "Seq #752 encoded correctly\n",
      "Seq #753 encoded correctly\n",
      "Seq #754 encoded correctly\n",
      "Seq #755 encoded correctly\n",
      "Seq #756 encoded correctly\n",
      "Seq #757 encoded correctly\n",
      "Seq #758 encoded correctly\n",
      "Seq #759 encoded correctly\n",
      "Seq #760 encoded correctly\n",
      "Seq #761 encoded correctly\n",
      "Seq #762 encoded correctly\n",
      "Seq #763 encoded correctly\n",
      "Seq #764 encoded correctly\n",
      "Seq #765 encoded correctly\n",
      "Seq #766 encoded correctly\n",
      "Seq #767 encoded correctly\n",
      "Seq #768 encoded correctly\n",
      "Seq #769 encoded correctly\n",
      "Seq #770 encoded correctly\n",
      "Seq #771 encoded correctly\n",
      "Seq #772 encoded correctly\n",
      "Seq #773 encoded correctly\n",
      "Seq #774 encoded correctly\n",
      "Seq #775 encoded correctly\n",
      "Seq #776 encoded correctly\n",
      "Seq #777 encoded correctly\n",
      "Seq #778 encoded correctly\n",
      "Seq #779 encoded correctly\n",
      "Seq #780 encoded correctly\n",
      "Seq #781 encoded correctly\n",
      "Seq #782 encoded correctly\n",
      "Seq #783 encoded correctly\n",
      "Seq #784 encoded correctly\n",
      "Seq #785 encoded correctly\n",
      "Seq #786 encoded correctly\n",
      "Seq #787 encoded correctly\n",
      "Seq #788 encoded correctly\n",
      "Seq #789 encoded correctly\n",
      "Seq #790 encoded correctly\n",
      "Seq #791 encoded correctly\n",
      "Seq #792 encoded correctly\n",
      "Seq #793 encoded correctly\n",
      "Seq #794 encoded correctly\n",
      "Seq #795 encoded correctly\n",
      "Seq #796 encoded correctly\n",
      "Seq #797 encoded correctly\n",
      "Seq #798 encoded correctly\n",
      "Seq #799 encoded correctly\n",
      "Seq #800 encoded correctly\n",
      "Seq #801 encoded correctly\n",
      "Seq #802 encoded correctly\n",
      "Seq #803 encoded correctly\n",
      "Seq #804 encoded correctly\n",
      "Seq #805 encoded correctly\n",
      "Seq #806 encoded correctly\n",
      "Seq #807 encoded correctly\n",
      "Seq #808 encoded correctly\n",
      "Seq #809 encoded correctly\n",
      "Seq #810 encoded correctly\n",
      "Seq #811 encoded correctly\n",
      "Seq #812 encoded correctly\n",
      "Seq #813 encoded correctly\n",
      "Seq #814 encoded correctly\n",
      "Seq #815 encoded correctly\n",
      "Seq #816 encoded correctly\n",
      "Seq #817 encoded correctly\n",
      "Seq #818 encoded correctly\n",
      "Seq #819 encoded correctly\n",
      "Seq #820 encoded correctly\n",
      "Seq #821 encoded correctly\n",
      "Seq #822 encoded correctly\n",
      "Seq #823 encoded correctly\n",
      "Seq #824 encoded correctly\n",
      "Seq #825 encoded correctly\n",
      "Seq #826 encoded correctly\n",
      "Seq #827 encoded correctly\n",
      "Seq #828 encoded correctly\n",
      "Seq #829 encoded correctly\n",
      "Seq #830 encoded correctly\n",
      "Seq #831 encoded correctly\n",
      "Seq #832 encoded correctly\n",
      "Seq #833 encoded correctly\n",
      "Seq #834 encoded correctly\n",
      "Seq #835 encoded correctly\n",
      "Seq #836 encoded correctly\n",
      "Seq #837 encoded correctly\n",
      "Seq #838 encoded correctly\n",
      "Seq #839 encoded correctly\n",
      "Seq #840 encoded correctly\n",
      "Seq #841 encoded correctly\n",
      "Seq #842 encoded correctly\n",
      "Seq #843 encoded correctly\n",
      "Seq #844 encoded correctly\n",
      "Seq #845 encoded correctly\n",
      "Seq #846 encoded correctly\n",
      "Seq #847 encoded correctly\n",
      "Seq #848 encoded correctly\n",
      "Seq #849 encoded correctly\n",
      "Seq #850 encoded correctly\n",
      "Seq #851 encoded correctly\n",
      "Seq #852 encoded correctly\n",
      "Seq #853 encoded correctly\n",
      "Seq #854 encoded correctly\n",
      "Seq #855 encoded correctly\n",
      "Seq #856 encoded correctly\n",
      "Seq #857 encoded correctly\n",
      "Seq #858 encoded correctly\n",
      "Seq #859 encoded correctly\n",
      "Seq #860 encoded correctly\n",
      "Seq #861 encoded correctly\n",
      "Seq #862 encoded correctly\n",
      "Seq #863 encoded correctly\n",
      "Seq #864 encoded correctly\n",
      "Seq #865 encoded correctly\n",
      "Seq #866 encoded correctly\n",
      "Seq #867 encoded correctly\n",
      "Seq #868 encoded correctly\n",
      "Seq #869 encoded correctly\n",
      "Seq #870 encoded correctly\n",
      "Seq #871 encoded correctly\n",
      "Seq #872 encoded correctly\n",
      "Seq #873 encoded correctly\n",
      "Seq #874 encoded correctly\n",
      "Seq #875 encoded correctly\n",
      "Seq #876 encoded correctly\n",
      "Seq #877 encoded correctly\n",
      "Seq #878 encoded correctly\n",
      "Seq #879 encoded correctly\n",
      "Seq #880 encoded correctly\n",
      "Seq #881 encoded correctly\n",
      "Seq #882 encoded correctly\n",
      "Seq #883 encoded correctly\n",
      "Seq #884 encoded correctly\n",
      "Seq #885 encoded correctly\n",
      "Seq #886 encoded correctly\n",
      "Seq #887 encoded correctly\n",
      "Seq #888 encoded correctly\n",
      "Seq #889 encoded correctly\n",
      "Seq #890 encoded correctly\n",
      "Seq #891 encoded correctly\n",
      "Seq #892 encoded correctly\n",
      "Seq #893 encoded correctly\n",
      "Seq #894 encoded correctly\n",
      "Seq #895 encoded correctly\n",
      "Seq #896 encoded correctly\n",
      "Seq #897 encoded correctly\n",
      "Seq #898 encoded correctly\n",
      "Seq #899 encoded correctly\n",
      "Seq #900 encoded correctly\n",
      "Seq #901 encoded correctly\n",
      "Seq #902 encoded correctly\n",
      "Seq #903 encoded correctly\n",
      "Seq #904 encoded correctly\n",
      "Seq #905 encoded correctly\n",
      "Seq #906 encoded correctly\n",
      "Seq #907 encoded correctly\n",
      "Seq #908 encoded correctly\n",
      "Seq #909 encoded correctly\n",
      "Seq #910 encoded correctly\n",
      "Seq #911 encoded correctly\n",
      "Seq #912 encoded correctly\n",
      "Seq #913 encoded correctly\n",
      "Seq #914 encoded correctly\n",
      "Seq #915 encoded correctly\n",
      "Seq #916 encoded correctly\n",
      "Seq #917 encoded correctly\n",
      "Seq #918 encoded correctly\n",
      "Seq #919 encoded correctly\n",
      "Seq #920 encoded correctly\n",
      "Seq #921 encoded correctly\n",
      "Seq #922 encoded correctly\n",
      "Seq #923 encoded correctly\n",
      "Seq #924 encoded correctly\n",
      "Seq #925 encoded correctly\n",
      "Seq #926 encoded correctly\n",
      "Seq #927 encoded correctly\n",
      "Seq #928 encoded correctly\n",
      "Seq #929 encoded correctly\n",
      "Seq #930 encoded correctly\n",
      "Seq #931 encoded correctly\n",
      "Seq #932 encoded correctly\n",
      "Seq #933 encoded correctly\n",
      "Seq #934 encoded correctly\n",
      "Seq #935 encoded correctly\n",
      "Seq #936 encoded correctly\n",
      "Seq #937 encoded correctly\n",
      "Seq #938 encoded correctly\n",
      "Seq #939 encoded correctly\n",
      "Seq #940 encoded correctly\n",
      "Seq #941 encoded correctly\n",
      "Seq #942 encoded correctly\n",
      "Seq #943 encoded correctly\n",
      "Seq #944 encoded correctly\n",
      "Seq #945 encoded correctly\n",
      "Seq #946 encoded correctly\n",
      "Seq #947 encoded correctly\n",
      "Seq #948 encoded correctly\n",
      "Seq #949 encoded correctly\n",
      "Seq #950 encoded correctly\n",
      "Seq #951 encoded correctly\n",
      "Seq #952 encoded correctly\n",
      "Seq #953 encoded correctly\n",
      "Seq #954 encoded correctly\n",
      "Seq #955 encoded correctly\n",
      "Seq #956 encoded correctly\n",
      "Seq #957 encoded correctly\n",
      "Seq #958 encoded correctly\n",
      "Seq #959 encoded correctly\n",
      "Seq #960 encoded correctly\n",
      "Seq #961 encoded correctly\n",
      "Seq #962 encoded correctly\n",
      "Seq #963 encoded correctly\n",
      "Seq #964 encoded correctly\n",
      "Seq #965 encoded correctly\n",
      "Seq #966 encoded correctly\n",
      "Seq #967 encoded correctly\n",
      "Seq #968 encoded correctly\n",
      "Seq #969 encoded correctly\n",
      "Seq #970 encoded correctly\n",
      "Seq #971 encoded correctly\n",
      "Seq #972 encoded correctly\n",
      "Seq #973 encoded correctly\n",
      "Seq #974 encoded correctly\n",
      "Seq #975 encoded correctly\n",
      "Seq #976 encoded correctly\n",
      "Seq #977 encoded correctly\n",
      "Seq #978 encoded correctly\n",
      "Seq #979 encoded correctly\n",
      "Seq #980 encoded correctly\n",
      "Seq #981 encoded correctly\n",
      "Seq #982 encoded correctly\n",
      "Seq #983 encoded correctly\n",
      "Seq #984 encoded correctly\n",
      "Seq #985 encoded correctly\n",
      "Seq #986 encoded correctly\n",
      "Seq #987 encoded correctly\n",
      "Seq #988 encoded correctly\n",
      "Seq #989 encoded correctly\n",
      "Seq #990 encoded correctly\n",
      "Seq #991 encoded correctly\n",
      "Seq #992 encoded correctly\n",
      "Seq #993 encoded correctly\n",
      "Seq #994 encoded correctly\n",
      "Seq #995 encoded correctly\n",
      "Seq #996 encoded correctly\n",
      "Seq #997 encoded correctly\n",
      "Seq #998 encoded correctly\n",
      "Seq #999 encoded correctly\n",
      "Seq #1000 encoded correctly\n",
      "Seq #1001 encoded correctly\n",
      "Seq #1002 encoded correctly\n",
      "Seq #1003 encoded correctly\n",
      "Seq #1004 encoded correctly\n",
      "Seq #1005 encoded correctly\n",
      "Seq #1006 encoded correctly\n",
      "Seq #1007 encoded correctly\n",
      "Seq #1008 encoded correctly\n",
      "Seq #1009 encoded correctly\n",
      "Seq #1010 encoded correctly\n",
      "Seq #1011 encoded correctly\n",
      "Seq #1012 encoded correctly\n",
      "Seq #1013 encoded correctly\n",
      "Seq #1014 encoded correctly\n",
      "Seq #1015 encoded correctly\n",
      "Seq #1016 encoded correctly\n",
      "Seq #1017 encoded correctly\n",
      "Seq #1018 encoded correctly\n",
      "Seq #1019 encoded correctly\n",
      "Seq #1020 encoded correctly\n",
      "Seq #1021 encoded correctly\n",
      "Seq #1022 encoded correctly\n",
      "Seq #1023 encoded correctly\n",
      "Seq #1024 encoded correctly\n",
      "Seq #1025 encoded correctly\n",
      "Seq #1026 encoded correctly\n",
      "Seq #1027 encoded correctly\n",
      "Seq #1028 encoded correctly\n",
      "Seq #1029 encoded correctly\n",
      "Seq #1030 encoded correctly\n",
      "Seq #1031 encoded correctly\n",
      "Seq #1032 encoded correctly\n",
      "Seq #1033 encoded correctly\n",
      "Seq #1034 encoded correctly\n",
      "Seq #1035 encoded correctly\n",
      "Seq #1036 encoded correctly\n",
      "Seq #1037 encoded correctly\n",
      "Seq #1038 encoded correctly\n",
      "Seq #1039 encoded correctly\n",
      "Seq #1040 encoded correctly\n",
      "Seq #1041 encoded correctly\n",
      "Seq #1042 encoded correctly\n",
      "Seq #1043 encoded correctly\n",
      "Seq #1044 encoded correctly\n",
      "Seq #1045 encoded correctly\n",
      "Seq #1046 encoded correctly\n",
      "Seq #1047 encoded correctly\n",
      "Seq #1048 encoded correctly\n",
      "Seq #1049 encoded correctly\n",
      "Seq #1050 encoded correctly\n",
      "Seq #1051 encoded correctly\n",
      "Seq #1052 encoded correctly\n",
      "Seq #1053 encoded correctly\n",
      "Seq #1054 encoded correctly\n",
      "Seq #1055 encoded correctly\n",
      "Seq #1056 encoded correctly\n",
      "Seq #1057 encoded correctly\n",
      "Seq #1058 encoded correctly\n",
      "Seq #1059 encoded correctly\n",
      "Seq #1060 encoded correctly\n",
      "Seq #1061 encoded correctly\n",
      "Seq #1062 encoded correctly\n",
      "Seq #1063 encoded correctly\n",
      "Seq #1064 encoded correctly\n",
      "Seq #1065 encoded correctly\n",
      "Seq #1066 encoded correctly\n",
      "Seq #1067 encoded correctly\n",
      "Seq #1068 encoded correctly\n",
      "Seq #1069 encoded correctly\n",
      "Seq #1070 encoded correctly\n",
      "Seq #1071 encoded correctly\n",
      "Seq #1072 encoded correctly\n",
      "Seq #1073 encoded correctly\n",
      "Seq #1074 encoded correctly\n",
      "Seq #1075 encoded correctly\n",
      "Seq #1076 encoded correctly\n",
      "Seq #1077 encoded correctly\n",
      "Seq #1078 encoded correctly\n",
      "Seq #1079 encoded correctly\n",
      "Seq #1080 encoded correctly\n",
      "Seq #1081 encoded correctly\n",
      "Seq #1082 encoded correctly\n",
      "Seq #1083 encoded correctly\n",
      "Seq #1084 encoded correctly\n",
      "Seq #1085 encoded correctly\n",
      "Seq #1086 encoded correctly\n",
      "Seq #1087 encoded correctly\n",
      "Seq #1088 encoded correctly\n",
      "Seq #1089 encoded correctly\n",
      "Seq #1090 encoded correctly\n",
      "Seq #1091 encoded correctly\n",
      "Seq #1092 encoded correctly\n",
      "Seq #1093 encoded correctly\n",
      "Seq #1094 encoded correctly\n",
      "Seq #1095 encoded correctly\n",
      "Seq #1096 encoded correctly\n",
      "Seq #1097 encoded correctly\n",
      "Seq #1098 encoded correctly\n",
      "Seq #1099 encoded correctly\n",
      "Seq #1100 encoded correctly\n",
      "Seq #1101 encoded correctly\n",
      "Seq #1102 encoded correctly\n",
      "Seq #1103 encoded correctly\n",
      "Seq #1104 encoded correctly\n",
      "Seq #1105 encoded correctly\n",
      "Seq #1106 encoded correctly\n",
      "Seq #1107 encoded correctly\n",
      "Seq #1108 encoded correctly\n",
      "Seq #1109 encoded correctly\n",
      "Seq #1110 encoded correctly\n",
      "Seq #1111 encoded correctly\n",
      "Seq #1112 encoded correctly\n",
      "Seq #1113 encoded correctly\n",
      "Seq #1114 encoded correctly\n",
      "Seq #1115 encoded correctly\n",
      "Seq #1116 encoded correctly\n",
      "Seq #1117 encoded correctly\n",
      "Seq #1118 encoded correctly\n",
      "Seq #1119 encoded correctly\n",
      "Seq #1120 encoded correctly\n",
      "Seq #1121 encoded correctly\n",
      "Seq #1122 encoded correctly\n",
      "Seq #1123 encoded correctly\n",
      "Seq #1124 encoded correctly\n",
      "Seq #1125 encoded correctly\n",
      "Seq #1126 encoded correctly\n",
      "Seq #1127 encoded correctly\n",
      "Seq #1128 encoded correctly\n",
      "Seq #1129 encoded correctly\n",
      "Seq #1130 encoded correctly\n",
      "Seq #1131 encoded correctly\n",
      "Seq #1132 encoded correctly\n",
      "Seq #1133 encoded correctly\n",
      "Seq #1134 encoded correctly\n",
      "Seq #1135 encoded correctly\n",
      "Seq #1136 encoded correctly\n",
      "Seq #1137 encoded correctly\n",
      "Seq #1138 encoded correctly\n",
      "Seq #1139 encoded correctly\n",
      "Seq #1140 encoded correctly\n",
      "Seq #1141 encoded correctly\n",
      "Seq #1142 encoded correctly\n",
      "Seq #1143 encoded correctly\n",
      "Seq #1144 encoded correctly\n",
      "Seq #1145 encoded correctly\n",
      "Seq #1146 encoded correctly\n",
      "Seq #1147 encoded correctly\n",
      "Seq #1148 encoded correctly\n",
      "Seq #1149 encoded correctly\n",
      "Seq #1150 encoded correctly\n",
      "Seq #1151 encoded correctly\n",
      "Seq #1152 encoded correctly\n",
      "Seq #1153 encoded correctly\n",
      "Seq #1154 encoded correctly\n",
      "Seq #1155 encoded correctly\n",
      "Seq #1156 encoded correctly\n",
      "Seq #1157 encoded correctly\n",
      "Seq #1158 encoded correctly\n",
      "Seq #1159 encoded correctly\n",
      "Seq #1160 encoded correctly\n",
      "Seq #1161 encoded correctly\n",
      "Seq #1162 encoded correctly\n",
      "Seq #1163 encoded correctly\n",
      "Seq #1164 encoded correctly\n",
      "Seq #1165 encoded correctly\n",
      "Seq #1166 encoded correctly\n",
      "Seq #1167 encoded correctly\n",
      "Seq #1168 encoded correctly\n",
      "Seq #1169 encoded correctly\n",
      "Seq #1170 encoded correctly\n",
      "Seq #1171 encoded correctly\n",
      "Seq #1172 encoded correctly\n",
      "Seq #1173 encoded correctly\n",
      "Seq #1174 encoded correctly\n",
      "Seq #1175 encoded correctly\n",
      "Seq #1176 encoded correctly\n",
      "Seq #1177 encoded correctly\n",
      "Seq #1178 encoded correctly\n",
      "Seq #1179 encoded correctly\n",
      "Seq #1180 encoded correctly\n",
      "Seq #1181 encoded correctly\n",
      "Seq #1182 encoded correctly\n",
      "Seq #1183 encoded correctly\n",
      "Seq #1184 encoded correctly\n",
      "Seq #1185 encoded correctly\n",
      "Seq #1186 encoded correctly\n",
      "Seq #1187 encoded correctly\n",
      "Seq #1188 encoded correctly\n",
      "Seq #1189 encoded correctly\n",
      "Seq #1190 encoded correctly\n",
      "Seq #1191 encoded correctly\n",
      "Seq #1192 encoded correctly\n",
      "Seq #1193 encoded correctly\n",
      "Seq #1194 encoded correctly\n",
      "Seq #1195 encoded correctly\n",
      "Seq #1196 encoded correctly\n",
      "Seq #1197 encoded correctly\n",
      "Seq #1198 encoded correctly\n",
      "Seq #1199 encoded correctly\n",
      "Seq #1200 encoded correctly\n",
      "Seq #1201 encoded correctly\n",
      "Seq #1202 encoded correctly\n",
      "Seq #1203 encoded correctly\n",
      "Seq #1204 encoded correctly\n",
      "Seq #1205 encoded correctly\n",
      "Seq #1206 encoded correctly\n",
      "Seq #1207 encoded correctly\n",
      "Seq #1208 encoded correctly\n",
      "Seq #1209 encoded correctly\n",
      "Seq #1210 encoded correctly\n",
      "Seq #1211 encoded correctly\n",
      "Seq #1212 encoded correctly\n",
      "Seq #1213 encoded correctly\n",
      "Seq #1214 encoded correctly\n",
      "Seq #1215 encoded correctly\n",
      "Seq #1216 encoded correctly\n",
      "Seq #1217 encoded correctly\n",
      "Seq #1218 encoded correctly\n",
      "Seq #1219 encoded correctly\n",
      "Seq #1220 encoded correctly\n",
      "Seq #1221 encoded correctly\n",
      "Seq #1222 encoded correctly\n",
      "Seq #1223 encoded correctly\n",
      "Seq #1224 encoded correctly\n",
      "Seq #1225 encoded correctly\n",
      "Seq #1226 encoded correctly\n",
      "Seq #1227 encoded correctly\n",
      "Seq #1228 encoded correctly\n",
      "Seq #1229 encoded correctly\n",
      "Seq #1230 encoded correctly\n",
      "Seq #1231 encoded correctly\n",
      "Seq #1232 encoded correctly\n",
      "Seq #1233 encoded correctly\n",
      "Seq #1234 encoded correctly\n",
      "Seq #1235 encoded correctly\n",
      "Seq #1236 encoded correctly\n",
      "Seq #1237 encoded correctly\n",
      "Seq #1238 encoded correctly\n",
      "Seq #1239 encoded correctly\n",
      "Seq #1240 encoded correctly\n",
      "Seq #1241 encoded correctly\n",
      "Seq #1242 encoded correctly\n",
      "Seq #1243 encoded correctly\n",
      "Seq #1244 encoded correctly\n",
      "Seq #1245 encoded correctly\n",
      "Seq #1246 encoded correctly\n",
      "Seq #1247 encoded correctly\n",
      "Seq #1248 encoded correctly\n",
      "Seq #1249 encoded correctly\n",
      "Seq #1250 encoded correctly\n",
      "Seq #1251 encoded correctly\n",
      "Seq #1252 encoded correctly\n",
      "Seq #1253 encoded correctly\n",
      "Seq #1254 encoded correctly\n",
      "Seq #1255 encoded correctly\n",
      "Seq #1256 encoded correctly\n",
      "Seq #1257 encoded correctly\n",
      "Seq #1258 encoded correctly\n",
      "Seq #1259 encoded correctly\n",
      "Seq #1260 encoded correctly\n",
      "Seq #1261 encoded correctly\n",
      "Seq #1262 encoded correctly\n",
      "Seq #1263 encoded correctly\n",
      "Seq #1264 encoded correctly\n",
      "Seq #1265 encoded correctly\n",
      "Seq #1266 encoded correctly\n",
      "Seq #1267 encoded correctly\n",
      "Seq #1268 encoded correctly\n",
      "Seq #1269 encoded correctly\n",
      "Seq #1270 encoded correctly\n",
      "Seq #1271 encoded correctly\n",
      "Seq #1272 encoded correctly\n",
      "Seq #1273 encoded correctly\n",
      "Seq #1274 encoded correctly\n",
      "Seq #1275 encoded correctly\n",
      "Seq #1276 encoded correctly\n",
      "Seq #1277 encoded correctly\n",
      "Seq #1278 encoded correctly\n",
      "Seq #1279 encoded correctly\n",
      "Seq #1280 encoded correctly\n",
      "Seq #1281 encoded correctly\n",
      "Seq #1282 encoded correctly\n",
      "Seq #1283 encoded correctly\n",
      "Seq #1284 encoded correctly\n",
      "Seq #1285 encoded correctly\n",
      "Seq #1286 encoded correctly\n",
      "Seq #1287 encoded correctly\n",
      "Seq #1288 encoded correctly\n",
      "Seq #1289 encoded correctly\n",
      "Seq #1290 encoded correctly\n",
      "Seq #1291 encoded correctly\n",
      "Seq #1292 encoded correctly\n",
      "Seq #1293 encoded correctly\n",
      "Seq #1294 encoded correctly\n",
      "Seq #1295 encoded correctly\n",
      "Seq #1296 encoded correctly\n",
      "Seq #1297 encoded correctly\n",
      "Seq #1298 encoded correctly\n",
      "Seq #1299 encoded correctly\n",
      "Seq #1300 encoded correctly\n",
      "Seq #1301 encoded correctly\n",
      "Seq #1302 encoded correctly\n",
      "Seq #1303 encoded correctly\n",
      "Seq #1304 encoded correctly\n",
      "Seq #1305 encoded correctly\n",
      "Seq #1306 encoded correctly\n",
      "Seq #1307 encoded correctly\n",
      "Seq #1308 encoded correctly\n",
      "Seq #1309 encoded correctly\n",
      "Seq #1310 encoded correctly\n",
      "Seq #1311 encoded correctly\n",
      "Seq #1312 encoded correctly\n",
      "Seq #1313 encoded correctly\n",
      "Seq #1314 encoded correctly\n",
      "Seq #1315 encoded correctly\n",
      "Seq #1316 encoded correctly\n",
      "Seq #1317 encoded correctly\n",
      "Seq #1318 encoded correctly\n",
      "Seq #1319 encoded correctly\n",
      "Seq #1320 encoded correctly\n",
      "Seq #1321 encoded correctly\n",
      "Seq #1322 encoded correctly\n",
      "Seq #1323 encoded correctly\n",
      "Seq #1324 encoded correctly\n",
      "Seq #1325 encoded correctly\n",
      "Seq #1326 encoded correctly\n",
      "Seq #1327 encoded correctly\n",
      "Seq #1328 encoded correctly\n",
      "Seq #1329 encoded correctly\n",
      "Seq #1330 encoded correctly\n",
      "Seq #1331 encoded correctly\n",
      "Seq #1332 encoded correctly\n",
      "Seq #1333 encoded correctly\n",
      "Seq #1334 encoded correctly\n",
      "Seq #1335 encoded correctly\n",
      "Seq #1336 encoded correctly\n",
      "Seq #1337 encoded correctly\n",
      "Seq #1338 encoded correctly\n",
      "Seq #1339 encoded correctly\n",
      "Seq #1340 encoded correctly\n",
      "Seq #1341 encoded correctly\n",
      "Seq #1342 encoded correctly\n",
      "Seq #1343 encoded correctly\n",
      "Seq #1344 encoded correctly\n",
      "Seq #1345 encoded correctly\n",
      "Seq #1346 encoded correctly\n",
      "Seq #1347 encoded correctly\n",
      "Seq #1348 encoded correctly\n",
      "Seq #1349 encoded correctly\n",
      "Seq #1350 encoded correctly\n",
      "Seq #1351 encoded correctly\n",
      "Seq #1352 encoded correctly\n",
      "Seq #1353 encoded correctly\n",
      "Seq #1354 encoded correctly\n",
      "Seq #1355 encoded correctly\n",
      "Seq #1356 encoded correctly\n",
      "Seq #1357 encoded correctly\n",
      "Seq #1358 encoded correctly\n",
      "Seq #1359 encoded correctly\n",
      "Seq #1360 encoded correctly\n",
      "Seq #1361 encoded correctly\n",
      "Seq #1362 encoded correctly\n",
      "Seq #1363 encoded correctly\n",
      "Seq #1364 encoded correctly\n",
      "Seq #1365 encoded correctly\n",
      "Seq #1366 encoded correctly\n",
      "Seq #1367 encoded correctly\n",
      "Seq #1368 encoded correctly\n",
      "Seq #1369 encoded correctly\n",
      "Seq #1370 encoded correctly\n",
      "Seq #1371 encoded correctly\n",
      "Seq #1372 encoded correctly\n",
      "Seq #1373 encoded correctly\n",
      "Seq #1374 encoded correctly\n",
      "Seq #1375 encoded correctly\n",
      "Seq #1376 encoded correctly\n",
      "Seq #1377 encoded correctly\n",
      "Seq #1378 encoded correctly\n",
      "Seq #1379 encoded correctly\n",
      "Seq #1380 encoded correctly\n",
      "Seq #1381 encoded correctly\n",
      "Seq #1382 encoded correctly\n",
      "Seq #1383 encoded correctly\n",
      "Seq #1384 encoded correctly\n",
      "Seq #1385 encoded correctly\n",
      "Seq #1386 encoded correctly\n",
      "Seq #1387 encoded correctly\n",
      "Seq #1388 encoded correctly\n",
      "Seq #1389 encoded correctly\n",
      "Seq #1390 encoded correctly\n",
      "Seq #1391 encoded correctly\n",
      "Seq #1392 encoded correctly\n",
      "Seq #1393 encoded correctly\n",
      "Seq #1394 encoded correctly\n",
      "Seq #1395 encoded correctly\n",
      "Seq #1396 encoded correctly\n",
      "Seq #1397 encoded correctly\n",
      "Seq #1398 encoded correctly\n",
      "Seq #1399 encoded correctly\n",
      "Seq #1400 encoded correctly\n",
      "Seq #1401 encoded correctly\n",
      "Seq #1402 encoded correctly\n",
      "Seq #1403 encoded correctly\n",
      "Seq #1404 encoded correctly\n",
      "Seq #1405 encoded correctly\n",
      "Seq #1406 encoded correctly\n",
      "Seq #1407 encoded correctly\n",
      "Seq #1408 encoded correctly\n",
      "Seq #1409 encoded correctly\n",
      "Seq #1410 encoded correctly\n",
      "Seq #1411 encoded correctly\n",
      "Seq #1412 encoded correctly\n",
      "Seq #1413 encoded correctly\n",
      "Seq #1414 encoded correctly\n",
      "Seq #1415 encoded correctly\n",
      "Seq #1416 encoded correctly\n",
      "Seq #1417 encoded correctly\n",
      "Seq #1418 encoded correctly\n",
      "Seq #1419 encoded correctly\n",
      "Seq #1420 encoded correctly\n",
      "Seq #1421 encoded correctly\n",
      "Seq #1422 encoded correctly\n",
      "Seq #1423 encoded correctly\n",
      "Seq #1424 encoded correctly\n",
      "Seq #1425 encoded correctly\n",
      "Seq #1426 encoded correctly\n",
      "Seq #1427 encoded correctly\n",
      "Seq #1428 encoded correctly\n",
      "Seq #1429 encoded correctly\n",
      "Seq #1430 encoded correctly\n",
      "Seq #1431 encoded correctly\n",
      "Seq #1432 encoded correctly\n",
      "Seq #1433 encoded correctly\n",
      "Seq #1434 encoded correctly\n",
      "Seq #1435 encoded correctly\n",
      "Seq #1436 encoded correctly\n",
      "Seq #1437 encoded correctly\n",
      "Seq #1438 encoded correctly\n",
      "Seq #1439 encoded correctly\n",
      "Seq #1440 encoded correctly\n",
      "Seq #1441 encoded correctly\n",
      "Seq #1442 encoded correctly\n",
      "Seq #1443 encoded correctly\n",
      "Seq #1444 encoded correctly\n",
      "Seq #1445 encoded correctly\n",
      "Seq #1446 encoded correctly\n",
      "Seq #1447 encoded correctly\n",
      "Seq #1448 encoded correctly\n",
      "Seq #1449 encoded correctly\n",
      "Seq #1450 encoded correctly\n",
      "Seq #1451 encoded correctly\n",
      "Seq #1452 encoded correctly\n",
      "Seq #1453 encoded correctly\n",
      "Seq #1454 encoded correctly\n",
      "Seq #1455 encoded correctly\n",
      "Seq #1456 encoded correctly\n",
      "Seq #1457 encoded correctly\n",
      "Seq #1458 encoded correctly\n",
      "Seq #1459 encoded correctly\n",
      "Seq #1460 encoded correctly\n",
      "Seq #1461 encoded correctly\n",
      "Seq #1462 encoded correctly\n",
      "Seq #1463 encoded correctly\n",
      "Seq #1464 encoded correctly\n",
      "Seq #1465 encoded correctly\n",
      "Seq #1466 encoded correctly\n",
      "Seq #1467 encoded correctly\n",
      "Seq #1468 encoded correctly\n",
      "Seq #1469 encoded correctly\n",
      "Seq #1470 encoded correctly\n",
      "Seq #1471 encoded correctly\n",
      "Seq #1472 encoded correctly\n",
      "Seq #1473 encoded correctly\n",
      "Seq #1474 encoded correctly\n",
      "Seq #1475 encoded correctly\n",
      "Seq #1476 encoded correctly\n",
      "Seq #1477 encoded correctly\n",
      "Seq #1478 encoded correctly\n",
      "Seq #1479 encoded correctly\n",
      "Seq #1480 encoded correctly\n",
      "Seq #1481 encoded correctly\n",
      "Seq #1482 encoded correctly\n",
      "Seq #1483 encoded correctly\n",
      "Seq #1484 encoded correctly\n",
      "Seq #1485 encoded correctly\n",
      "Seq #1486 encoded correctly\n",
      "Seq #1487 encoded correctly\n",
      "Seq #1488 encoded correctly\n",
      "Seq #1489 encoded correctly\n",
      "Seq #1490 encoded correctly\n",
      "Seq #1491 encoded correctly\n",
      "Seq #1492 encoded correctly\n",
      "Seq #1493 encoded correctly\n",
      "Seq #1494 encoded correctly\n",
      "Seq #1495 encoded correctly\n",
      "Seq #1496 encoded correctly\n",
      "Seq #1497 encoded correctly\n",
      "Seq #1498 encoded correctly\n",
      "Seq #1499 encoded correctly\n",
      "Seq #1500 encoded correctly\n",
      "Seq #1501 encoded correctly\n",
      "Seq #1502 encoded correctly\n",
      "Seq #1503 encoded correctly\n",
      "Seq #1504 encoded correctly\n",
      "Seq #1505 encoded correctly\n",
      "Seq #1506 encoded correctly\n",
      "Seq #1507 encoded correctly\n",
      "Seq #1508 encoded correctly\n",
      "Seq #1509 encoded correctly\n",
      "Seq #1510 encoded correctly\n",
      "Seq #1511 encoded correctly\n",
      "Seq #1512 encoded correctly\n",
      "Seq #1513 encoded correctly\n",
      "Seq #1514 encoded correctly\n",
      "Seq #1515 encoded correctly\n",
      "Seq #1516 encoded correctly\n",
      "Seq #1517 encoded correctly\n",
      "Seq #1518 encoded correctly\n",
      "Seq #1519 encoded correctly\n",
      "Seq #1520 encoded correctly\n",
      "Seq #1521 encoded correctly\n",
      "Seq #1522 encoded correctly\n",
      "Seq #1523 encoded correctly\n",
      "Seq #1524 encoded correctly\n",
      "Seq #1525 encoded correctly\n",
      "Seq #1526 encoded correctly\n",
      "Seq #1527 encoded correctly\n",
      "Seq #1528 encoded correctly\n",
      "Seq #1529 encoded correctly\n",
      "Seq #1530 encoded correctly\n",
      "Seq #1531 encoded correctly\n",
      "Seq #1532 encoded correctly\n",
      "Seq #1533 encoded correctly\n",
      "Seq #1534 encoded correctly\n",
      "Seq #1535 encoded correctly\n",
      "Seq #1536 encoded correctly\n",
      "Seq #1537 encoded correctly\n",
      "Seq #1538 encoded correctly\n",
      "Seq #1539 encoded correctly\n",
      "Seq #1540 encoded correctly\n",
      "Seq #1541 encoded correctly\n",
      "Seq #1542 encoded correctly\n",
      "Seq #1543 encoded correctly\n",
      "Seq #1544 encoded correctly\n",
      "Seq #1545 encoded correctly\n",
      "Seq #1546 encoded correctly\n",
      "Seq #1547 encoded correctly\n",
      "Seq #1548 encoded correctly\n",
      "Seq #1549 encoded correctly\n",
      "Seq #1550 encoded correctly\n",
      "Seq #1551 encoded correctly\n",
      "Seq #1552 encoded correctly\n",
      "Seq #1553 encoded correctly\n",
      "Seq #1554 encoded correctly\n",
      "Seq #1555 encoded correctly\n",
      "Seq #1556 encoded correctly\n",
      "Seq #1557 encoded correctly\n",
      "Seq #1558 encoded correctly\n",
      "Seq #1559 encoded correctly\n",
      "Seq #1560 encoded correctly\n",
      "Seq #1561 encoded correctly\n",
      "Seq #1562 encoded correctly\n",
      "Seq #1563 encoded correctly\n",
      "Seq #1564 encoded correctly\n",
      "Seq #1565 encoded correctly\n",
      "Seq #1566 encoded correctly\n",
      "Seq #1567 encoded correctly\n",
      "Seq #1568 encoded correctly\n",
      "Seq #1569 encoded correctly\n",
      "Seq #1570 encoded correctly\n",
      "Seq #1571 encoded correctly\n",
      "Seq #1572 encoded correctly\n",
      "Seq #1573 encoded correctly\n",
      "Seq #1574 encoded correctly\n",
      "Seq #1575 encoded correctly\n",
      "Seq #1576 encoded correctly\n",
      "Seq #1577 encoded correctly\n",
      "Seq #1578 encoded correctly\n",
      "Seq #1579 encoded correctly\n",
      "Seq #1580 encoded correctly\n",
      "Seq #1581 encoded correctly\n",
      "Seq #1582 encoded correctly\n",
      "Seq #1583 encoded correctly\n",
      "Seq #1584 encoded correctly\n",
      "Seq #1585 encoded correctly\n",
      "Seq #1586 encoded correctly\n",
      "Seq #1587 encoded correctly\n",
      "Seq #1588 encoded correctly\n",
      "Seq #1589 encoded correctly\n",
      "Seq #1590 encoded correctly\n",
      "Seq #1591 encoded correctly\n",
      "Seq #1592 encoded correctly\n",
      "Seq #1593 encoded correctly\n",
      "Seq #1594 encoded correctly\n",
      "Seq #1595 encoded correctly\n",
      "Seq #1596 encoded correctly\n",
      "Seq #1597 encoded correctly\n",
      "Seq #1598 encoded correctly\n",
      "Seq #1599 encoded correctly\n",
      "Seq #1600 encoded correctly\n",
      "Seq #1601 encoded correctly\n",
      "Seq #1602 encoded correctly\n",
      "Seq #1603 encoded correctly\n",
      "Seq #1604 encoded correctly\n",
      "Seq #1605 encoded correctly\n",
      "Seq #1606 encoded correctly\n",
      "Seq #1607 encoded correctly\n",
      "Seq #1608 encoded correctly\n",
      "Seq #1609 encoded correctly\n",
      "Seq #1610 encoded correctly\n",
      "Seq #1611 encoded correctly\n",
      "Seq #1612 encoded correctly\n",
      "Seq #1613 encoded correctly\n",
      "Seq #1614 encoded correctly\n",
      "Seq #1615 encoded correctly\n",
      "Seq #1616 encoded correctly\n",
      "Seq #1617 encoded correctly\n",
      "Seq #1618 encoded correctly\n",
      "Seq #1619 encoded correctly\n",
      "Seq #1620 encoded correctly\n",
      "Seq #1621 encoded correctly\n",
      "Seq #1622 encoded correctly\n",
      "Seq #1623 encoded correctly\n",
      "Seq #1624 encoded correctly\n",
      "Seq #1625 encoded correctly\n",
      "Seq #1626 encoded correctly\n",
      "Seq #1627 encoded correctly\n",
      "Seq #1628 encoded correctly\n",
      "Seq #1629 encoded correctly\n",
      "Seq #1630 encoded correctly\n",
      "Seq #1631 encoded correctly\n",
      "Seq #1632 encoded correctly\n",
      "Seq #1633 encoded correctly\n",
      "Seq #1634 encoded correctly\n",
      "Seq #1635 encoded correctly\n",
      "Seq #1636 encoded correctly\n",
      "Seq #1637 encoded correctly\n",
      "Seq #1638 encoded correctly\n",
      "Seq #1639 encoded correctly\n",
      "Seq #1640 encoded correctly\n",
      "Seq #1641 encoded correctly\n",
      "Seq #1642 encoded correctly\n",
      "Seq #1643 encoded correctly\n",
      "Seq #1644 encoded correctly\n",
      "Seq #1645 encoded correctly\n",
      "Seq #1646 encoded correctly\n",
      "Seq #1647 encoded correctly\n",
      "Seq #1648 encoded correctly\n",
      "Seq #1649 encoded correctly\n",
      "Seq #1650 encoded correctly\n",
      "Seq #1651 encoded correctly\n",
      "Seq #1652 encoded correctly\n",
      "Seq #1653 encoded correctly\n",
      "Seq #1654 encoded correctly\n",
      "Seq #1655 encoded correctly\n",
      "Seq #1656 encoded correctly\n",
      "Seq #1657 encoded correctly\n",
      "Seq #1658 encoded correctly\n",
      "Seq #1659 encoded correctly\n",
      "Seq #1660 encoded correctly\n",
      "Seq #1661 encoded correctly\n",
      "Seq #1662 encoded correctly\n",
      "Seq #1663 encoded correctly\n",
      "Seq #1664 encoded correctly\n",
      "Seq #1665 encoded correctly\n",
      "Seq #1666 encoded correctly\n",
      "Seq #1667 encoded correctly\n",
      "Seq #1668 encoded correctly\n",
      "Seq #1669 encoded correctly\n",
      "Seq #1670 encoded correctly\n",
      "Seq #1671 encoded correctly\n",
      "Seq #1672 encoded correctly\n",
      "Seq #1673 encoded correctly\n",
      "Seq #1674 encoded correctly\n",
      "Seq #1675 encoded correctly\n",
      "Seq #1676 encoded correctly\n",
      "Seq #1677 encoded correctly\n",
      "Seq #1678 encoded correctly\n",
      "Seq #1679 encoded correctly\n",
      "Seq #1680 encoded correctly\n",
      "Seq #1681 encoded correctly\n",
      "Seq #1682 encoded correctly\n",
      "Seq #1683 encoded correctly\n",
      "Seq #1684 encoded correctly\n",
      "Seq #1685 encoded correctly\n",
      "Seq #1686 encoded correctly\n",
      "Seq #1687 encoded correctly\n",
      "Seq #1688 encoded correctly\n",
      "Seq #1689 encoded correctly\n",
      "Seq #1690 encoded correctly\n",
      "Seq #1691 encoded correctly\n",
      "Seq #1692 encoded correctly\n",
      "Seq #1693 encoded correctly\n",
      "Seq #1694 encoded correctly\n",
      "Seq #1695 encoded correctly\n",
      "Seq #1696 encoded correctly\n",
      "Seq #1697 encoded correctly\n",
      "Seq #1698 encoded correctly\n",
      "Seq #1699 encoded correctly\n",
      "Seq #1700 encoded correctly\n",
      "Seq #1701 encoded correctly\n",
      "Seq #1702 encoded correctly\n",
      "Seq #1703 encoded correctly\n",
      "Seq #1704 encoded correctly\n",
      "Seq #1705 encoded correctly\n",
      "Seq #1706 encoded correctly\n",
      "Seq #1707 encoded correctly\n",
      "Seq #1708 encoded correctly\n",
      "Seq #1709 encoded correctly\n",
      "Seq #1710 encoded correctly\n",
      "Seq #1711 encoded correctly\n",
      "Seq #1712 encoded correctly\n",
      "Seq #1713 encoded correctly\n",
      "Seq #1714 encoded correctly\n",
      "Seq #1715 encoded correctly\n",
      "Seq #1716 encoded correctly\n",
      "Seq #1717 encoded correctly\n",
      "Seq #1718 encoded correctly\n",
      "Seq #1719 encoded correctly\n",
      "Seq #1720 encoded correctly\n",
      "Seq #1721 encoded correctly\n",
      "Seq #1722 encoded correctly\n",
      "Seq #1723 encoded correctly\n",
      "Seq #1724 encoded correctly\n",
      "Seq #1725 encoded correctly\n",
      "Seq #1726 encoded correctly\n",
      "Seq #1727 encoded correctly\n",
      "Seq #1728 encoded correctly\n",
      "Seq #1729 encoded correctly\n",
      "Seq #1730 encoded correctly\n",
      "Seq #1731 encoded correctly\n",
      "Seq #1732 encoded correctly\n",
      "Seq #1733 encoded correctly\n",
      "Seq #1734 encoded correctly\n",
      "Seq #1735 encoded correctly\n",
      "Seq #1736 encoded correctly\n",
      "Seq #1737 encoded correctly\n",
      "Seq #1738 encoded correctly\n",
      "Seq #1739 encoded correctly\n",
      "Seq #1740 encoded correctly\n",
      "Seq #1741 encoded correctly\n",
      "Seq #1742 encoded correctly\n",
      "Seq #1743 encoded correctly\n",
      "Seq #1744 encoded correctly\n",
      "Seq #1745 encoded correctly\n",
      "Seq #1746 encoded correctly\n",
      "Seq #1747 encoded correctly\n",
      "Seq #1748 encoded correctly\n",
      "Seq #1749 encoded correctly\n",
      "Seq #1750 encoded correctly\n",
      "Seq #1751 encoded correctly\n",
      "Seq #1752 encoded correctly\n",
      "Seq #1753 encoded correctly\n",
      "Seq #1754 encoded correctly\n",
      "Seq #1755 encoded correctly\n",
      "Seq #1756 encoded correctly\n",
      "Seq #1757 encoded correctly\n",
      "Seq #1758 encoded correctly\n",
      "Seq #1759 encoded correctly\n",
      "Seq #1760 encoded correctly\n",
      "Seq #1761 encoded correctly\n",
      "Seq #1762 encoded correctly\n",
      "Seq #1763 encoded correctly\n",
      "Seq #1764 encoded correctly\n",
      "Seq #1765 encoded correctly\n",
      "Seq #1766 encoded correctly\n",
      "Seq #1767 encoded correctly\n",
      "Seq #1768 encoded correctly\n",
      "Seq #1769 encoded correctly\n",
      "Seq #1770 encoded correctly\n",
      "Seq #1771 encoded correctly\n",
      "Seq #1772 encoded correctly\n",
      "Seq #1773 encoded correctly\n",
      "Seq #1774 encoded correctly\n",
      "Seq #1775 encoded correctly\n",
      "Seq #1776 encoded correctly\n",
      "Seq #1777 encoded correctly\n",
      "Seq #1778 encoded correctly\n",
      "Seq #1779 encoded correctly\n",
      "Seq #1780 encoded correctly\n",
      "Seq #1781 encoded correctly\n",
      "Seq #1782 encoded correctly\n",
      "Seq #1783 encoded correctly\n",
      "Seq #1784 encoded correctly\n",
      "Seq #1785 encoded correctly\n",
      "Seq #1786 encoded correctly\n",
      "Seq #1787 encoded correctly\n",
      "Seq #1788 encoded correctly\n",
      "Seq #1789 encoded correctly\n",
      "Seq #1790 encoded correctly\n",
      "Seq #1791 encoded correctly\n",
      "Seq #1792 encoded correctly\n",
      "Seq #1793 encoded correctly\n",
      "Seq #1794 encoded correctly\n",
      "Seq #1795 encoded correctly\n",
      "Seq #1796 encoded correctly\n",
      "Seq #1797 encoded correctly\n",
      "Seq #1798 encoded correctly\n",
      "Seq #1799 encoded correctly\n",
      "Seq #1800 encoded correctly\n",
      "Seq #1801 encoded correctly\n",
      "Seq #1802 encoded correctly\n",
      "Seq #1803 encoded correctly\n",
      "Seq #1804 encoded correctly\n",
      "Seq #1805 encoded correctly\n",
      "Seq #1806 encoded correctly\n",
      "Seq #1807 encoded correctly\n",
      "Seq #1808 encoded correctly\n",
      "Seq #1809 encoded correctly\n",
      "Seq #1810 encoded correctly\n",
      "Seq #1811 encoded correctly\n",
      "Seq #1812 encoded correctly\n",
      "Seq #1813 encoded correctly\n",
      "Seq #1814 encoded correctly\n",
      "Seq #1815 encoded correctly\n",
      "Seq #1816 encoded correctly\n",
      "Seq #1817 encoded correctly\n",
      "Seq #1818 encoded correctly\n",
      "Seq #1819 encoded correctly\n",
      "Seq #1820 encoded correctly\n",
      "Seq #1821 encoded correctly\n",
      "Seq #1822 encoded correctly\n",
      "Seq #1823 encoded correctly\n",
      "Seq #1824 encoded correctly\n",
      "Seq #1825 encoded correctly\n",
      "Seq #1826 encoded correctly\n",
      "Seq #1827 encoded correctly\n",
      "Seq #1828 encoded correctly\n",
      "Seq #1829 encoded correctly\n",
      "Seq #1830 encoded correctly\n",
      "Seq #1831 encoded correctly\n",
      "Seq #1832 encoded correctly\n",
      "Seq #1833 encoded correctly\n",
      "Seq #1834 encoded correctly\n",
      "Seq #1835 encoded correctly\n",
      "Seq #1836 encoded correctly\n",
      "Seq #1837 encoded correctly\n",
      "Seq #1838 encoded correctly\n",
      "Seq #1839 encoded correctly\n",
      "Seq #1840 encoded correctly\n",
      "Seq #1841 encoded correctly\n",
      "Seq #1842 encoded correctly\n",
      "Seq #1843 encoded correctly\n",
      "Seq #1844 encoded correctly\n",
      "Seq #1845 encoded correctly\n",
      "Seq #1846 encoded correctly\n",
      "Seq #1847 encoded correctly\n",
      "Seq #1848 encoded correctly\n",
      "Seq #1849 encoded correctly\n",
      "Seq #1850 encoded correctly\n",
      "Seq #1851 encoded correctly\n",
      "Seq #1852 encoded correctly\n",
      "Seq #1853 encoded correctly\n",
      "Seq #1854 encoded correctly\n",
      "Seq #1855 encoded correctly\n",
      "Seq #1856 encoded correctly\n",
      "Seq #1857 encoded correctly\n",
      "Seq #1858 encoded correctly\n",
      "Seq #1859 encoded correctly\n",
      "Seq #1860 encoded correctly\n",
      "Seq #1861 encoded correctly\n",
      "Seq #1862 encoded correctly\n",
      "Seq #1863 encoded correctly\n",
      "Seq #1864 encoded correctly\n",
      "Seq #1865 encoded correctly\n",
      "Seq #1866 encoded correctly\n",
      "Seq #1867 encoded correctly\n",
      "Seq #1868 encoded correctly\n",
      "Seq #1869 encoded correctly\n",
      "Seq #1870 encoded correctly\n",
      "Seq #1871 encoded correctly\n",
      "Seq #1872 encoded correctly\n",
      "Seq #1873 encoded correctly\n",
      "Seq #1874 encoded correctly\n",
      "Seq #1875 encoded correctly\n",
      "Seq #1876 encoded correctly\n",
      "Seq #1877 encoded correctly\n",
      "Seq #1878 encoded correctly\n",
      "Seq #1879 encoded correctly\n",
      "Seq #1880 encoded correctly\n",
      "Seq #1881 encoded correctly\n",
      "Seq #1882 encoded correctly\n",
      "Seq #1883 encoded correctly\n",
      "Seq #1884 encoded correctly\n",
      "Seq #1885 encoded correctly\n",
      "Seq #1886 encoded correctly\n",
      "Seq #1887 encoded correctly\n",
      "Seq #1888 encoded correctly\n",
      "Seq #1889 encoded correctly\n",
      "Seq #1890 encoded correctly\n",
      "Seq #1891 encoded correctly\n",
      "Seq #1892 encoded correctly\n",
      "Seq #1893 encoded correctly\n",
      "Seq #1894 encoded correctly\n",
      "Seq #1895 encoded correctly\n",
      "Seq #1896 encoded correctly\n",
      "Seq #1897 encoded correctly\n",
      "Seq #1898 encoded correctly\n",
      "Seq #1899 encoded correctly\n",
      "Seq #1900 encoded correctly\n",
      "Seq #1901 encoded correctly\n",
      "Seq #1902 encoded correctly\n",
      "Seq #1903 encoded correctly\n",
      "Seq #1904 encoded correctly\n",
      "Seq #1905 encoded correctly\n",
      "Seq #1906 encoded correctly\n",
      "Seq #1907 encoded correctly\n",
      "Seq #1908 encoded correctly\n",
      "Seq #1909 encoded correctly\n",
      "Seq #1910 encoded correctly\n",
      "Seq #1911 encoded correctly\n",
      "Seq #1912 encoded correctly\n",
      "Seq #1913 encoded correctly\n",
      "Seq #1914 encoded correctly\n",
      "Seq #1915 encoded correctly\n",
      "Seq #1916 encoded correctly\n",
      "Seq #1917 encoded correctly\n",
      "Seq #1918 encoded correctly\n",
      "Seq #1919 encoded correctly\n",
      "Seq #1920 encoded correctly\n",
      "Seq #1921 encoded correctly\n",
      "Seq #1922 encoded correctly\n",
      "Seq #1923 encoded correctly\n",
      "Seq #1924 encoded correctly\n",
      "Seq #1925 encoded correctly\n",
      "Seq #1926 encoded correctly\n",
      "Seq #1927 encoded correctly\n",
      "Seq #1928 encoded correctly\n",
      "Seq #1929 encoded correctly\n",
      "Seq #1930 encoded correctly\n",
      "Seq #1931 encoded correctly\n",
      "Seq #1932 encoded correctly\n",
      "Seq #1933 encoded correctly\n",
      "Seq #1934 encoded correctly\n",
      "Seq #1935 encoded correctly\n",
      "Seq #1936 encoded correctly\n",
      "Seq #1937 encoded correctly\n",
      "Seq #1938 encoded correctly\n",
      "Seq #1939 encoded correctly\n",
      "Seq #1940 encoded correctly\n",
      "Seq #1941 encoded correctly\n",
      "Seq #1942 encoded correctly\n",
      "Seq #1943 encoded correctly\n",
      "Seq #1944 encoded correctly\n",
      "Seq #1945 encoded correctly\n",
      "Seq #1946 encoded correctly\n",
      "Seq #1947 encoded correctly\n",
      "Seq #1948 encoded correctly\n",
      "Seq #1949 encoded correctly\n",
      "Seq #1950 encoded correctly\n",
      "Seq #1951 encoded correctly\n",
      "Seq #1952 encoded correctly\n",
      "Seq #1953 encoded correctly\n",
      "Seq #1954 encoded correctly\n",
      "Seq #1955 encoded correctly\n",
      "Seq #1956 encoded correctly\n",
      "Seq #1957 encoded correctly\n",
      "Seq #1958 encoded correctly\n",
      "Seq #1959 encoded correctly\n",
      "Seq #1960 encoded correctly\n",
      "Seq #1961 encoded correctly\n",
      "Seq #1962 encoded correctly\n",
      "Seq #1963 encoded correctly\n",
      "Seq #1964 encoded correctly\n",
      "Seq #1965 encoded correctly\n",
      "Seq #1966 encoded correctly\n",
      "Seq #1967 encoded correctly\n",
      "Seq #1968 encoded correctly\n",
      "Seq #1969 encoded correctly\n",
      "Seq #1970 encoded correctly\n",
      "Seq #1971 encoded correctly\n",
      "Seq #1972 encoded correctly\n",
      "Seq #1973 encoded correctly\n",
      "Seq #1974 encoded correctly\n",
      "Seq #1975 encoded correctly\n",
      "Seq #1976 encoded correctly\n",
      "Seq #1977 encoded correctly\n",
      "Seq #1978 encoded correctly\n",
      "Seq #1979 encoded correctly\n",
      "Seq #1980 encoded correctly\n",
      "Seq #1981 encoded correctly\n",
      "Seq #1982 encoded correctly\n",
      "Seq #1983 encoded correctly\n",
      "Seq #1984 encoded correctly\n",
      "Seq #1985 encoded correctly\n",
      "Seq #1986 encoded correctly\n",
      "Seq #1987 encoded correctly\n",
      "Seq #1988 encoded correctly\n",
      "Seq #1989 encoded correctly\n",
      "Seq #1990 encoded correctly\n",
      "Seq #1991 encoded correctly\n",
      "Seq #1992 encoded correctly\n",
      "Seq #1993 encoded correctly\n",
      "Seq #1994 encoded correctly\n",
      "Seq #1995 encoded correctly\n",
      "Seq #1996 encoded correctly\n",
      "Seq #1997 encoded correctly\n",
      "Seq #1998 encoded correctly\n",
      "Seq #1999 encoded correctly\n",
      "Seq #2000 encoded correctly\n",
      "Seq #2001 encoded correctly\n",
      "Seq #2002 encoded correctly\n",
      "Seq #2003 encoded correctly\n",
      "Seq #2004 encoded correctly\n",
      "Seq #2005 encoded correctly\n",
      "Seq #2006 encoded correctly\n",
      "Seq #2007 encoded correctly\n",
      "Seq #2008 encoded correctly\n",
      "Seq #2009 encoded correctly\n",
      "Seq #2010 encoded correctly\n",
      "Seq #2011 encoded correctly\n",
      "Seq #2012 encoded correctly\n",
      "Seq #2013 encoded correctly\n",
      "Seq #2014 encoded correctly\n",
      "Seq #2015 encoded correctly\n",
      "Seq #2016 encoded correctly\n",
      "Seq #2017 encoded correctly\n",
      "Seq #2018 encoded correctly\n",
      "Seq #2019 encoded correctly\n",
      "Seq #2020 encoded correctly\n",
      "Seq #2021 encoded correctly\n",
      "Seq #2022 encoded correctly\n",
      "Seq #2023 encoded correctly\n",
      "Seq #2024 encoded correctly\n",
      "Seq #2025 encoded correctly\n",
      "Seq #2026 encoded correctly\n",
      "Seq #2027 encoded correctly\n",
      "Seq #2028 encoded correctly\n",
      "Seq #2029 encoded correctly\n",
      "Seq #2030 encoded correctly\n",
      "Seq #2031 encoded correctly\n",
      "Seq #2032 encoded correctly\n",
      "Seq #2033 encoded correctly\n",
      "Seq #2034 encoded correctly\n",
      "Seq #2035 encoded correctly\n",
      "Seq #2036 encoded correctly\n",
      "Seq #2037 encoded correctly\n",
      "Seq #2038 encoded correctly\n",
      "Seq #2039 encoded correctly\n",
      "Seq #2040 encoded correctly\n",
      "Seq #2041 encoded correctly\n",
      "Seq #2042 encoded correctly\n",
      "Seq #2043 encoded correctly\n",
      "Seq #2044 encoded correctly\n",
      "Seq #2045 encoded correctly\n",
      "Seq #2046 encoded correctly\n",
      "Seq #2047 encoded correctly\n",
      "Seq #2048 encoded correctly\n",
      "Seq #2049 encoded correctly\n",
      "Seq #2050 encoded correctly\n",
      "Seq #2051 encoded correctly\n",
      "Seq #2052 encoded correctly\n",
      "Seq #2053 encoded correctly\n",
      "Seq #2054 encoded correctly\n",
      "Seq #2055 encoded correctly\n",
      "Seq #2056 encoded correctly\n",
      "Seq #2057 encoded correctly\n",
      "Seq #2058 encoded correctly\n",
      "Seq #2059 encoded correctly\n",
      "Seq #2060 encoded correctly\n",
      "Seq #2061 encoded correctly\n",
      "Seq #2062 encoded correctly\n",
      "Seq #2063 encoded correctly\n",
      "Seq #2064 encoded correctly\n",
      "Seq #2065 encoded correctly\n",
      "Seq #2066 encoded correctly\n",
      "Seq #2067 encoded correctly\n",
      "Seq #2068 encoded correctly\n",
      "Seq #2069 encoded correctly\n",
      "Seq #2070 encoded correctly\n",
      "Seq #2071 encoded correctly\n",
      "Seq #2072 encoded correctly\n",
      "Seq #2073 encoded correctly\n",
      "Seq #2074 encoded correctly\n",
      "Seq #2075 encoded correctly\n",
      "Seq #2076 encoded correctly\n",
      "Seq #2077 encoded correctly\n",
      "Seq #2078 encoded correctly\n",
      "Seq #2079 encoded correctly\n",
      "Seq #2080 encoded correctly\n",
      "Seq #2081 encoded correctly\n",
      "Seq #2082 encoded correctly\n",
      "Seq #2083 encoded correctly\n",
      "Seq #2084 encoded correctly\n",
      "Seq #2085 encoded correctly\n",
      "Seq #2086 encoded correctly\n",
      "Seq #2087 encoded correctly\n",
      "Seq #2088 encoded correctly\n",
      "Seq #2089 encoded correctly\n",
      "Seq #2090 encoded correctly\n",
      "Seq #2091 encoded correctly\n",
      "Seq #2092 encoded correctly\n",
      "Seq #2093 encoded correctly\n",
      "Seq #2094 encoded correctly\n",
      "Seq #2095 encoded correctly\n",
      "Seq #2096 encoded correctly\n",
      "Seq #2097 encoded correctly\n",
      "Seq #2098 encoded correctly\n",
      "Seq #2099 encoded correctly\n",
      "Seq #2100 encoded correctly\n",
      "Seq #2101 encoded correctly\n",
      "Seq #2102 encoded correctly\n",
      "Seq #2103 encoded correctly\n",
      "Seq #2104 encoded correctly\n",
      "Seq #2105 encoded correctly\n",
      "Seq #2106 encoded correctly\n",
      "Seq #2107 encoded correctly\n",
      "Seq #2108 encoded correctly\n",
      "Seq #2109 encoded correctly\n",
      "Seq #2110 encoded correctly\n",
      "Seq #2111 encoded correctly\n",
      "Seq #2112 encoded correctly\n",
      "Seq #2113 encoded correctly\n",
      "Seq #2114 encoded correctly\n",
      "Seq #2115 encoded correctly\n",
      "Seq #2116 encoded correctly\n",
      "Seq #2117 encoded correctly\n",
      "Seq #2118 encoded correctly\n",
      "Seq #2119 encoded correctly\n",
      "Seq #2120 encoded correctly\n",
      "Seq #2121 encoded correctly\n",
      "Seq #2122 encoded correctly\n",
      "Seq #2123 encoded correctly\n",
      "Seq #2124 encoded correctly\n",
      "Seq #2125 encoded correctly\n",
      "Seq #2126 encoded correctly\n",
      "Seq #2127 encoded correctly\n",
      "Seq #2128 encoded correctly\n",
      "Seq #2129 encoded correctly\n",
      "Seq #2130 encoded correctly\n",
      "Seq #2131 encoded correctly\n",
      "Seq #2132 encoded correctly\n",
      "Seq #2133 encoded correctly\n",
      "Seq #2134 encoded correctly\n",
      "Seq #2135 encoded correctly\n",
      "Seq #2136 encoded correctly\n",
      "Seq #2137 encoded correctly\n",
      "Seq #2138 encoded correctly\n",
      "Seq #2139 encoded correctly\n",
      "Seq #2140 encoded correctly\n",
      "Seq #2141 encoded correctly\n",
      "Seq #2142 encoded correctly\n",
      "Seq #2143 encoded correctly\n",
      "Seq #2144 encoded correctly\n",
      "Seq #2145 encoded correctly\n",
      "Seq #2146 encoded correctly\n",
      "Seq #2147 encoded correctly\n",
      "Seq #2148 encoded correctly\n",
      "Seq #2149 encoded correctly\n",
      "Seq #2150 encoded correctly\n",
      "Seq #2151 encoded correctly\n",
      "Seq #2152 encoded correctly\n",
      "Seq #2153 encoded correctly\n",
      "Seq #2154 encoded correctly\n",
      "Seq #2155 encoded correctly\n",
      "Seq #2156 encoded correctly\n",
      "Seq #2157 encoded correctly\n",
      "Seq #2158 encoded correctly\n",
      "Seq #2159 encoded correctly\n",
      "Seq #2160 encoded correctly\n",
      "Seq #2161 encoded correctly\n",
      "Seq #2162 encoded correctly\n",
      "Seq #2163 encoded correctly\n",
      "Seq #2164 encoded correctly\n",
      "Seq #2165 encoded correctly\n",
      "Seq #2166 encoded correctly\n",
      "Seq #2167 encoded correctly\n",
      "Seq #2168 encoded correctly\n",
      "Seq #2169 encoded correctly\n",
      "Seq #2170 encoded correctly\n",
      "Seq #2171 encoded correctly\n",
      "Seq #2172 encoded correctly\n",
      "Seq #2173 encoded correctly\n",
      "Seq #2174 encoded correctly\n",
      "Seq #2175 encoded correctly\n",
      "Seq #2176 encoded correctly\n",
      "Seq #2177 encoded correctly\n",
      "Seq #2178 encoded correctly\n",
      "Seq #2179 encoded correctly\n",
      "Seq #2180 encoded correctly\n",
      "Seq #2181 encoded correctly\n",
      "Seq #2182 encoded correctly\n",
      "Seq #2183 encoded correctly\n",
      "Seq #2184 encoded correctly\n",
      "Seq #2185 encoded correctly\n",
      "Seq #2186 encoded correctly\n",
      "Seq #2187 encoded correctly\n",
      "Seq #2188 encoded correctly\n",
      "Seq #2189 encoded correctly\n",
      "Seq #2190 encoded correctly\n",
      "Seq #2191 encoded correctly\n",
      "Seq #2192 encoded correctly\n",
      "Seq #2193 encoded correctly\n",
      "Seq #2194 encoded correctly\n",
      "Seq #2195 encoded correctly\n",
      "Seq #2196 encoded correctly\n",
      "Seq #2197 encoded correctly\n",
      "Seq #2198 encoded correctly\n",
      "Seq #2199 encoded correctly\n",
      "Seq #2200 encoded correctly\n",
      "Seq #2201 encoded correctly\n",
      "Seq #2202 encoded correctly\n",
      "Seq #2203 encoded correctly\n",
      "Seq #2204 encoded correctly\n",
      "Seq #2205 encoded correctly\n",
      "Seq #2206 encoded correctly\n",
      "Seq #2207 encoded correctly\n",
      "Seq #2208 encoded correctly\n",
      "Seq #2209 encoded correctly\n",
      "Seq #2210 encoded correctly\n",
      "Seq #2211 encoded correctly\n",
      "Seq #2212 encoded correctly\n",
      "Seq #2213 encoded correctly\n",
      "Seq #2214 encoded correctly\n",
      "Seq #2215 encoded correctly\n",
      "Seq #2216 encoded correctly\n",
      "Seq #2217 encoded correctly\n",
      "Seq #2218 encoded correctly\n",
      "Seq #2219 encoded correctly\n",
      "Seq #2220 encoded correctly\n",
      "Seq #2221 encoded correctly\n",
      "Seq #2222 encoded correctly\n",
      "Seq #2223 encoded correctly\n",
      "Seq #2224 encoded correctly\n",
      "Seq #2225 encoded correctly\n",
      "Seq #2226 encoded correctly\n",
      "Seq #2227 encoded correctly\n",
      "Seq #2228 encoded correctly\n",
      "Seq #2229 encoded correctly\n",
      "Seq #2230 encoded correctly\n",
      "Seq #2231 encoded correctly\n",
      "Seq #2232 encoded correctly\n",
      "Seq #2233 encoded correctly\n",
      "Seq #2234 encoded correctly\n",
      "Seq #2235 encoded correctly\n",
      "Seq #2236 encoded correctly\n",
      "Seq #2237 encoded correctly\n",
      "Seq #2238 encoded correctly\n",
      "Seq #2239 encoded correctly\n",
      "Seq #2240 encoded correctly\n",
      "Seq #2241 encoded correctly\n",
      "Seq #2242 encoded correctly\n",
      "Seq #2243 encoded correctly\n",
      "Seq #2244 encoded correctly\n",
      "Seq #2245 encoded correctly\n",
      "Seq #2246 encoded correctly\n",
      "Seq #2247 encoded correctly\n",
      "Seq #2248 encoded correctly\n",
      "Seq #2249 encoded correctly\n",
      "Seq #2250 encoded correctly\n",
      "Seq #2251 encoded correctly\n",
      "Seq #2252 encoded correctly\n",
      "Seq #2253 encoded correctly\n",
      "Seq #2254 encoded correctly\n",
      "Seq #2255 encoded correctly\n",
      "Seq #2256 encoded correctly\n",
      "Seq #2257 encoded correctly\n",
      "Seq #2258 encoded correctly\n",
      "Seq #2259 encoded correctly\n",
      "Seq #2260 encoded correctly\n",
      "Seq #2261 encoded correctly\n",
      "Seq #2262 encoded correctly\n",
      "Seq #2263 encoded correctly\n",
      "Seq #2264 encoded correctly\n",
      "Seq #2265 encoded correctly\n",
      "Seq #2266 encoded correctly\n",
      "Seq #2267 encoded correctly\n",
      "Seq #2268 encoded correctly\n",
      "Seq #2269 encoded correctly\n",
      "Seq #2270 encoded correctly\n",
      "Seq #2271 encoded correctly\n",
      "Seq #2272 encoded correctly\n",
      "Seq #2273 encoded correctly\n",
      "Seq #2274 encoded correctly\n",
      "Seq #2275 encoded correctly\n",
      "Seq #2276 encoded correctly\n",
      "Seq #2277 encoded correctly\n",
      "Seq #2278 encoded correctly\n",
      "Seq #2279 encoded correctly\n",
      "Seq #2280 encoded correctly\n",
      "Seq #2281 encoded correctly\n",
      "Seq #2282 encoded correctly\n",
      "Seq #2283 encoded correctly\n",
      "Seq #2284 encoded correctly\n",
      "Seq #2285 encoded correctly\n",
      "Seq #2286 encoded correctly\n",
      "Seq #2287 encoded correctly\n",
      "Seq #2288 encoded correctly\n",
      "Seq #2289 encoded correctly\n",
      "Seq #2290 encoded correctly\n",
      "Seq #2291 encoded correctly\n",
      "Seq #2292 encoded correctly\n",
      "Seq #2293 encoded correctly\n",
      "Seq #2294 encoded correctly\n",
      "Seq #2295 encoded correctly\n",
      "Seq #2296 encoded correctly\n",
      "Seq #2297 encoded correctly\n",
      "Seq #2298 encoded correctly\n",
      "Seq #2299 encoded correctly\n",
      "Seq #2300 encoded correctly\n",
      "Seq #2301 encoded correctly\n",
      "Seq #2302 encoded correctly\n",
      "Seq #2303 encoded correctly\n",
      "Seq #2304 encoded correctly\n",
      "Seq #2305 encoded correctly\n",
      "Seq #2306 encoded correctly\n",
      "Seq #2307 encoded correctly\n",
      "Seq #2308 encoded correctly\n",
      "Seq #2309 encoded correctly\n",
      "Seq #2310 encoded correctly\n",
      "Seq #2311 encoded correctly\n",
      "Seq #2312 encoded correctly\n",
      "Seq #2313 encoded correctly\n",
      "Seq #2314 encoded correctly\n",
      "Seq #2315 encoded correctly\n",
      "Seq #2316 encoded correctly\n",
      "Seq #2317 encoded correctly\n",
      "Seq #2318 encoded correctly\n",
      "Seq #2319 encoded correctly\n",
      "Seq #2320 encoded correctly\n",
      "Seq #2321 encoded correctly\n",
      "Seq #2322 encoded correctly\n",
      "Seq #2323 encoded correctly\n",
      "Seq #2324 encoded correctly\n",
      "Seq #2325 encoded correctly\n",
      "Seq #2326 encoded correctly\n",
      "Seq #2327 encoded correctly\n",
      "Seq #2328 encoded correctly\n",
      "Seq #2329 encoded correctly\n",
      "Seq #2330 encoded correctly\n",
      "Seq #2331 encoded correctly\n",
      "Seq #2332 encoded correctly\n",
      "Seq #2333 encoded correctly\n",
      "Seq #2334 encoded correctly\n",
      "Seq #2335 encoded correctly\n",
      "Seq #2336 encoded correctly\n",
      "Seq #2337 encoded correctly\n",
      "Seq #2338 encoded correctly\n",
      "Seq #2339 encoded correctly\n",
      "Seq #2340 encoded correctly\n",
      "Seq #2341 encoded correctly\n",
      "Seq #2342 encoded correctly\n",
      "Seq #2343 encoded correctly\n",
      "Seq #2344 encoded correctly\n",
      "Seq #2345 encoded correctly\n",
      "Seq #2346 encoded correctly\n",
      "Seq #2347 encoded correctly\n",
      "Seq #2348 encoded correctly\n",
      "Seq #2349 encoded correctly\n",
      "Seq #2350 encoded correctly\n",
      "Seq #2351 encoded correctly\n",
      "Seq #2352 encoded correctly\n",
      "Seq #2353 encoded correctly\n",
      "Seq #2354 encoded correctly\n",
      "Seq #2355 encoded correctly\n",
      "Seq #2356 encoded correctly\n",
      "Seq #2357 encoded correctly\n",
      "Seq #2358 encoded correctly\n",
      "Seq #2359 encoded correctly\n",
      "Seq #2360 encoded correctly\n",
      "Seq #2361 encoded correctly\n",
      "Seq #2362 encoded correctly\n",
      "Seq #2363 encoded correctly\n",
      "Seq #2364 encoded correctly\n",
      "Seq #2365 encoded correctly\n",
      "Seq #2366 encoded correctly\n",
      "Seq #2367 encoded correctly\n",
      "Seq #2368 encoded correctly\n",
      "Seq #2369 encoded correctly\n",
      "Seq #2370 encoded correctly\n",
      "Seq #2371 encoded correctly\n",
      "Seq #2372 encoded correctly\n",
      "Seq #2373 encoded correctly\n",
      "Seq #2374 encoded correctly\n",
      "Seq #2375 encoded correctly\n",
      "Seq #2376 encoded correctly\n",
      "Seq #2377 encoded correctly\n",
      "Seq #2378 encoded correctly\n",
      "Seq #2379 encoded correctly\n",
      "Seq #2380 encoded correctly\n",
      "Seq #2381 encoded correctly\n",
      "Seq #2382 encoded correctly\n",
      "Seq #2383 encoded correctly\n",
      "Seq #2384 encoded correctly\n",
      "Seq #2385 encoded correctly\n",
      "Seq #2386 encoded correctly\n",
      "Seq #2387 encoded correctly\n",
      "Seq #2388 encoded correctly\n",
      "Seq #2389 encoded correctly\n",
      "Seq #2390 encoded correctly\n",
      "Seq #2391 encoded correctly\n",
      "Seq #2392 encoded correctly\n",
      "Seq #2393 encoded correctly\n",
      "Seq #2394 encoded correctly\n",
      "Seq #2395 encoded correctly\n",
      "Seq #2396 encoded correctly\n",
      "Seq #2397 encoded correctly\n",
      "Seq #2398 encoded correctly\n",
      "Seq #2399 encoded correctly\n",
      "Seq #2400 encoded correctly\n",
      "Seq #2401 encoded correctly\n",
      "Seq #2402 encoded correctly\n",
      "Seq #2403 encoded correctly\n",
      "Seq #2404 encoded correctly\n",
      "Seq #2405 encoded correctly\n",
      "Seq #2406 encoded correctly\n",
      "Seq #2407 encoded correctly\n",
      "Seq #2408 encoded correctly\n",
      "Seq #2409 encoded correctly\n",
      "Seq #2410 encoded correctly\n",
      "Seq #2411 encoded correctly\n",
      "Seq #2412 encoded correctly\n",
      "Seq #2413 encoded correctly\n",
      "Seq #2414 encoded correctly\n",
      "Seq #2415 encoded correctly\n",
      "Seq #2416 encoded correctly\n",
      "Seq #2417 encoded correctly\n",
      "Seq #2418 encoded correctly\n",
      "Seq #2419 encoded correctly\n",
      "Seq #2420 encoded correctly\n",
      "Seq #2421 encoded correctly\n",
      "Seq #2422 encoded correctly\n",
      "Seq #2423 encoded correctly\n",
      "Seq #2424 encoded correctly\n",
      "Seq #2425 encoded correctly\n",
      "Seq #2426 encoded correctly\n",
      "Seq #2427 encoded correctly\n",
      "Seq #2428 encoded correctly\n",
      "Seq #2429 encoded correctly\n",
      "Seq #2430 encoded correctly\n",
      "Seq #2431 encoded correctly\n",
      "Seq #2432 encoded correctly\n",
      "Seq #2433 encoded correctly\n",
      "Seq #2434 encoded correctly\n",
      "Seq #2435 encoded correctly\n",
      "Seq #2436 encoded correctly\n",
      "Seq #2437 encoded correctly\n",
      "Seq #2438 encoded correctly\n",
      "Seq #2439 encoded correctly\n",
      "Seq #2440 encoded correctly\n",
      "Seq #2441 encoded correctly\n",
      "Seq #2442 encoded correctly\n",
      "Seq #2443 encoded correctly\n",
      "Seq #2444 encoded correctly\n",
      "Seq #2445 encoded correctly\n",
      "Seq #2446 encoded correctly\n",
      "Seq #2447 encoded correctly\n",
      "Seq #2448 encoded correctly\n",
      "Seq #2449 encoded correctly\n",
      "Seq #2450 encoded correctly\n",
      "Seq #2451 encoded correctly\n",
      "Seq #2452 encoded correctly\n",
      "Seq #2453 encoded correctly\n",
      "Seq #2454 encoded correctly\n",
      "Seq #2455 encoded correctly\n",
      "Seq #2456 encoded correctly\n",
      "Seq #2457 encoded correctly\n",
      "Seq #2458 encoded correctly\n",
      "Seq #2459 encoded correctly\n",
      "Seq #2460 encoded correctly\n",
      "Seq #2461 encoded correctly\n",
      "Seq #2462 encoded correctly\n",
      "Seq #2463 encoded correctly\n",
      "Seq #2464 encoded correctly\n",
      "Seq #2465 encoded correctly\n",
      "Seq #2466 encoded correctly\n",
      "Seq #2467 encoded correctly\n",
      "Seq #2468 encoded correctly\n",
      "Seq #2469 encoded correctly\n",
      "Seq #2470 encoded correctly\n",
      "Seq #2471 encoded correctly\n",
      "Seq #2472 encoded correctly\n",
      "Seq #2473 encoded correctly\n",
      "Seq #2474 encoded correctly\n",
      "Seq #2475 encoded correctly\n",
      "Seq #2476 encoded correctly\n",
      "Seq #2477 encoded correctly\n",
      "Seq #2478 encoded correctly\n",
      "Seq #2479 encoded correctly\n",
      "Seq #2480 encoded correctly\n",
      "Seq #2481 encoded correctly\n",
      "Seq #2482 encoded correctly\n",
      "Seq #2483 encoded correctly\n",
      "Seq #2484 encoded correctly\n",
      "Seq #2485 encoded correctly\n",
      "Seq #2486 encoded correctly\n",
      "Seq #2487 encoded correctly\n",
      "Seq #2488 encoded correctly\n",
      "Seq #2489 encoded correctly\n",
      "Seq #2490 encoded correctly\n",
      "Seq #2491 encoded correctly\n",
      "Seq #2492 encoded correctly\n",
      "Seq #2493 encoded correctly\n",
      "Seq #2494 encoded correctly\n",
      "Seq #2495 encoded correctly\n",
      "Seq #2496 encoded correctly\n",
      "Seq #2497 encoded correctly\n",
      "Seq #2498 encoded correctly\n",
      "Seq #2499 encoded correctly\n",
      "Seq #2500 encoded correctly\n",
      "Seq #2501 encoded correctly\n",
      "Seq #2502 encoded correctly\n",
      "Seq #2503 encoded correctly\n",
      "Seq #2504 encoded correctly\n",
      "Seq #2505 encoded correctly\n",
      "Seq #2506 encoded correctly\n",
      "Seq #2507 encoded correctly\n",
      "Seq #2508 encoded correctly\n",
      "Seq #2509 encoded correctly\n",
      "Seq #2510 encoded correctly\n",
      "Seq #2511 encoded correctly\n",
      "Seq #2512 encoded correctly\n",
      "Seq #2513 encoded correctly\n",
      "Seq #2514 encoded correctly\n",
      "Seq #2515 encoded correctly\n",
      "Seq #2516 encoded correctly\n",
      "Seq #2517 encoded correctly\n",
      "Seq #2518 encoded correctly\n",
      "Seq #2519 encoded correctly\n",
      "Seq #2520 encoded correctly\n",
      "Seq #2521 encoded correctly\n",
      "Seq #2522 encoded correctly\n",
      "Seq #2523 encoded correctly\n",
      "Seq #2524 encoded correctly\n",
      "Seq #2525 encoded correctly\n",
      "Seq #2526 encoded correctly\n",
      "Seq #2527 encoded correctly\n",
      "Seq #2528 encoded correctly\n",
      "Seq #2529 encoded correctly\n",
      "Seq #2530 encoded correctly\n",
      "Seq #2531 encoded correctly\n",
      "Seq #2532 encoded correctly\n",
      "Seq #2533 encoded correctly\n",
      "Seq #2534 encoded correctly\n",
      "Seq #2535 encoded correctly\n",
      "Seq #2536 encoded correctly\n",
      "Seq #2537 encoded correctly\n",
      "Seq #2538 encoded correctly\n",
      "Seq #2539 encoded correctly\n",
      "Seq #2540 encoded correctly\n",
      "Seq #2541 encoded correctly\n",
      "Seq #2542 encoded correctly\n",
      "Seq #2543 encoded correctly\n",
      "Seq #2544 encoded correctly\n",
      "Seq #2545 encoded correctly\n",
      "Seq #2546 encoded correctly\n",
      "Seq #2547 encoded correctly\n",
      "Seq #2548 encoded correctly\n",
      "Seq #2549 encoded correctly\n",
      "Seq #2550 encoded correctly\n",
      "Seq #2551 encoded correctly\n",
      "Seq #2552 encoded correctly\n",
      "Seq #2553 encoded correctly\n",
      "Seq #2554 encoded correctly\n",
      "Seq #2555 encoded correctly\n",
      "Seq #2556 encoded correctly\n",
      "Seq #2557 encoded correctly\n",
      "Seq #2558 encoded correctly\n",
      "Seq #2559 encoded correctly\n",
      "Seq #2560 encoded correctly\n",
      "Seq #2561 encoded correctly\n",
      "Seq #2562 encoded correctly\n",
      "Seq #2563 encoded correctly\n",
      "Seq #2564 encoded correctly\n",
      "Seq #2565 encoded correctly\n",
      "Seq #2566 encoded correctly\n",
      "Seq #2567 encoded correctly\n",
      "Seq #2568 encoded correctly\n",
      "Seq #2569 encoded correctly\n",
      "Seq #2570 encoded correctly\n",
      "Seq #2571 encoded correctly\n",
      "Seq #2572 encoded correctly\n",
      "Seq #2573 encoded correctly\n",
      "Seq #2574 encoded correctly\n",
      "Seq #2575 encoded correctly\n",
      "Seq #2576 encoded correctly\n",
      "Seq #2577 encoded correctly\n",
      "Seq #2578 encoded correctly\n",
      "Seq #2579 encoded correctly\n",
      "Seq #2580 encoded correctly\n",
      "Seq #2581 encoded correctly\n",
      "Seq #2582 encoded correctly\n",
      "Seq #2583 encoded correctly\n",
      "Seq #2584 encoded correctly\n",
      "Seq #2585 encoded correctly\n",
      "Seq #2586 encoded correctly\n",
      "Seq #2587 encoded correctly\n",
      "Seq #2588 encoded correctly\n",
      "Seq #2589 encoded correctly\n",
      "Seq #2590 encoded correctly\n",
      "Seq #2591 encoded correctly\n",
      "Seq #2592 encoded correctly\n",
      "Seq #2593 encoded correctly\n",
      "Seq #2594 encoded correctly\n",
      "Seq #2595 encoded correctly\n",
      "Seq #2596 encoded correctly\n",
      "Seq #2597 encoded correctly\n",
      "Seq #2598 encoded correctly\n",
      "Seq #2599 encoded correctly\n",
      "Seq #2600 encoded correctly\n",
      "Seq #2601 encoded correctly\n",
      "Seq #2602 encoded correctly\n",
      "Seq #2603 encoded correctly\n",
      "Seq #2604 encoded correctly\n",
      "Seq #2605 encoded correctly\n",
      "Seq #2606 encoded correctly\n",
      "Seq #2607 encoded correctly\n",
      "Seq #2608 encoded correctly\n",
      "Seq #2609 encoded correctly\n",
      "Seq #2610 encoded correctly\n",
      "Seq #2611 encoded correctly\n",
      "Seq #2612 encoded correctly\n",
      "Seq #2613 encoded correctly\n",
      "Seq #2614 encoded correctly\n",
      "Seq #2615 encoded correctly\n",
      "Seq #2616 encoded correctly\n",
      "Seq #2617 encoded correctly\n",
      "Seq #2618 encoded correctly\n",
      "Seq #2619 encoded correctly\n",
      "Seq #2620 encoded correctly\n",
      "Seq #2621 encoded correctly\n",
      "Seq #2622 encoded correctly\n",
      "Seq #2623 encoded correctly\n",
      "Seq #2624 encoded correctly\n",
      "Seq #2625 encoded correctly\n",
      "Seq #2626 encoded correctly\n",
      "Seq #2627 encoded correctly\n",
      "Seq #2628 encoded correctly\n",
      "Seq #2629 encoded correctly\n",
      "Seq #2630 encoded correctly\n",
      "Seq #2631 encoded correctly\n",
      "Seq #2632 encoded correctly\n",
      "Seq #2633 encoded correctly\n",
      "Seq #2634 encoded correctly\n",
      "Seq #2635 encoded correctly\n",
      "Seq #2636 encoded correctly\n",
      "Seq #2637 encoded correctly\n",
      "Seq #2638 encoded correctly\n",
      "Seq #2639 encoded correctly\n",
      "Seq #2640 encoded correctly\n",
      "Seq #2641 encoded correctly\n",
      "Seq #2642 encoded correctly\n",
      "Seq #2643 encoded correctly\n",
      "Seq #2644 encoded correctly\n",
      "Seq #2645 encoded correctly\n",
      "Seq #2646 encoded correctly\n",
      "Seq #2647 encoded correctly\n",
      "Seq #2648 encoded correctly\n",
      "Seq #2649 encoded correctly\n",
      "Seq #2650 encoded correctly\n",
      "Seq #2651 encoded correctly\n",
      "Seq #2652 encoded correctly\n",
      "Seq #2653 encoded correctly\n",
      "Seq #2654 encoded correctly\n",
      "Seq #2655 encoded correctly\n",
      "Seq #2656 encoded correctly\n",
      "Seq #2657 encoded correctly\n",
      "Seq #2658 encoded correctly\n",
      "Seq #2659 encoded correctly\n",
      "Seq #2660 encoded correctly\n",
      "Seq #2661 encoded correctly\n",
      "Seq #2662 encoded correctly\n",
      "Seq #2663 encoded correctly\n",
      "Seq #2664 encoded correctly\n",
      "Seq #2665 encoded correctly\n",
      "Seq #2666 encoded correctly\n",
      "Seq #2667 encoded correctly\n",
      "Seq #2668 encoded correctly\n",
      "Seq #2669 encoded correctly\n",
      "Seq #2670 encoded correctly\n",
      "Seq #2671 encoded correctly\n",
      "Seq #2672 encoded correctly\n",
      "Seq #2673 encoded correctly\n",
      "Seq #2674 encoded correctly\n",
      "Seq #2675 encoded correctly\n",
      "Seq #2676 encoded correctly\n",
      "Seq #2677 encoded correctly\n",
      "Seq #2678 encoded correctly\n",
      "Seq #2679 encoded correctly\n",
      "Seq #2680 encoded correctly\n",
      "Seq #2681 encoded correctly\n",
      "Seq #2682 encoded correctly\n",
      "Seq #2683 encoded correctly\n",
      "Seq #2684 encoded correctly\n",
      "Seq #2685 encoded correctly\n",
      "Seq #2686 encoded correctly\n",
      "Seq #2687 encoded correctly\n",
      "Seq #2688 encoded correctly\n",
      "Seq #2689 encoded correctly\n",
      "Seq #2690 encoded correctly\n",
      "Seq #2691 encoded correctly\n",
      "Seq #2692 encoded correctly\n",
      "Seq #2693 encoded correctly\n",
      "Seq #2694 encoded correctly\n",
      "Seq #2695 encoded correctly\n",
      "Seq #2696 encoded correctly\n",
      "Seq #2697 encoded correctly\n",
      "Seq #2698 encoded correctly\n",
      "Seq #2699 encoded correctly\n",
      "Seq #2700 encoded correctly\n",
      "Seq #2701 encoded correctly\n",
      "Seq #2702 encoded correctly\n",
      "Seq #2703 encoded correctly\n",
      "Seq #2704 encoded correctly\n",
      "Seq #2705 encoded correctly\n",
      "Seq #2706 encoded correctly\n",
      "Seq #2707 encoded correctly\n",
      "Seq #2708 encoded correctly\n",
      "Seq #2709 encoded correctly\n",
      "Seq #2710 encoded correctly\n",
      "Seq #2711 encoded correctly\n",
      "Seq #2712 encoded correctly\n",
      "Seq #2713 encoded correctly\n",
      "Seq #2714 encoded correctly\n",
      "Seq #2715 encoded correctly\n",
      "Seq #2716 encoded correctly\n",
      "Seq #2717 encoded correctly\n",
      "Seq #2718 encoded correctly\n",
      "Seq #2719 encoded correctly\n",
      "Seq #2720 encoded correctly\n",
      "Seq #2721 encoded correctly\n",
      "Seq #2722 encoded correctly\n",
      "Seq #2723 encoded correctly\n",
      "Seq #2724 encoded correctly\n",
      "Seq #2725 encoded correctly\n",
      "Seq #2726 encoded correctly\n",
      "Seq #2727 encoded correctly\n",
      "Seq #2728 encoded correctly\n",
      "Seq #2729 encoded correctly\n",
      "Seq #2730 encoded correctly\n",
      "Seq #2731 encoded correctly\n",
      "Seq #2732 encoded correctly\n",
      "Seq #2733 encoded correctly\n",
      "Seq #2734 encoded correctly\n",
      "Seq #2735 encoded correctly\n",
      "Seq #2736 encoded correctly\n",
      "Seq #2737 encoded correctly\n",
      "Seq #2738 encoded correctly\n",
      "Seq #2739 encoded correctly\n",
      "Seq #2740 encoded correctly\n",
      "Seq #2741 encoded correctly\n",
      "Seq #2742 encoded correctly\n",
      "Seq #2743 encoded correctly\n",
      "Seq #2744 encoded correctly\n",
      "Seq #2745 encoded correctly\n",
      "Seq #2746 encoded correctly\n",
      "Seq #2747 encoded correctly\n",
      "Seq #2748 encoded correctly\n",
      "Seq #2749 encoded correctly\n",
      "Seq #2750 encoded correctly\n",
      "Seq #2751 encoded correctly\n",
      "Seq #2752 encoded correctly\n",
      "Seq #2753 encoded correctly\n",
      "Seq #2754 encoded correctly\n",
      "Seq #2755 encoded correctly\n",
      "Seq #2756 encoded correctly\n",
      "Seq #2757 encoded correctly\n",
      "Seq #2758 encoded correctly\n",
      "Seq #2759 encoded correctly\n",
      "Seq #2760 encoded correctly\n",
      "Seq #2761 encoded correctly\n",
      "Seq #2762 encoded correctly\n",
      "Seq #2763 encoded correctly\n",
      "Seq #2764 encoded correctly\n",
      "Seq #2765 encoded correctly\n",
      "Seq #2766 encoded correctly\n",
      "Seq #2767 encoded correctly\n",
      "Seq #2768 encoded correctly\n",
      "Seq #2769 encoded correctly\n",
      "Seq #2770 encoded correctly\n",
      "Seq #2771 encoded correctly\n",
      "Seq #2772 encoded correctly\n",
      "Seq #2773 encoded correctly\n",
      "Seq #2774 encoded correctly\n",
      "Seq #2775 encoded correctly\n",
      "Seq #2776 encoded correctly\n",
      "Seq #2777 encoded correctly\n",
      "Seq #2778 encoded correctly\n",
      "Seq #2779 encoded correctly\n",
      "Seq #2780 encoded correctly\n",
      "Seq #2781 encoded correctly\n",
      "Seq #2782 encoded correctly\n",
      "Seq #2783 encoded correctly\n",
      "Seq #2784 encoded correctly\n",
      "Seq #2785 encoded correctly\n",
      "Seq #2786 encoded correctly\n",
      "Seq #2787 encoded correctly\n",
      "Seq #2788 encoded correctly\n",
      "Seq #2789 encoded correctly\n",
      "Seq #2790 encoded correctly\n",
      "Seq #2791 encoded correctly\n",
      "Seq #2792 encoded correctly\n",
      "Seq #2793 encoded correctly\n",
      "Seq #2794 encoded correctly\n",
      "Seq #2795 encoded correctly\n",
      "Seq #2796 encoded correctly\n",
      "Seq #2797 encoded correctly\n",
      "Seq #2798 encoded correctly\n",
      "Seq #2799 encoded correctly\n",
      "Seq #2800 encoded correctly\n",
      "Seq #2801 encoded correctly\n",
      "Seq #2802 encoded correctly\n",
      "Seq #2803 encoded correctly\n",
      "Seq #2804 encoded correctly\n",
      "Seq #2805 encoded correctly\n",
      "Seq #2806 encoded correctly\n",
      "Seq #2807 encoded correctly\n",
      "Seq #2808 encoded correctly\n",
      "Seq #2809 encoded correctly\n",
      "Seq #2810 encoded correctly\n",
      "Seq #2811 encoded correctly\n",
      "Seq #2812 encoded correctly\n",
      "Seq #2813 encoded correctly\n",
      "Seq #2814 encoded correctly\n",
      "Seq #2815 encoded correctly\n",
      "Seq #2816 encoded correctly\n",
      "Seq #2817 encoded correctly\n",
      "Seq #2818 encoded correctly\n",
      "Seq #2819 encoded correctly\n",
      "Seq #2820 encoded correctly\n",
      "Seq #2821 encoded correctly\n",
      "Seq #2822 encoded correctly\n",
      "Seq #2823 encoded correctly\n",
      "Seq #2824 encoded correctly\n",
      "Seq #2825 encoded correctly\n",
      "Seq #2826 encoded correctly\n",
      "Seq #2827 encoded correctly\n",
      "Seq #2828 encoded correctly\n",
      "Seq #2829 encoded correctly\n",
      "Seq #2830 encoded correctly\n",
      "Seq #2831 encoded correctly\n",
      "Seq #2832 encoded correctly\n",
      "Seq #2833 encoded correctly\n",
      "Seq #2834 encoded correctly\n",
      "Seq #2835 encoded correctly\n",
      "Seq #2836 encoded correctly\n",
      "Seq #2837 encoded correctly\n",
      "Seq #2838 encoded correctly\n",
      "Seq #2839 encoded correctly\n",
      "Seq #2840 encoded correctly\n",
      "Seq #2841 encoded correctly\n",
      "Seq #2842 encoded correctly\n",
      "Seq #2843 encoded correctly\n",
      "Seq #2844 encoded correctly\n",
      "Seq #2845 encoded correctly\n",
      "Seq #2846 encoded correctly\n",
      "Seq #2847 encoded correctly\n",
      "Seq #2848 encoded correctly\n",
      "Seq #2849 encoded correctly\n",
      "Seq #2850 encoded correctly\n",
      "Seq #2851 encoded correctly\n",
      "Seq #2852 encoded correctly\n",
      "Seq #2853 encoded correctly\n",
      "Seq #2854 encoded correctly\n",
      "Seq #2855 encoded correctly\n",
      "Seq #2856 encoded correctly\n",
      "Seq #2857 encoded correctly\n",
      "Seq #2858 encoded correctly\n",
      "Seq #2859 encoded correctly\n",
      "Seq #2860 encoded correctly\n",
      "Seq #2861 encoded correctly\n",
      "Seq #2862 encoded correctly\n",
      "Seq #2863 encoded correctly\n",
      "Seq #2864 encoded correctly\n",
      "Seq #2865 encoded correctly\n",
      "Seq #2866 encoded correctly\n",
      "Seq #2867 encoded correctly\n",
      "Seq #2868 encoded correctly\n",
      "Seq #2869 encoded correctly\n",
      "Seq #2870 encoded correctly\n",
      "Seq #2871 encoded correctly\n",
      "Seq #2872 encoded correctly\n",
      "Seq #2873 encoded correctly\n",
      "Seq #2874 encoded correctly\n",
      "Seq #2875 encoded correctly\n",
      "Seq #2876 encoded correctly\n",
      "Seq #2877 encoded correctly\n",
      "Seq #2878 encoded correctly\n",
      "Seq #2879 encoded correctly\n",
      "Seq #2880 encoded correctly\n",
      "Seq #2881 encoded correctly\n",
      "Seq #2882 encoded correctly\n",
      "Seq #2883 encoded correctly\n",
      "Seq #2884 encoded correctly\n",
      "Seq #2885 encoded correctly\n",
      "Seq #2886 encoded correctly\n",
      "Seq #2887 encoded correctly\n",
      "Seq #2888 encoded correctly\n",
      "Seq #2889 encoded correctly\n",
      "Seq #2890 encoded correctly\n",
      "Seq #2891 encoded correctly\n",
      "Seq #2892 encoded correctly\n",
      "Seq #2893 encoded correctly\n",
      "Seq #2894 encoded correctly\n",
      "Seq #2895 encoded correctly\n",
      "Seq #2896 encoded correctly\n",
      "Seq #2897 encoded correctly\n",
      "Seq #2898 encoded correctly\n",
      "Seq #2899 encoded correctly\n",
      "Seq #2900 encoded correctly\n",
      "Seq #2901 encoded correctly\n",
      "Seq #2902 encoded correctly\n",
      "Seq #2903 encoded correctly\n",
      "Seq #2904 encoded correctly\n",
      "Seq #2905 encoded correctly\n",
      "Seq #2906 encoded correctly\n",
      "Seq #2907 encoded correctly\n",
      "Seq #2908 encoded correctly\n",
      "Seq #2909 encoded correctly\n",
      "Seq #2910 encoded correctly\n",
      "Seq #2911 encoded correctly\n",
      "Seq #2912 encoded correctly\n",
      "Seq #2913 encoded correctly\n",
      "Seq #2914 encoded correctly\n",
      "Seq #2915 encoded correctly\n",
      "Seq #2916 encoded correctly\n",
      "Seq #2917 encoded correctly\n",
      "Seq #2918 encoded correctly\n",
      "Seq #2919 encoded correctly\n",
      "Seq #2920 encoded correctly\n",
      "Seq #2921 encoded correctly\n",
      "Seq #2922 encoded correctly\n",
      "Seq #2923 encoded correctly\n",
      "Seq #2924 encoded correctly\n",
      "Seq #2925 encoded correctly\n",
      "Seq #2926 encoded correctly\n",
      "Seq #2927 encoded correctly\n",
      "Seq #2928 encoded correctly\n",
      "Seq #2929 encoded correctly\n",
      "Seq #2930 encoded correctly\n",
      "Seq #2931 encoded correctly\n",
      "Seq #2932 encoded correctly\n",
      "Seq #2933 encoded correctly\n",
      "Seq #2934 encoded correctly\n",
      "Seq #2935 encoded correctly\n",
      "Seq #2936 encoded correctly\n",
      "Seq #2937 encoded correctly\n",
      "Seq #2938 encoded correctly\n",
      "Seq #2939 encoded correctly\n",
      "Seq #2940 encoded correctly\n",
      "Seq #2941 encoded correctly\n",
      "Seq #2942 encoded correctly\n",
      "Seq #2943 encoded correctly\n",
      "Seq #2944 encoded correctly\n",
      "Seq #2945 encoded correctly\n",
      "Seq #2946 encoded correctly\n",
      "Seq #2947 encoded correctly\n",
      "Seq #2948 encoded correctly\n",
      "Seq #2949 encoded correctly\n",
      "Seq #2950 encoded correctly\n",
      "Seq #2951 encoded correctly\n",
      "Seq #2952 encoded correctly\n",
      "Seq #2953 encoded correctly\n",
      "Seq #2954 encoded correctly\n",
      "Seq #2955 encoded correctly\n",
      "Seq #2956 encoded correctly\n",
      "Seq #2957 encoded correctly\n",
      "Seq #2958 encoded correctly\n",
      "Seq #2959 encoded correctly\n",
      "Seq #2960 encoded correctly\n",
      "Seq #2961 encoded correctly\n",
      "Seq #2962 encoded correctly\n",
      "Seq #2963 encoded correctly\n",
      "Seq #2964 encoded correctly\n",
      "Seq #2965 encoded correctly\n",
      "Seq #2966 encoded correctly\n",
      "Seq #2967 encoded correctly\n",
      "Seq #2968 encoded correctly\n",
      "Seq #2969 encoded correctly\n",
      "Seq #2970 encoded correctly\n",
      "Seq #2971 encoded correctly\n",
      "Seq #2972 encoded correctly\n",
      "Seq #2973 encoded correctly\n",
      "Seq #2974 encoded correctly\n",
      "Seq #2975 encoded correctly\n",
      "Seq #2976 encoded correctly\n",
      "Seq #2977 encoded correctly\n",
      "Seq #2978 encoded correctly\n",
      "Seq #2979 encoded correctly\n",
      "Seq #2980 encoded correctly\n",
      "Seq #2981 encoded correctly\n",
      "Seq #2982 encoded correctly\n",
      "Seq #2983 encoded correctly\n",
      "Seq #2984 encoded correctly\n",
      "Seq #2985 encoded correctly\n",
      "Seq #2986 encoded correctly\n",
      "Seq #2987 encoded correctly\n",
      "Seq #2988 encoded correctly\n",
      "Seq #2989 encoded correctly\n",
      "Seq #2990 encoded correctly\n",
      "Seq #2991 encoded correctly\n",
      "Seq #2992 encoded correctly\n",
      "Seq #2993 encoded correctly\n",
      "Seq #2994 encoded correctly\n",
      "Seq #2995 encoded correctly\n",
      "Seq #2996 encoded correctly\n",
      "Seq #2997 encoded correctly\n",
      "Seq #2998 encoded correctly\n",
      "Seq #2999 encoded correctly\n",
      "Seq #3000 encoded correctly\n",
      "Seq #3001 encoded correctly\n",
      "Seq #3002 encoded correctly\n",
      "Seq #3003 encoded correctly\n",
      "Seq #3004 encoded correctly\n",
      "Seq #3005 encoded correctly\n",
      "Seq #3006 encoded correctly\n",
      "Seq #3007 encoded correctly\n",
      "Seq #3008 encoded correctly\n",
      "Seq #3009 encoded correctly\n",
      "Seq #3010 encoded correctly\n",
      "Seq #3011 encoded correctly\n",
      "Seq #3012 encoded correctly\n",
      "Seq #3013 encoded correctly\n",
      "Seq #3014 encoded correctly\n",
      "Seq #3015 encoded correctly\n",
      "Seq #3016 encoded correctly\n",
      "Seq #3017 encoded correctly\n",
      "Seq #3018 encoded correctly\n",
      "Seq #3019 encoded correctly\n",
      "Seq #3020 encoded correctly\n",
      "Seq #3021 encoded correctly\n",
      "Seq #3022 encoded correctly\n",
      "Seq #3023 encoded correctly\n",
      "Seq #3024 encoded correctly\n",
      "Seq #3025 encoded correctly\n",
      "Seq #3026 encoded correctly\n",
      "Seq #3027 encoded correctly\n",
      "Seq #3028 encoded correctly\n",
      "Seq #3029 encoded correctly\n"
     ]
    }
   ],
   "source": [
    "# Sanity check encoding for randomly chosens sequences\n",
    "indeces = np.random.choice(len(X_features), size=len(X_features) // 100)\n",
    "for j, ind in enumerate(indeces):\n",
    "    seq = X_seqs[ind]\n",
    "    one_hot_seq = X_features[ind]\n",
    "    for i, bp in enumerate(seq):\n",
    "        if bp == \"A\":\n",
    "            if (one_hot_seq[i] != [1.0, 0.0, 0.0, 0.0]).all():\n",
    "                print(\"You one hot encoded wrong dummy!\")\n",
    "                print(seq, one_hot_seq)\n",
    "        elif bp == \"C\":\n",
    "            if (one_hot_seq[i] != [0.0, 1.0, 0.0, 0.0]).all():\n",
    "                print(\"You one hot encoded wrong dummy!\")\n",
    "                print(seq, one_hot_seq)\n",
    "        elif bp == \"G\":\n",
    "            if (one_hot_seq[i] != [0.0, 0.0, 1.0, 0.0]).all():\n",
    "                print(\"You one hot encoded wrong dummy!\")\n",
    "                print(seq, one_hot_seq)\n",
    "        elif bp == \"T\":\n",
    "            if (one_hot_seq[i] != [0.0, 0.0, 0.0, 1.0]).all():\n",
    "                print(\"You one hot encoded wrong dummy!\")\n",
    "                print(seq, one_hot_seq)\n",
    "        elif bp == \"N\":\n",
    "            if (one_hot_seq[i] != [0.0, 0.0, 0.0, 0.0]).all():\n",
    "                print(\"You one hot encoded wrong dummy!\")\n",
    "                print(seq, one_hot_seq)\n",
    "        else:\n",
    "            print(bp)\n",
    "    print(\"Seq #{} encoded correctly\".format(j + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "541a76c2-1fe7-4bd0-b043-91196ca1cb83",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:43:25.659244Z",
     "iopub.status.busy": "2021-10-28T17:43:25.659077Z",
     "iopub.status.idle": "2021-10-28T17:43:26.479101Z",
     "shell.execute_reply": "2021-10-28T17:43:26.478754Z",
     "shell.execute_reply.started": "2021-10-28T17:43:25.659226Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(302936, 4, 66)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format in way compatible with PyTorch CNN and RNN\n",
    "X_ohe_seq = np.transpose(X_ohe_seq, axes=(0, 2, 1))\n",
    "X_ohe_seq.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "99cc4277-a9ab-4234-ba6e-7e9ad6bda555",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-10-28T17:43:29.533847Z",
     "iopub.status.busy": "2021-10-28T17:43:29.533684Z",
     "iopub.status.idle": "2021-10-28T17:43:37.284848Z",
     "shell.execute_reply": "2021-10-28T17:43:37.284287Z",
     "shell.execute_reply.started": "2021-10-28T17:43:29.533815Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Save in binary format\n",
    "np.save(\"ohe_seq/{}_X_rev-ohe-seq\".format(PREPROCESS), X_ohe_seq)\n",
    "np.save(\"ohe_seq/{}_X-train-{}_rev-ohe-seq\".format(PREPROCESS, SPLIT), arr=X_ohe_seq[train_index, :, :])\n",
    "np.save(\"ohe_seq/{}_X-test-{}_rev-ohe-seq\".format(PREPROCESS, round(1-SPLIT, 1)), arr=X_ohe_seq[test_index, :, :])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b80d0307-0240-4b78-a28c-c1f8c98978a9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Old test fasta code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "6bfa5b1d-fb70-431d-a095-e0c696d2acc5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:18:35.433231Z",
     "iopub.status.busy": "2021-11-10T00:18:35.432986Z",
     "iopub.status.idle": "2021-11-10T00:18:36.978735Z",
     "shell.execute_reply": "2021-11-10T00:18:36.977904Z",
     "shell.execute_reply.started": "2021-11-10T00:18:35.433201Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Negative seqs\n",
    "X_test_neg = OLS_data.loc[test_index][test_neg_mask][\"SEQUENCE\"].values\n",
    "y_test_neg = y_test[test_neg_mask]\n",
    "id_test_neg = ID[test_index][test_neg_mask]\n",
    "\n",
    "# Positive seqs\n",
    "X_test_pos = OLS_data.loc[test_index][~test_neg_mask][\"SEQUENCE\"].values\n",
    "y_test_pos = y_test[~test_neg_mask]\n",
    "id_test_pos = ID[test_index][~test_neg_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "05ad40c3-a96f-45d5-b249-6a3fe39e9de6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:18:36.980027Z",
     "iopub.status.busy": "2021-11-10T00:18:36.979792Z",
     "iopub.status.idle": "2021-11-10T00:18:38.332349Z",
     "shell.execute_reply": "2021-11-10T00:18:38.331417Z",
     "shell.execute_reply.started": "2021-11-10T00:18:36.979998Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10395,) (10395,)\n",
      "(9360,) (9360,)\n",
      "We good: (19755, 18), (19755,). (19755,)\n"
     ]
    }
   ],
   "source": [
    "# Check\n",
    "print(X_test_neg.shape, y_test_neg.shape)\n",
    "print(X_test_pos.shape, y_test_pos.shape)\n",
    "if (X_test_neg.shape[0] + X_test_pos.shape[0] == X_test.shape[0]):\n",
    "    print(\"We good: {}, {}. {}\".format(X_test.shape, y_test.shape, ID[test_index].shape))\n",
    "else:\n",
    "    print(\"The game is afoot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e13322f-1038-4130-b748-92ddfe5ce780",
   "metadata": {},
   "source": [
    "*Positive test sequences*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "825ec466-9754-41c0-b894-7e76bb1cc0f1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:18:38.334147Z",
     "iopub.status.busy": "2021-11-10T00:18:38.333761Z",
     "iopub.status.idle": "2021-11-10T00:18:39.651785Z",
     "shell.execute_reply": "2021-11-10T00:18:39.651094Z",
     "shell.execute_reply.started": "2021-11-10T00:18:38.334099Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test_file = open(\"fasta/{}_X-test-{}_fasta-pos.fa\".format(PREPROCESS, round(1-SPLIT, 1)), \"w\")\n",
    "for i in range(len(X_test_pos)):\n",
    "    test_file.write(\">\" + id_test_pos[i] + \"\\n\" + X_test_pos[i] + \"\\n\")\n",
    "test_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c7e007e-6039-47a5-a82d-b7b6bb8f8a7b",
   "metadata": {},
   "source": [
    "*Negative test sequences*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ec01b50b-aa99-4c4c-9dcf-26968ab5e0b4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:18:43.189458Z",
     "iopub.status.busy": "2021-11-10T00:18:43.189179Z",
     "iopub.status.idle": "2021-11-10T00:18:44.446567Z",
     "shell.execute_reply": "2021-11-10T00:18:44.445809Z",
     "shell.execute_reply.started": "2021-11-10T00:18:43.189427Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "test_neg_file = open(\"fasta/{}_X-test-{}_fasta-neg.fa\".format(PREPROCESS, round(1-SPLIT, 1)), \"w\")\n",
    "for i in range(len(X_test_neg)):\n",
    "    test_neg_file.write(\">\" + id_test_neg[i] + \"\\n\" + X_test_neg[i] + \"\\n\")\n",
    "test_neg_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "c5ebb360-1ece-4405-840d-7804128468ca",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:19:28.849600Z",
     "iopub.status.busy": "2021-11-10T00:19:28.849339Z",
     "iopub.status.idle": "2021-11-10T00:19:30.859041Z",
     "shell.execute_reply": "2021-11-10T00:19:30.857801Z",
     "shell.execute_reply.started": "2021-11-10T00:19:28.849570Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20790 fasta/0.09-0.4_X-test-0.1_fasta-neg.fa\n"
     ]
    }
   ],
   "source": [
    "!wc -l fasta/0.09-0.4_X-test-0.1_fasta-neg.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "7ddc2f13-0716-493f-a0f7-4c6f8be92e17",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-10T00:19:30.861579Z",
     "iopub.status.busy": "2021-11-10T00:19:30.861126Z",
     "iopub.status.idle": "2021-11-10T00:19:32.256166Z",
     "shell.execute_reply": "2021-11-10T00:19:32.255428Z",
     "shell.execute_reply.started": "2021-11-10T00:19:30.861523Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20790"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(X_test_neg)*2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "915fa1bc-0b13-4f7f-aa2a-54cb44d34f57",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-26T23:19:15.582622Z",
     "iopub.status.busy": "2021-11-26T23:19:15.582328Z",
     "iopub.status.idle": "2021-11-26T23:19:16.868713Z",
     "shell.execute_reply": "2021-11-26T23:19:16.867858Z",
     "shell.execute_reply.started": "2021-11-26T23:19:15.582589Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(208690,) (208690,)\n",
      "(94189,) (94189,)\n",
      "We good: (302879,), (302879,), (302879,)\n"
     ]
    }
   ],
   "source": [
    "# Mask\n",
    "training_neg_mask = (y_training == 0)\n",
    "\n",
    "# Negative seqs\n",
    "X_training_neg = X_seqs[training_index][training_neg_mask]\n",
    "y_training_neg = y_training[training_neg_mask]\n",
    "id_training_neg = ID[training_index][training_neg_mask]\n",
    "\n",
    "# Positive seqs\n",
    "X_training_pos = X_seqs[training_index][~training_neg_mask]\n",
    "y_training_pos = y_training[~training_neg_mask]\n",
    "id_training_pos = ID[training_index][~training_neg_mask]\n",
    "\n",
    "# Check\n",
    "print(X_training_neg.shape, y_training_neg.shape)\n",
    "print(X_training_pos.shape, y_training_pos.shape)\n",
    "if (X_training_neg.shape[0] + X_training_pos.shape[0] == X_seqs[training_index].shape[0]):\n",
    "    print(\"We good: {}, {}, {}\".format(X_seqs[training_index].shape, y_training.shape, ID[training_index].shape))\n",
    "else:\n",
    "    print(\"The game is afoot\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2916c9f0-c531-4083-bac6-2f5a9b17e803",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:03.169479Z",
     "iopub.status.busy": "2021-11-25T02:36:03.169277Z",
     "iopub.status.idle": "2021-11-25T02:36:05.359979Z",
     "shell.execute_reply": "2021-11-25T02:36:05.359255Z",
     "shell.execute_reply.started": "2021-11-25T02:36:03.169452Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "169522\n",
      "169658 fasta/0.09-0.4_X-train-0.9_fasta-pos.fa\n"
     ]
    }
   ],
   "source": [
    "# Save positive sequences in fasta format\n",
    "tr_file = open(\"fasta/{}_X-train-{}_fasta-pos.fa\".format(PREPROCESS, SPLIT), \"w\")\n",
    "for i in range(len(X_train_pos)):\n",
    "    tr_file.write(\">\" + id_train_pos[i] + \"\\n\" + X_train_pos[i] + \"\\n\")\n",
    "tr_file.close()\n",
    "\n",
    "# Should equal below\n",
    "print(len(X_train_pos)*2)\n",
    "\n",
    "# Double check\n",
    "!wc -l fasta/0.09-0.4_X-train-0.9_fasta-pos.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad57394-db84-4a32-9958-20c89f3b0bab",
   "metadata": {},
   "source": [
    "*Negative training sequences*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "b2899c9e-8471-45b6-a562-981753fa43a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:05.361344Z",
     "iopub.status.busy": "2021-11-25T02:36:05.361096Z",
     "iopub.status.idle": "2021-11-25T02:36:08.008848Z",
     "shell.execute_reply": "2021-11-25T02:36:08.008073Z",
     "shell.execute_reply.started": "2021-11-25T02:36:05.361311Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "375660\n",
      "185928 fasta/0.09-0.4_X-train-0.9_fasta-neg.fa\n"
     ]
    }
   ],
   "source": [
    "# Save negative sequences in fasta format\n",
    "tr_neg_file = open(\"fasta/{}_X-train-{}_fasta-neg.fa\".format(PREPROCESS, SPLIT), \"w\")\n",
    "for i in range(len(X_train_neg)):\n",
    "    tr_neg_file.write(\">\" + id_train_neg[i] + \"\\n\" + X_train_neg[i] + \"\\n\")\n",
    "tr_neg_file.close()\n",
    "\n",
    "# Should equal below\n",
    "print(len(X_train_neg)*2)\n",
    "\n",
    "# Double check\n",
    "!wc -l fasta/0.09-0.4_X-train-0.9_fasta-neg.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0248f72-4cee-4e61-b5ca-2d9bf2d4d371",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### <u> **Train** </u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "597000d7-da1e-4827-8198-fcd2aae30e51",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:01.605659Z",
     "iopub.status.busy": "2021-11-25T02:36:01.605288Z",
     "iopub.status.idle": "2021-11-25T02:36:03.168333Z",
     "shell.execute_reply": "2021-11-25T02:36:03.167675Z",
     "shell.execute_reply.started": "2021-11-25T02:36:01.605611Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(187830,) (187830,)\n",
      "(84761,) (84761,)\n",
      "The game is afoot\n"
     ]
    }
   ],
   "source": [
    "# Mask\n",
    "train_neg_mask = (y_train == 0)\n",
    "\n",
    "# Negative seqs\n",
    "X_train_neg = X_seqs[train_index][train_neg_mask]\n",
    "y_train_neg = y_train[train_neg_mask]\n",
    "id_train_neg = ID[train_index][train_neg_mask]\n",
    "\n",
    "# Positive seqs\n",
    "X_train_pos = X_seqs[train_index][~train_neg_mask]\n",
    "y_train_pos = y_train[~train_neg_mask]\n",
    "id_train_pos = ID[train_index][~train_neg_mask]\n",
    "\n",
    "# Check\n",
    "print(X_train_neg.shape, y_train_neg.shape)\n",
    "print(X_train_pos.shape, y_train_pos.shape)\n",
    "if (X_train_neg.shape[0] + X_train_pos.shape[0] == X_train.shape[0]):\n",
    "    print(\"We good: {}, {}, {}\".format(X_train.shape, y_train.shape, ID[train_index].shape))\n",
    "else:\n",
    "    print(\"The game is afoot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f89b145-40fd-420e-84f1-0856d494a96c",
   "metadata": {},
   "source": [
    "*Positive training sequences*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2866c9dc-0c2c-4b51-899d-d54a9182db42",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:03.169479Z",
     "iopub.status.busy": "2021-11-25T02:36:03.169277Z",
     "iopub.status.idle": "2021-11-25T02:36:05.359979Z",
     "shell.execute_reply": "2021-11-25T02:36:05.359255Z",
     "shell.execute_reply.started": "2021-11-25T02:36:03.169452Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "169522\n",
      "169658 fasta/0.09-0.4_X-train-0.9_fasta-pos.fa\n"
     ]
    }
   ],
   "source": [
    "# Save positive sequences in fasta format\n",
    "tr_file = open(\"fasta/{}_X-train-{}_fasta-pos.fa\".format(PREPROCESS, SPLIT), \"w\")\n",
    "for i in range(len(X_train_pos)):\n",
    "    tr_file.write(\">\" + id_train_pos[i] + \"\\n\" + X_train_pos[i] + \"\\n\")\n",
    "tr_file.close()\n",
    "\n",
    "# Should equal below\n",
    "print(len(X_train_pos)*2)\n",
    "\n",
    "# Double check\n",
    "!wc -l fasta/0.09-0.4_X-train-0.9_fasta-pos.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f2c123-b6de-4298-b6bb-832d1f3eed37",
   "metadata": {},
   "source": [
    "*Negative training sequences*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2b09cd80-320d-4363-a3a9-a52fd7addf66",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:05.361344Z",
     "iopub.status.busy": "2021-11-25T02:36:05.361096Z",
     "iopub.status.idle": "2021-11-25T02:36:08.008848Z",
     "shell.execute_reply": "2021-11-25T02:36:08.008073Z",
     "shell.execute_reply.started": "2021-11-25T02:36:05.361311Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "375660\n",
      "185928 fasta/0.09-0.4_X-train-0.9_fasta-neg.fa\n"
     ]
    }
   ],
   "source": [
    "# Save negative sequences in fasta format\n",
    "tr_neg_file = open(\"fasta/{}_X-train-{}_fasta-neg.fa\".format(PREPROCESS, SPLIT), \"w\")\n",
    "for i in range(len(X_train_neg)):\n",
    "    tr_neg_file.write(\">\" + id_train_neg[i] + \"\\n\" + X_train_neg[i] + \"\\n\")\n",
    "tr_neg_file.close()\n",
    "\n",
    "# Should equal below\n",
    "print(len(X_train_neg)*2)\n",
    "\n",
    "# Double check\n",
    "!wc -l fasta/0.09-0.4_X-train-0.9_fasta-neg.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "162abe4c-76d7-4148-ac5d-c8d46dac6f83",
   "metadata": {},
   "source": [
    "*Positive training sequences*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62970426-b226-48c1-ab5a-9f1ec2b71f55",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-05T04:56:26.512653Z",
     "iopub.status.busy": "2021-08-05T04:56:26.512337Z",
     "iopub.status.idle": "2021-08-05T04:56:26.519366Z",
     "shell.execute_reply": "2021-08-05T04:56:26.517654Z",
     "shell.execute_reply.started": "2021-08-05T04:56:26.512618Z"
    },
    "tags": []
   },
   "source": [
    "#### <u> **Test** </u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "30a3d7d2-fcd9-410e-aae0-72dd48ee0bdb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:08.010748Z",
     "iopub.status.busy": "2021-11-25T02:36:08.010393Z",
     "iopub.status.idle": "2021-11-25T02:36:10.406919Z",
     "shell.execute_reply": "2021-11-25T02:36:10.406130Z",
     "shell.execute_reply.started": "2021-11-25T02:36:08.010698Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39510 fasta/0.09-0.4_X-test-0.1_fasta.fa\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "60576"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only need to split for training\n",
    "X_test = X_seqs[test_index]\n",
    "id_test = ID[test_index]\n",
    "\n",
    "test_file = open(\"fasta/{}_X-test-{}_fasta.fa\".format(PREPROCESS, round(1-SPLIT, 1)), \"w\")\n",
    "for i in range(len(X_test)):\n",
    "    test_file.write(\">\" + id_test[i] + \"\\n\" + X_test[i] + \"\\n\")\n",
    "test_file.close()\n",
    "\n",
    "!wc -l fasta/0.09-0.4_X-test-0.1_fasta.fa\n",
    "\n",
    "len(X_test)*2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5372c5e5-89f8-4fc3-8334-336428372745",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### <u> **Hold-out** </u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "60efedf4-89ea-4b8e-ad1c-dce90f0d91a5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-11-25T02:36:10.408791Z",
     "iopub.status.busy": "2021-11-25T02:36:10.408454Z",
     "iopub.status.idle": "2021-11-25T02:36:13.466823Z",
     "shell.execute_reply": "2021-11-25T02:36:13.466045Z",
     "shell.execute_reply.started": "2021-11-25T02:36:10.408741Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wc: fasta/0.09-0.4_X-holdout-0.1_fasta.fa: No such file or directory\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "315842"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only need to split for training\n",
    "X_holdout = X_seqs[holdout_index]\n",
    "id_holdout = ID[holdout_index]\n",
    "\n",
    "test_file = open(\"fasta/{}_X-holdout_fasta.fa\".format(PREPROCESS), \"w\")\n",
    "for i in range(len(X_holdout)):\n",
    "    test_file.write(\">\" + id_holdout[i] + \"\\n\" + X_holdout[i] + \"\\n\")\n",
    "test_file.close()\n",
    "\n",
    "!wc -l fasta/0.09-0.4_X-holdout-0.1_fasta.fa\n",
    "\n",
    "len(X_holdout)*2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 pytorch_dev",
   "language": "python",
   "name": "pytorch_dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
